---
title: "Figures and analyses for evolution of mutation spectra across eukaryotes"
author: "Fabian Ramos-Almodovar"
date: "2025-05-19"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(ggpubr)
library(dplyr)
library(shiny)
library(ggtree)
library(ggtreeExtra)
library(tidytree)
library(tidyverse)
library(ape)
library(vegan)
library(reshape2)
library(qgraph)
library(phylolm)
library(readxl)
library(pheatmap)
library(MCMCglmm)
library(viridis)
library(scico)
library(gridExtra)
library(phytools)
library(cowplot)
library(ggrepel)
library(patchwork)
library(paletteer)
library(forcats)
library(stringr)
library(grid)
library(ggimage)
setwd("~/Downloads/MathiesonLab/Temporary_100Species_Downloads")
#to end knitting early use knitr::knit_exit()
```

## Species label map (to include common names)
```{r, echo=FALSE}
species_label_map <- list("Haplochromini" = "Haplochromini_(Cichlid)",
                          "Haplochromine_cichlid" = "Haplochromini_(Cichlid)",
                          "Labidochromis_vellicans" = "Haplochromini_(Cichlid)",
                          "Sander_lucioperca" = "Sander_lucioperca_(Pikeperch)",
                          "Perca_fluviatilis" = "Perca_fluviatilis_(Perch)",
                          "Seriola_dumerili" = "Seriola_dumerili_(Greater_Amberjack)",
                          "Canis_lupus" = "Canis_lupus_(Wolf_and_Domestic_Dog)",
                          "Canis_lupus_familiaris" = "Canis_lupus_(Wolf_and_Domestic_Dog)",
                          "canis_lupus_familiaris" = "Canis_lupus_(Wolf_and_Domestic_Dog)",
                          "Canis_familiaris" = "Canis_lupus_(Wolf_and_Domestic_Dog)",
                          "Equus_caballus" = "Equus_caballus_(Horse)",
                          "Ovis_orientalis" = "Ovis_orientalis_(Asiatic_Mouflon)",
                          "Ovis_aries" = "Ovis_aries_(Sheep)",
                          "Sus_scrofa" = "Sus_scrofa_(Pig)",
                          "Homo_sapiens" = "Homo_sapiens_(Human)",
                          "Pan_troglodytes" = "Pan_troglodytes_(Chimpanzee)",
                          "Papio_anubis" = "Papio_anubis_(Baboon)",
                          "Anas_platyrhynchos" = "Anas_platyrhynchos_(Duck)",
                          "Gallus_gallus" = "Gallus_gallus_(Chicken)",
                          "Galgal6" = "Gallus_gallus_(Chicken)",
                          "Apis_mellifera" = "Apis_mellifera_(Honey_Bee)",
                          "Brassica_oleracea" = "Brassica_oleracea_(Wild_Cabbage)",
                          "Brassica_oleracea_rerun_for_test" = "Brassica_oleracea_(Wild_Cabbage)",
                          "Theobroma_cacao" = "Theobroma_cacao_(Cacao_Tree)",
                          "Glycine_max" = "Glycine_max_(Soy_Bean)",
                          "Oryza_rufipogon" = "Oryza_rufipogon_(Wild_Rice)",
                          "Sorghum_bicolor" = "Sorghum_bicolor_(Sorghum)",
                          "Sorghum_Bicolor" = "Sorghum_bicolor_(Sorghum)",
                          "Haliaeetus_albicilla" = "Haliaeetus_albicilla_(White-tailed_Eagle)",
                          "Saccharomyces_cerevisiae" = "Saccharomyces_cerevisiae_(Yeast)",
                          "Drosophila_melanogaster" = "Drosophila_melanogaster_(Fruit_Fly)",
                          "Panicum_italica" = "Panicum_italica_(Foxtail_Millet)",
                          "Setaria_italica" = "Panicum_italica_(Foxtail_Millet)",
                          "Setarica_italica" = "Panicum_italica_(Foxtail_Millet)",
                          "Setarica_italica_2" = "Panicum_italica_(Foxtail_Millet)",
                          "Bos_taurus" = "Bos_taurus_(Cow)",
                          "Oryza_sativa" = "Oryza_sativa_(Cultivated_Rice)",
                          "Oryza_sativa_Japonica_Group" = "Oryza_sativa_(Cultivated_Rice)",
                          "Zea_mays" = "Zea_mays_(Corn)",
                          "Caenorhabditis_elegans" = "Caenorhabditis_elegans",
                          "Mus_musculus" = "Mus_musculus_(Mouse)",
                          "Acropora_tenuis" = "Acropora_tenuis_(Reef_Coral)",
                          "Plasmodium_vivax" = "Plasmodium_vivax_(Malaria)",
                          "Acropora_millepora" = "Acropora_millepora_(Stony_Coral)",
                          "Tursiops_truncatus" = "Tursiops_truncatus_(Bottle-nose_Dolphin)",
                          "Delphinus_tropicalis" = "Tursiops_truncatus_(Bottle-nose_Dolphin)",
                          "Tursiops_truncatus" = "Tursiops_truncatus_(Bottle-nose_Dolphin)",
                          "Ursus_arctos" = "Ursus_arctos_(Brown_Bear)",
                          "Mimulus_ringens" = "Mimulus_ringens_(Monkeyflower)",
                          "Chlorocebus_sabaeus" = "Chlorocebus_sabaeus_(Vervet_Monkey)",
                          "Oreochromis_niloticus" = "Oreochromis_niloticus_(Tilapia)",
                          "Salmo_salar" = "Salmo_salar_(Salmon)",
                          "Aegilops_tauschii" = "Aegilops_tauschii_(Wild_Wheat)",
                          "Bos_indicus" = "Bos_indicus_(Zebu_Cattle)",
                          "Prunus_yedoensis" = "Prunus_yedoensis_(King_Cherry_Flower)",
                          "Anser_cygnoides" = "Anser_cygnoides_(Goose)",
                          "Anolis_sagrei" = "Anolis_sagrei_(Brown_Lizard)",
                          "Phaseolus_vulgaris" = "Phaseolus_vulgaris_(Common_Bean)",
                          "Phocoena_phocoena" = "Phocoena_phocoena_(Harbor_Porpoise)",
                          "Drosophila_simulans" = "Drosophila_simulans",
                          "Rattus_norvegicus" = "Rattus_norvegicus_(City_Rat)",
                          "Anopheles_gambiae" = "Anopheles_gambiae_(Malaria_Mosquito)",
                          "Solanum_lycopersicum" = "Solanum_lycopersicum_(Tomato)",
                          "Caenorhabditis_briggsae" = "Caenorhabditis_briggsae",
                          "Labeo_rohita" = "Labeo_rohita_(Carp)",
                          "Schistosoma_sinensium" = "Schistosoma_mansoni_(Blood_Fluke)",
                          "Schistosoma_mansoni" = "Schistosoma_mansoni_(Blood_Fluke)",
                          "Capra_hircus" = "Capra_hircus_(Goat)",
                          "Eucalyptus_pellita" = "Eucalyptus_pellita_(Red_Mahogany)",
                          "Anolis_sagrei" = "Anolis_sagrei_(Brown_Lizard)",
                          "Prunus_persica" = "Prunus_persica_(Peach)",
                          "Camellia_sinensis" = "Camellia_sinensis_(Lincang_Tea)",
                          "Chenopodium_quinoa" = "Chenopodium_quinoa_(Quinoa)",
                          "Capra_aegagrus" = "Capra_aegagrus_(Wild_Goat)",
                          "Camelus_ferus" = "Camelus_ferus_(Camel)",
                          "Camelus_ferrus" = "Camelus_ferus_(Camel)",
                          "Pantherophis_guttatus" = "Pantherophis_guttatus_(Corn_Snake)",
                          "Strongylocentrotus_intermedius" = "Strongylocentrotus_purpuratus_(Sea_Urchin)",
                          "Strongylocentrotus_purpuratus" = "Strongylocentrotus_purpuratus_(Sea_Urchin)",
                          "Oryza_glaberrima" = "Oryza_glaberrima_(African_Rice)",
                          "Vicugna_pacos" = "Vicugna_pacos_(Alpaca)",
                          "Sistrurus_catenatus" = "Sistrurus_catenatus_(Rattlesnake)",
                          "Sistrurus_catenatus_catenatus" = "Sistrurus_catenatus_(Rattlesnake)",
                          "Salmo_trutta" = "Salmo_trutta_(Brown_Trout)",
                          "Salmo_trutta_trutta" = "Salmo_trutta_(Brown_Trout)",
                          "Plasmodium_falciparum" = "Plasmodium_falciparum_(Malaria)",
                          "Loxodonta_africana" = "Loxodonta_africana_(Elephant)",
                          "Sarotherodon_galilaeus_sanagaensis" = "Sarotherodon_galilaeus_sanagaensis_(Mango_Tilapia)",
                          "Caperea_marginata" = "Caperea_marginata_(Pygmy_Whale)",
                          "Conogethes_punctiferalis" = "Conogethes_punctiferalis_(Peach_Moth)",
                          "Spilomelinae" = "Conogethes_punctiferalis_(Peach_Moth)",
                          "Oncorhynchus_nerka" = "Oncorhynchus_nerka_(Sockeye_Salmon)",
                          "Spodoptera_frugiperda" = "Spodoptera_frugiperda_(Fall_Armyworm)",
                          "Varanus_komodoensis" = "Varanus_komodoensis_(Komodo_Dragon)",
                          "Lepus_timidus" = "Lepus_timidus_(Mountain_Hare)",
                          "Cucurbita_pepo" = "Cucurbita_pepo_(Field_Pumpkin)",
                          "Puma_concolor" = "Puma_concolor_(Cougar)",
                          "Mirounga_angustirostris" = "Mirounga_angustirostris_(Elephant_Seal)",
                          "Cannabis_sativa" = "Cannabis_sativa",
                          "Daucus_carota" = "Daucus_carota_(Carrot)",
                          "Bos_mutus" = "Bos_mutus_(Wild_Yak)",
                          "Manis_javanica" = "Manis_javanica_(Malayan_Pangolin)",
                          "Manihot_esculenta" = "Manihot_esculenta_(Cassava)",
                          "Prunus_mume" = "Prunus_mume_(Plum_Blossom)",
                          "Brassica_rapa" = "Brassica_rapa_(Field_Mustard)",
                          "Equus_ferus" = "Equus_ferus_(Wild_Horse)",
                          "Jatropha_curcas" = "Jatropha_curcas_(Barbados_Nut)",
                          "Oryzias_latipes" = "Oryzias_latipes_(Medaka)",
                          "Brassica_napus" = "Brassica_napus_(Oilseed)",
                          "Felis_catus" = "Felis_catus_(Domestic_Cat)",
                          "Populus_trichocarpa" = "Populus_trichocarpa_(Black_Cottonwood)",
                          "Arabis_alpina" = "Arabis_alpina_(Alpine_Rock_Cress)",
                          "Arabidopsis_thaliana" = "Arabidopsis_thaliana",
                          "Hordeum_vulgare" = "Hordeum_vulgare_(Barley)",
                          "Ambystosoma_mexicanum" = "Ambystoma_mexicanum_(Axolotl)",
                          "Ambystoma_mexicanum" = "Ambystoma_mexicanum_(Axolotl)",
                          "Terebelliformia" = "Alvinella_pompejana_(Pompeii_Worm)",
                          "Alvinella_pompejana" = "Alvinella_pompejana_(Pompeii_Worm)",
                          "Vitis_vinifera" = "Vitis_vinifera_(Wine_Grape)",
                          "Equus_asinus" = "Equus_asinus_(Donkey)",
                          "Eurytemora_affinis" = "Eurytemora_affinis_(Copepod)",
                          "Trichaptum" = "Trichaptum_abietinum_(Purple_Tooth_Fungia)",
                          "Larimichthys_polyactis" = "Larimichthys_polyactis_(Yellow_Croaker)",
                          "Geospiza_magnirostris" = "Geospiza_magnirostris_(Large_Ground_Finch)",
                          "Saccharomyces_paradoxus" = "Saccharomyces_paradoxus_(Wild_Yeast)",
                          "Cicer_arietinum" = "Cicer_arietinum_(Chickpea)",
                          "Orcinus_orca" = "Orcinus_orca_(Orca)",
                          "Peromyscus_maniculatus" = "Peromyscus_maniculatus_(Prairie_Deer_Mouse)",
                          "Danio_rerio" = "Danio_rerio_(Zebrafish)",
                          "Clupea_harengus" = "Clupea_harengus_(Atlantic_Herring)",
                          "Bison_bison" = "Bison_bison_(American_Bison)",
                          "Gouania" = "Gouania_(Clingfish)",
                          "Gouania_willdenowi" = "Gouania_(Clingfish)",
                          "Triticum_aestivus" = "Triticum_aestivum_(Common_Wheat)",
                          "Triticum_aestivum" = "Triticum_aestivum_(Common_Wheat)",
                          "Triticum_carthlicum" = "Triticum_aestivum_(Common_Wheat)",
                          "Oncorhynchus_anatinus" = "Ornithorhynchus_anatinus_(Platypus)",
                          "Ornithorhynchus_anatinus" = "Ornithorhynchus_anatinus_(Platypus)",
                          "Ailuropoda_melanoleuca" = "Ailuropoda_melanoleuca_(Giant_Panda)",
                          "Meganyctiphanes_norvegica" = "Meganyctiphanes_norvegica_(Krill)",
                          "Aspergillus_flavus" = "Aspergillus_flavus")

species_labels <- c("Haplochromini_(Cichlid)",
                    "Sander_lucioperca_(Pikeperch)",
                    "Perca_fluviatilis_(Perch)",
                    "Seriola_dumerili_(Greater_Amberjack)",
                    "Canis_lupus_(Wolf_and_Domestic_Dog)",
                    "Equus_caballus_(Horse)",
                    "Ovis_orientalis_(Asiatic_Mouflon)",
                    "Ovis_aries_(Sheep)",
                    "Sus_scrofa_(Pig)",
                    "Bos_taurus_(Cow)",
                    "Homo_sapiens_(Human)",
                    "Pan_troglodytes_(Chimpanzee)",
                    "Mus_musculus_(Mouse)",
                    "Papio_anubis_(Baboon)",
                    "Anas_platyrhynchos_(Duck)",
                    "Haliaeetus_albicilla_(White-tailed_Eagle)",
                    "Gallus_gallus_(Chicken)",
                    "Apis_mellifera_(Honey_Bee)",
                    "Drosophila_melanogaster_(Fruit_Fly)",
                    "Caenorhabditis_elegans",
                    "Brassica_oleracea_(Wild_Cabbage)",
                    "Theobroma_cacao_(Cacao_Tree)",
                    "Glycine_max_(Soy_Bean)",
                    "Oryza_rufipogon_(Wild_Rice)",
                    "Sorghum_bicolor_(Sorghum)",
                    "Panicum_italica_(Foxtail_Millet)",
                    "Oryza_sativa_(Cultivated_Rice)",
                    "Zea_mays_(Corn)",
                    "Saccharomyces_cerevisiae_(Yeast)",
                    "Acropora_tenuis_(Reef_Coral)",
                    "Plasmodium_vivax_(Malaria)",
                    "Acropora_millepora_(Stony_Coral)",
                    "Tursiops_truncatus_(Bottle-nose_Dolphin)",
                    "Ursus_arctos_(Brown_Bear)",
                    "Mimulus_ringens_(Monkeyflower)",
                    "Chlorocebus_sabaeus_(Vervet_Monkey)",
                    "Oreochromis_niloticus_(Tilapia)",
                    "Salmo_salar_(Salmon)",
                    "Aegilops_tauschii_(Wild_Wheat)",
                    "Bos_indicus_(Zebu_Cattle)",
                    "Prunus_yedoensis_(King_Cherry_Flower)",
                    "Anser_cygnoides_(Goose)",
                    "Anolis_sagrei_(Brown_Lizard)",
                    "Phaseolus_vulgaris_(Common_Bean)",
                    "Phocoena_phocoena_(Harbor_Porpoise)",
                    "Drosophila_simulans",
                    "Rattus_norvegicus_(City_Rat)",
                    "Anopheles_gambiae_(Malaria_Mosquito)",
                    "Solanum_lycopersicum_(Tomato)",
                    "Caenorhabditis_briggsae",
                    "Labeo_rohita_(Carp)",
                    "Schistosoma_mansoni_(Blood_Fluke)",
                    "Capra_hircus_(Goat)",
                    "Eucalyptus_pellita_(Red_Mahogany)",
                    "Prunus_persica_(Peach)",
                    "Camellia_sinensis_(Lincang_Tea)",
                    "Chenopodium_quinoa_(Quinoa)",
                    "Capra_aegagrus_(Wild_Goat)",
                    "Camelus_ferus_(Camel)",
                    "Pantherophis_guttatus_(Corn_Snake)",
                    "Strongylocentrotus_purpuratus_(Sea_Urchin)",
                    "Oryza_glaberrima_(African_Rice)",
                    "Vicugna_pacos_(Alpaca)",
                    "Sistrurus_catenatus_(Rattlesnake)",
                    "Salmo_trutta_(Brown_Trout)",
                    "Plasmodium_falciparum_(Malaria)",
                    "Loxodonta_africana_(Elephant)",
                    "Sarotherodon_galilaeus_sanagaensis_(Mango_Tilapia)",
                    "Caperea_marginata_(Pygmy_Whale)",
                    "Conogethes_punctiferalis_(Peach_Moth)",
                    "Oncorhynchus_nerka_(Sockeye_Salmon)",
                    "Spodoptera_frugiperda_(Fall_Armyworm)",
                    "Varanus_komodoensis_(Komodo_Dragon)",
                    "Lepus_timidus_(Mountain_Hare)",
                    "Cucurbita_pepo_(Field_Pumpkin)",
                    "Puma_concolor_(Cougar)",
                    "Mirounga_angustirostris_(Elephant_Seal)",
                    "Cannabis_sativa",
                    "Daucus_carota_(Carrot)",
                    "Bos_mutus_(Wild_Yak)",
                    "Manis_javanica_(Malayan_Pangolin)",
                    "Manihot_esculenta_(Cassava)",
                    "Prunus_mume_(Plum_Blossom)",
                    "Brassica_rapa_(Field_Mustard)",
                    "Equus_ferus_(Wild_Horse)",
                    "Jatropha_curcas_(Barbados_Nut)",
                    "Oryzias_latipes_(Medaka)",
                    "Brassica_napus_(Oilseed)",
                    "Felis_catus_(Domestic_Cat)",
                    "Populus_trichocarpa_(Black_Cottonwood)",
                    "Arabis_alpina_(Alpine_Rock_Cress)",
                    "Arabidopsis_thaliana",
                    "Hordeum_vulgare_(Barley)",
                    "Ambystoma_mexicanum_(Axolotl)",
                    "Alvinella_pompejana_(Pompeii_Worm)",
                    "Vitis_vinifera_(Wine_Grape)",
                    "Equus_asinus_(Donkey)",
                    "Eurytemora_affinis_(Copepod)",
                    "Trichaptum_abietinum_(Purple_Tooth_Fungia)",
                    "Larimichthys_polyactis_(Yellow_Croaker)",
                    "Geospiza_magnirostris_(Large_Ground_Finch)",
                    "Saccharomyces_paradoxus_(Wild_Yeast)",
                    "Cicer_arietinum_(Chickpea)",
                    "Orcinus_orca_(Orca)",
                    "Peromyscus_maniculatus_(Prairie_Deer_Mouse)",
                    "Danio_rerio_(Zebrafish)",
                    "Clupea_harengus_(Atlantic_Herring)",
                    "Bison_bison_(American_Bison)",
                    "Gouania_(Clingfish)",
                    "Triticum_aestivum_(Common_Wheat)",
                    "Ornithorhynchus_anatinus_(Platypus)",
                    "Ailuropoda_melanoleuca_(Giant_Panda)",
                    "Meganyctiphanes_norvegica_(Krill)",
                    "Aspergillus_flavus")
species_broad_class <- c("Osteichthyes",
                         "Osteichthyes",
                         "Osteichthyes",
                         "Osteichthyes",
                         "Mammalia",
                         "Mammalia",
                         "Mammalia",
                         "Mammalia",
                         "Mammalia",
                         "Mammalia",
                         "Mammalia",
                         "Mammalia",
                         "Mammalia",
                         "Mammalia",
                         "Aves",
                         "Aves",
                         "Aves",
                         "Invertebrata",
                         "Invertebrata",
                         "Invertebrata",
                         "Plantae",
                         "Plantae",
                         "Plantae",
                         "Plantae",
                         "Plantae",
                         "Plantae",
                         "Plantae",
                         "Plantae",
                         "Fungia",
                         "Cnidaria",
                         "Protozoa",
                         "Cnidaria",
                         "Mammalia",
                         "Mammalia",
                         "Plantae",
                         "Mammalia",
                         "Osteichthyes",
                         "Osteichthyes",
                         "Plantae",
                         "Mammalia",
                         "Plantae",
                         "Aves",
                         "Reptilia",
                         "Plantae",
                         "Mammalia",
                         "Invertebrata",
                         "Mammalia",
                         "Invertebrata",
                         "Plantae",
                         "Invertebrata",
                         "Osteichthyes",
                         "Invertebrata",
                         "Mammalia",
                         "Plantae",
                         "Plantae",
                         "Plantae",
                         "Plantae",
                         "Mammalia",
                         "Mammalia",
                         "Reptilia",
                         "Echinodermata",
                         "Plantae",
                         "Mammalia",
                         "Reptilia",
                         "Osteichthyes",
                         "Protozoa",
                         "Mammalia",
                         "Osteichthyes",
                         "Mammalia",
                         "Invertebrata",
                         "Osteichthyes",
                         "Invertebrata",
                         "Reptilia",
                         "Mammalia",
                         "Plantae",
                         "Mammalia",
                         "Mammalia",
                         "Plantae",
                         "Plantae",
                         "Mammalia",
                         "Mammalia",
                         "Plantae",
                         "Plantae",
                         "Plantae",
                         "Mammalia",
                         "Plantae",
                         "Osteichthyes",
                         "Plantae",
                         "Mammalia",
                         "Plantae",
                         "Plantae",
                         "Plantae",
                         "Plantae",
                         "Amphibia",
                         "Invertebrata",
                         "Plantae",
                         "Mammalia",
                         "Invertebrata",
                         "Fungia",
                         "Osteichthyes",
                         "Aves",
                         "Fungia",
                         "Plantae",
                         "Mammalia",
                         "Mammalia",
                         "Osteichthyes",
                         "Osteichthyes",
                         "Mammalia",
                         "Osteichthyes",
                         "Plantae",
                         "Mammalia",
                         "Mammalia",
                         "Invertebrata",
                         "Fungia")

classified_species <- data.frame(species_labels,species_broad_class)
colnames(classified_species) <- c("Species", "Class")
rownames(classified_species) <- classified_species$Species

ColoredClades <- c("Aves" = "#F9C000FF",
                                "Cnidaria" = "#FF4D6FFF",
                                "Mammalia" = "#C8350DFF",
                                "Reptilia" = "#5D7298FF",
                                "Protozoa" = "#BD777AFF",
                                "Plantae" = "#86AD34FF",
                                "Osteichthyes" = "#579EA4FF",
                                "Fungia" = "#DF7713FF",
                                "Invertebrata" = "#7E1A2FFF",
                                "Echinodermata" = "#81B28DFF",
                                "Amphibia" = "#2D2651FF")
```

#Load tree!!
```{r}
tree <- read.tree("MutationEvolution_108Species.nwk")
#####

tip_data <- data.frame(
  label = classified_species[tree$tip.label,]$Species,
  category = classified_species[tree$tip.label,]$Class,
  imagePlatypus = NA,
  imageAxolotl = NA,
  imageCougar = NA,
  imageKrill = NA,
  imageArabidopsis = NA,
  imageYeast = NA,
  imageCoral = NA,
  imageElephant = NA,
  imageBee = NA,
  imageCorn = NA,
  imageSalmon = NA,
  imageMedaka = NA,
  imagePorpoise = NA,
  imageChicken = NA,
  imageAmberjack = NA,
  imageRattlesnake = NA,
  imageFalciparum = NA,
  imagePompeii = NA,
  imageUrchin = NA,
  imageBear = NA,
  imageBison = NA,
  imageRice = NA,
  imagePumpkin = NA,
  imageEucalyptus = NA,
  imageBarley = NA
)
# adding some phylopic images for specific species
  # image url
tip_data[tip_data$label=="Ornithorhynchus_anatinus_(Platypus)",]$imagePlatypus <- "https://images.phylopic.org/images/61932f57-1fd2-49d9-bb86-042d6005581a/raster/1536x727.png?v=1933f4c556b"
tip_data[tip_data$label=="Ambystoma_mexicanum_(Axolotl)",]$imageAxolotl <- "https://images.phylopic.org/images/575eaa51-6c9b-4d36-9881-b8463c68ebbc/raster/1536x787.png?v=15af88502ad"
tip_data[tip_data$label=="Puma_concolor_(Cougar)",]$imageCougar <- "https://images.phylopic.org/images/ba8012a3-cfc9-4403-b4bb-1959988766ca/raster/488x266.png?v=1776a708772"
tip_data[tip_data$label=="Meganyctiphanes_norvegica_(Krill)",]$imageKrill <- "https://images.phylopic.org/images/44a3628d-aafd-45cc-97a6-1cb74bd43dec/raster/1536x542.png?v=18d2eeb25db"
tip_data[tip_data$label=="Arabidopsis_thaliana",]$imageArabidopsis <- "https://images.phylopic.org/images/7fb6a9de-0024-4e45-bb8d-1cc32fb2470f/raster/1536x1384.png?v=1577cf19483"
tip_data[tip_data$label=="Saccharomyces_cerevisiae_(Yeast)",]$imageYeast <- "https://images.phylopic.org/images/35db2572-a27d-4f83-9b45-11195d6fd5af/raster/1536x1412.png?v=158fc047eda"
tip_data[tip_data$label=="Acropora_millepora_(Stony_Coral)",]$imageCoral <- "https://images.phylopic.org/images/f6a243aa-5cb1-41a2-a52c-c8d4c4300104/raster/1536x922.png?v=18dbbefb8d1"
tip_data[tip_data$label=="Loxodonta_africana_(Elephant)",]$imageElephant <- "https://images.phylopic.org/images/2a070808-0874-409d-81c0-d0c1995e54bb/raster/1536x1434.png?v=178987a6aa6"
tip_data[tip_data$label=="Apis_mellifera_(Honey_Bee)",]$imageBee <- "https://images.phylopic.org/images/d98ca79d-443e-466b-b78e-c8a6f8ecfa40/raster/1536x1009.png?v=18810533287"
tip_data[tip_data$label=="Zea_mays_(Corn)",]$imageCorn <- "https://images.phylopic.org/images/89dc273a-1a26-4abc-819a-0ba8b28ff111/raster/1246x1536.png?v=18f553279fb"
tip_data[tip_data$label=="Salmo_salar_(Salmon)",]$imageSalmon <- "https://images.phylopic.org/images/5034d9bf-a395-433d-b967-e335fd7185ca/raster/1536x529.png?v=1837e797724"
tip_data[tip_data$label=="Oryzias_latipes_(Medaka)",]$imageMedaka <- "https://images.phylopic.org/images/b7ae3bbf-c0a1-43bf-80ac-babf90554073/raster/1172x338.png?v=15634f376f3"
tip_data[tip_data$label=="Phocoena_phocoena_(Harbor_Porpoise)",]$imagePorpoise <- "https://images.phylopic.org/images/970f7e8d-a823-45b9-85b9-767704d0c13f/raster/1536x521.png?v=1352c1bce1f"
tip_data[tip_data$label=="Gallus_gallus_(Chicken)",]$imageChicken <- "https://images.phylopic.org/images/aff847b0-ecbd-4d41-98ce-665921a6d96e/raster/1080x1536.png?v=13571db5163"
tip_data[tip_data$label=="Seriola_dumerili_(Greater_Amberjack)",]$imageAmberjack <- "https://images.phylopic.org/images/d95d697f-e9a8-4f6c-a746-a7bb76cb98db/raster/1536x519.png?v=18bc18db182"
tip_data[tip_data$label=="Sistrurus_catenatus_(Rattlesnake)",]$imageRattlesnake <- "https://images.phylopic.org/images/5f0a231c-f254-468b-a8c3-83a7ae6171b9/raster/1536x1086.png?v=13faa3ad0ae"
tip_data[tip_data$label=="Plasmodium_falciparum_(Malaria)",]$imageFalciparum <- "https://images.phylopic.org/images/e6014244-4dd5-4785-bf2e-c67dc4d05ca8/raster/1536x1536.png?v=170850a964d"
tip_data[tip_data$label=="Alvinella_pompejana_(Pompeii_Worm)",]$imagePompeii <- "https://images.phylopic.org/images/e7d59ad8-887c-4017-bad5-6e2b5b167dc6/raster/1536x802.png?v=183b9e71a4f"
tip_data[tip_data$label=="Strongylocentrotus_purpuratus_(Sea_Urchin)",]$imageUrchin <- "https://images.phylopic.org/images/06119510-7641-4a0b-8585-93ccb5ca9447/raster/1192x1019.png?v=1893ca617ca"
tip_data[tip_data$label=="Ursus_arctos_(Brown_Bear)",]$imageBear <- "https://images.phylopic.org/images/5629fae2-1da6-479d-bfeb-08a59a58121a/raster/1536x857.png?v=178afe7322c"
tip_data[tip_data$label=="Bison_bison_(American_Bison)",]$imageBison <- "https://images.phylopic.org/images/30afd55c-df49-44d9-bb60-5f8712fdde7d/raster/1536x1000.png?v=17fe450bb86"
tip_data[tip_data$label=="Oryza_sativa_(Cultivated_Rice)",]$imageRice <- "https://images.phylopic.org/images/461f7280-3636-42c0-98fd-4fca668460c5/raster/1536x1230.png?v=1708508f129"
tip_data[tip_data$label=="Cucurbita_pepo_(Field_Pumpkin)",]$imagePumpkin <- "https://images.phylopic.org/images/5812f7a2-5be6-461c-b048-e9352a70ebdf/raster/1148x1536.png?v=1425be153d6"
tip_data[tip_data$label=="Eucalyptus_pellita_(Red_Mahogany)",]$imageEucalyptus <- "https://images.phylopic.org/images/d739ba86-4b2a-4dc1-918e-76d0283c8089/raster/1223x1536.png?v=18d3bcf41cb"
tip_data[tip_data$label=="Hordeum_vulgare_(Barley)",]$imageBarley <- "https://images.phylopic.org/images/c9d31948-819e-401a-abe5-8a4268421b96/raster/692x1536.png?v=19156038f61"

# Plot the tree with colored shapes at the tips
treeForPowerpoint <- ggtree(tree, layout = "rectangular", show.tip.label = FALSE, color="black", lwd=1.5) %<+% tip_data + #  open.angle = 180
  geom_tippoint(aes(color = category), size = 3, shape= 15, offset = 0.5) +
    scale_color_manual(values = ColoredClades, na.translate = FALSE) +
  theme(legend.position = c(0.8,0.80),
        legend.direction = "horizontal",
        legend.title = element_blank(),
        legend.text = element_text(size=12, color="black"),
        legend.background = element_rect(fill="transparent", color="transparent"),
        legend.frame = element_rect(fill="transparent", color="transparent"),
        plot.margin = margin(0,0,0,0)) +
  guides(color = guide_legend(ncol = 5)) +
  geom_image(aes(image=imagePlatypus, colour = category), size=0.23, nudge_x = -200, nudge_y = 3) +
  geom_image(aes(image=imageCougar, colour = category), size=0.22, nudge_x = -200, nudge_y = 3) +
  geom_image(aes(image=imageAxolotl, colour = category), size=0.22, nudge_x = -220, nudge_y = -0.5) +
  geom_image(aes(image=imageKrill, colour = category), size=0.2, nudge_x = -180, nudge_y = 0) +
  geom_image(aes(image=imageArabidopsis, colour = category), size=0.15, nudge_x = -320, nudge_y = 0) +
  geom_image(aes(image=imageYeast, colour = category), size=0.1, nudge_x = -275, nudge_y = -3) +
  geom_image(aes(image=imageCoral, colour = category), size=0.21, nudge_x = -190, nudge_y = 0) +
  geom_image(aes(image=imageElephant, colour = category), size=0.2, nudge_x = -280, nudge_y = 9.5) +
  geom_image(aes(image=imageBee, colour = category), size=0.13, nudge_x = -310, nudge_y = 2.2) +
  geom_image(aes(image=imageCorn, colour = category), size=0.13, nudge_x = -260, nudge_y = 2) +
  geom_image(aes(image=imageSalmon, colour = category), size=0.2, nudge_x = -200, nudge_y = -2.4) +
  geom_image(aes(image=imageMedaka, colour = category), size=0.2, nudge_x = -130, nudge_y = -3) +
  geom_image(aes(image=imagePorpoise, colour = category), size=0.22, nudge_x = -150, nudge_y = 4) +
  geom_image(aes(image=imageChicken, colour = category), size=0.12, nudge_x = -280, nudge_y = 1) +
  geom_image(aes(image=imageAmberjack, colour = category), size=0.16, nudge_x = -300, nudge_y = -0.5) +
  geom_image(aes(image=imageRattlesnake, colour = category), size=0.16, nudge_x = -300, nudge_y = 0) +
  geom_image(aes(image=imageFalciparum, colour = category), size=0.12, nudge_x = -230, nudge_y = 1) +
  geom_image(aes(image=imagePompeii, colour = category), size=0.16, nudge_x = -320, nudge_y = 2) +
  geom_image(aes(image=imageUrchin, colour = category), size=0.14, nudge_x = -340, nudge_y = 1) +
  geom_image(aes(image=imageBear, colour = category), size=0.2, nudge_x = -350, nudge_y = 3.9) +
  geom_image(aes(image=imageBison, colour = category), size=0.2, nudge_x = -370, nudge_y = 3) +
  geom_image(aes(image=imageRice, colour = category), size=0.2, nudge_x = -250, nudge_y = 9.6) +
  geom_image(aes(image=imagePumpkin, colour = category), size=0.12, nudge_x = -265, nudge_y = 1) +
  geom_image(aes(image=imageEucalyptus, colour = category), size=0.2, nudge_x = -350, nudge_y = 7) +
  geom_image(aes(image=imageBarley, colour = category), size=0.12, nudge_x = -350, nudge_y = 2) +
  coord_flip() +
  scale_x_reverse() +
  expand_limits(x = 2100)

pdf("~/Downloads/MathiesonLab/100SpeciesPaperFigures/treeForPowerPoint.pdf", 12, 4)
treeForPowerpoint
dev.off()
```
## Master table summary
```{r}
Master_table <- read_xlsx('MutationRate_Species_Analyses_MasterTable.xlsx')

Master_table$Species <- sapply(Master_table$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})

summary_table <- data.frame(Master_table$Species, Master_table$Class,
                            log10(as.numeric(Master_table$NonGenic_SNPs)),
                            log10(as.numeric(Master_table$Sample_Size)),
                            as.numeric(Master_table$GC_Percent_at_NonGenicSites),
                            as.numeric(Master_table$Average_CpG_Methylation_perc),
                            as.numeric(Master_table$thermy),
                            as.numeric(Master_table$body_temp),
                            as.numeric(Master_table$Average_CHG_Methylation_perc),
                            as.numeric(Master_table$Contig_N50),
                            as.numeric(Master_table$Contig_L50),
                            as.numeric(Master_table$Average_CpH_or_CHHinPlants_Methylation_perc))
colnames(summary_table) <- c("Species", "Clade", "Non-Genic SNPs (log10)", "No. Individuals (log10)", "Non-Genic GC Content", "Average CpG Methylation", "Thermy", "Body Temperature", "Average CHG Methylation", "ContigN50", "ContigL50", "Average CHH Methylation")

Species <- sort(tree$tip.label)

rownames(summary_table) <- summary_table$Species
merged_summary_table <- summary_table[Species,]

CpG_data <- read.delim('CpG_Phis.txt', sep = "\t", header = FALSE)
colnames(CpG_data) <- c("Species","Context", "Context_To_A_or_C","Context_To_G","Context_To_T")
CpG_data$Species <- sapply(CpG_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})

CA_data <- read.delim('CA_Phis.txt', sep = "\t", header = FALSE)
colnames(CA_data) <- c("Species","Context", "Context_To_A_or_C","Context_To_G","Context_To_T")
CA_data$Species <- sapply(CA_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})

CC_data <- read.delim('CC_Phis.txt', sep = "\t", header = FALSE)
colnames(CC_data) <- c("Species","Context", "Context_To_A_or_C","Context_To_G","Context_To_T")
CC_data$Species <- sapply(CC_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})

CT_data <- read.delim('CT_Phis.txt', sep = "\t", header = FALSE)
colnames(CT_data) <- c("Species","Context", "Context_To_A_or_C","Context_To_G","Context_To_T")
CT_data$Species <- sapply(CT_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})

CpGProb_data <- read.delim('CpG_Probabilities.txt', sep = "\t", header = FALSE)
colnames(CpGProb_data) <- c("Species","Context", "Context_To_A_or_C","Context_To_G","Context_To_T")
CpGProb_data$Species <- sapply(CpGProb_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})

# load CHG to THG phi data
CHG_to_THG_phi_data <- as.data.frame(read.delim("NCHG_NTHG_averaged_Phis.txt", sep = "\t", header = FALSE))
colnames(CHG_to_THG_phi_data) <- c("Species", "CHG_to_THG_Phi")

CHG_to_THG_phi_data$Species <- sapply(CHG_to_THG_phi_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})
# load CHH to THH phi data
CHH_to_THH_phi_data <- as.data.frame(read.delim("NCHH_NTHH_averaged_Phis.txt", sep = "\t", header = FALSE))
colnames(CHH_to_THH_phi_data) <- c("Species", "CHH_to_THH_Phi")

CHH_to_THH_phi_data$Species <- sapply(CHH_to_THH_phi_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})

# load averaged C to T probability data
  # CG > TG
CG_to_TG_prob_data <- as.data.frame(read.delim("NNCGN_NNTGN_averaged_Probabilities.txt", sep = "\t", header = FALSE))
colnames(CG_to_TG_prob_data) <- c("Species", "CG_to_TG_Prob")

CG_to_TG_prob_data$Species <- sapply(CG_to_TG_prob_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})
  # CHN > THN
CH_to_TH_prob_data <- as.data.frame(read.delim("NNCHN_NNTHN_averaged_Probabilities.txt", sep = "\t", header = FALSE))
colnames(CH_to_TH_prob_data) <- c("Species", "CH_to_TH_Prob")

CH_to_TH_prob_data$Species <- sapply(CH_to_TH_prob_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})
  # CHH > THH
CHH_to_THH_prob_data <- as.data.frame(read.delim("NNCHH_NNTHH_averaged_Probabilities.txt", sep = "\t", header = FALSE))
colnames(CHH_to_THH_prob_data) <- c("Species", "CHH_to_THH_Prob")

CHH_to_THH_prob_data$Species <- sapply(CHH_to_THH_prob_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})
  #CHG > THG
CHG_to_THG_prob_data <- as.data.frame(read.delim("NNCHG_NNTHG_averaged_Probabilities.txt", sep = "\t", header = FALSE))
colnames(CHG_to_THG_prob_data) <- c("Species", "CHG_to_THG_Prob")

CHG_to_THG_prob_data$Species <- sapply(CHG_to_THG_prob_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})

# load O/E CHG data
OE_CHG_content_data <- read.delim('NonGenic_O_E_CHG_CompositionList.txt', sep = "\t", header = FALSE)
colnames(OE_CHG_content_data) <- c("Species", "O/E CHG")
OE_CHG_content_data$Species <- sapply(OE_CHG_content_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})

# load O/E CpG data
OE_CpG_content_data <- read.delim('NonGenic_O_E_CpG_CompositionList.txt', sep = "\t", header = FALSE)
colnames(OE_CpG_content_data) <- c("Species", "O/E CpG")
OE_CpG_content_data$Species <- sapply(OE_CpG_content_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})

# load CpG frequency data
CpG_frequency_data <- read.delim('NonCoding_CpG_frquencyList.txt', sep = "\t", header = FALSE)
colnames(CpG_frequency_data) <- c("Species", "CpG_Frequency")
CpG_frequency_data$Species <- sapply(CpG_frequency_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})

# load CpG Expectation data
CpG_Expectation_data <- read.delim('NonCoding_CpG_ExpectationList.txt', sep = "\t", header = FALSE)
colnames(CpG_Expectation_data) <- c("Species", "CpG_Expectation")
CpG_Expectation_data$Species <- sapply(CpG_Expectation_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})

# load CpG Binned Expectation data
CpG_BinnedExpectation_data <- read.delim('NonCoding_Binned_CpG_ExpectationList.txt', sep = "\t", header = FALSE)
colnames(CpG_BinnedExpectation_data) <- c("Species", "CpG_BinnedExpectation")
CpG_BinnedExpectation_data$Species <- sapply(CpG_BinnedExpectation_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})

# load ODD vs EVEN model RMSEs for each species
ODDvsEVEN_RMSE <- (read.delim('ODDvsEVEN_RMSEs.txt', sep = " ", header = FALSE))[,c(1,3)]
colnames(ODDvsEVEN_RMSE) <- c("Species", "OddEvenRMSE")
ODDvsEVEN_RMSE$Species <- sapply(ODDvsEVEN_RMSE$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})

CpGtoTPhi <- CpG_data$Context_To_T
names(CpGtoTPhi) <- CpG_data$Species

CAtoTPhi <- CA_data$Context_To_T
names(CAtoTPhi) <- CA_data$Species
CCtoTPhi <- CC_data$Context_To_T
names(CCtoTPhi) <- CC_data$Species
CTtoTPhi <- CT_data$Context_To_T
names(CTtoTPhi) <- CT_data$Species

CpGtoTProb <- CpGProb_data$Context_To_T
names(CpGtoTProb) <- CpGProb_data$Species

CHGtoTPhi <- CHG_to_THG_phi_data$CHG_to_THG_Phi
names(CHGtoTPhi) <- CHG_to_THG_phi_data$Species

CHHtoTPhi <- CHH_to_THH_phi_data$CHH_to_THH_Phi
names(CHHtoTPhi) <- CHH_to_THH_phi_data$Species

CGtoTProb <- CG_to_TG_prob_data$CG_to_TG_Prob
names(CGtoTProb) <- CG_to_TG_prob_data$Species

CHtoTProb <- CH_to_TH_prob_data$CH_to_TH_Prob
names(CHtoTProb) <- CH_to_TH_prob_data$Species

CHHtoTProb <- CHH_to_THH_prob_data$CHH_to_THH_Prob
names(CHHtoTProb) <- CHH_to_THH_prob_data$Species

CHGtoTProb <- CHG_to_THG_prob_data$CHG_to_THG_Prob
names(CHGtoTProb) <- CHG_to_THG_prob_data$Species

OE_CHG <- OE_CHG_content_data$`O/E CHG`
names(OE_CHG) <- OE_CHG_content_data$Species

OE_CpG <- OE_CpG_content_data$`O/E CpG`
names(OE_CpG) <- OE_CpG_content_data$Species

OddEvenRMSE <- ODDvsEVEN_RMSE$OddEvenRMSE
names(OddEvenRMSE) <- ODDvsEVEN_RMSE$Species

CpGFreq <- CpG_frequency_data$CpG_Frequency
names(CpGFreq) <- CpG_frequency_data$Species

CpGExpect <- CpG_Expectation_data$CpG_Expectation
names(CpGExpect) <- CpG_Expectation_data$Species

CpGBinExpect <- CpG_BinnedExpectation_data$CpG_BinnedExpectation
names(CpGBinExpect) <- CpG_BinnedExpectation_data$Species

merged_summary_table$CpGtoTPhi <- CpGtoTPhi[merged_summary_table$Species]
merged_summary_table$CAtoTPhi <- CAtoTPhi[merged_summary_table$Species]
merged_summary_table$CCtoTPhi <- CCtoTPhi[merged_summary_table$Species]
merged_summary_table$CTtoTPhi <- CTtoTPhi[merged_summary_table$Species]
merged_summary_table$CpGtoTProb <- CpGtoTProb[merged_summary_table$Species]
merged_summary_table$CHGtoTPhi <- CHGtoTPhi[merged_summary_table$Species]
merged_summary_table$CHHtoTPhi <-CHHtoTPhi[merged_summary_table$Species]
merged_summary_table$CGtoTProb <- CGtoTProb[merged_summary_table$Species]
merged_summary_table$CHtoTProb <- CHtoTProb[merged_summary_table$Species]
merged_summary_table$CHHtoTProb <- CHHtoTProb[merged_summary_table$Species]
merged_summary_table$CHGtoTProb <- CHGtoTProb[merged_summary_table$Species]
merged_summary_table$`O/E CHG content` <- OE_CHG[merged_summary_table$Species]
merged_summary_table$`O/E CpG in Non-Genic` <- OE_CpG[merged_summary_table$Species]
merged_summary_table$`Odd vs Even RMSE` <- OddEvenRMSE[merged_summary_table$Species]
merged_summary_table$`CpG Frequency` <- CpGFreq[merged_summary_table$Species]
merged_summary_table$`CpG Expectation` <- CpGExpect[merged_summary_table$Species]
merged_summary_table$`CpG Binned Expectation` <- CpGBinExpect[merged_summary_table$Species]


sorted_merged_summary_table <- merged_summary_table %>% arrange(desc(match(rownames(merged_summary_table), tree$tip.label)))


longsummary <- gather(sorted_merged_summary_table, key = "measure", value = "value", c("Non-Genic SNPs (log10)", "No. Individuals (log10)"))
species_sorted_for_summary_plot <- longsummary$Species[1:length(sorted_merged_summary_table[,1])]
## ##
```
# summary plot
```{r}
####
sorted_merged_summary_table$BroaderGroup <- sorted_merged_summary_table$Clade

ColoredClades <- c("Aves" = "#F9C000FF",
                                "Cnidaria" = "#FF4D6FFF",
                                "Mammalia" = "#C8350DFF",
                                "Reptilia" = "#5D7298FF",
                                "Protozoa" = "#BD777AFF",
                                "Plantae" = "#86AD34FF",
                                "Osteichthyes" = "#579EA4FF",
                                "Fungia" = "#DF7713FF",
                                "Invertebrata" = "#7E1A2FFF",
                                "Echinodermata" = "#81B28DFF",
                                "Amphibia" = "#2D2651FF")

cladeOrder <- c("Protozoa", "Plantae", "Fungia", "Cnidaria", "Invertebrata", "Echinodermata", "Amphibia", "Osteichthyes", "Reptilia", "Aves", "Mammalia")

customNumberedLabels <- c("Aves" = paste0("Aves (",sum(str_count(sorted_merged_summary_table$Clade, "Aves")),")"),
                              "Cnidaria" = paste0("Cnidaria (",sum(str_count(sorted_merged_summary_table$Clade, "Cnidaria")),")"),
                              "Mammalia" = paste0("Mammalia (",sum(str_count(sorted_merged_summary_table$Clade, "Mammalia")),")"),
                              "Reptilia" = paste0("Reptilia (",sum(str_count(sorted_merged_summary_table$Clade, "Reptilia")),")"),
                              "Protozoa" = paste0("Protozoa (",sum(str_count(sorted_merged_summary_table$Clade, "Protozoa")),")"),
                              "Plantae" = paste0("Plantae (",sum(str_count(sorted_merged_summary_table$Clade, "Plantae")),")"),
                              "Osteichthyes" = paste0("Osteichthyes (",sum(str_count(sorted_merged_summary_table$Clade, "Osteichthyes")),")"),
                              "Fungia" = paste0("Fungia (",sum(str_count(sorted_merged_summary_table$Clade, "Fungia")),")"),
                              "Invertebrata" = paste0("Invertebrata (",sum(str_count(sorted_merged_summary_table$Clade, "Invertebrata")),")"),
                              "Echinodermata" = paste0("Echinodermata (",sum(str_count(sorted_merged_summary_table$Clade, "Echinodermata")),")"),
                              "Amphibia" = paste0("Amphibia (",sum(str_count(sorted_merged_summary_table$Clade, "Amphibia")),")"))

scatter_plot2 <- ggplot(sorted_merged_summary_table, 
                       aes(x = factor(Clade,cladeOrder), 
                           y = `Non-Genic SNPs (log10)`, 
                           size = `No. Individuals (log10)`, 
                           color = Clade)) +
  geom_point(position = position_jitter(width = 0.2, height = 0), alpha = 0.7) +  # Jitter to spread out points
  theme(axis.text.x = element_text(angle = 30, hjust = 1, face = "italic", size = 12, color = "black"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 15, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        panel.grid.minor = element_line(color = "snow3", linewidth = 0.45, linetype = 1),
        panel.grid.major.y = element_line(color = "snow4", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_line(color = "tan", linewidth = 0.2, linetype = 2),
        panel.background = element_rect(fill = "white"),
        plot.background = element_rect(color = "transparent", size = 2),
        plot.margin = margin(0, 0, 0, 0), # plot.margin = margin(-8, 3, -2, 8)
        #legend.position = c(0.85,0.97),
        #legend.position = c(0.5,0.95),
        legend.position = "bottom",
        legend.background = element_rect(fill = "transparent", color = "transparent"),
        legend.direction = "horizontal",
        legend.text = element_text(size=12, color="black"),
        legend.title = element_text(size=12, color="black"),
        legend.key = element_rect(fill="transparent")) +
  scale_size_continuous(name = expression(paste("No. Individuals (",log[10],")")), range = c(2, 10)) +  # Adjust the size of the points
  scale_color_manual(values = ColoredClades) +
  scale_x_discrete(expand = expansion(mult = c(0.04, 0.1))) + #labels=customNumberedLabels
  guides(color = guide_none()) +
  labs(title = "",
       x = "Clade",
       y = expression(paste("Non-Coding SNPs (",log[10],")")))

#pdf("~/Downloads/MathiesonLab/100SpeciesPaperFigures/summarydata.pdf", 11, 5)
#scatter_plot2
#dev.off()

cladeCounts <- c("Protozoa" = sum(str_count(sorted_merged_summary_table$Clade, "Protozoa")),
                 "Plantae" = sum(str_count(sorted_merged_summary_table$Clade, "Plantae")),
                 "Fungia" = sum(str_count(sorted_merged_summary_table$Clade, "Fungia")),
                 "Cnidaria" = sum(str_count(sorted_merged_summary_table$Clade, "Cnidaria")),
                 "Invertebrata" = sum(str_count(sorted_merged_summary_table$Clade, "Invertebrata")),
                 "Echinodermata" = sum(str_count(sorted_merged_summary_table$Clade, "Echinodermata")),
                 "Amphibia" = sum(str_count(sorted_merged_summary_table$Clade, "Amphibia")),
                 "Osteichthyes" = sum(str_count(sorted_merged_summary_table$Clade, "Osteichthyes")),
                 "Reptilia" = sum(str_count(sorted_merged_summary_table$Clade, "Reptilia")),
                 "Aves" = sum(str_count(sorted_merged_summary_table$Clade, "Aves")),
                 "Mammalia" = sum(str_count(sorted_merged_summary_table$Clade, "Mammalia")))

cladecounts_df <- data.frame(
  Category = names(cladeCounts),
  Value = cladeCounts
)

####
```
# RMSE error plots
```{r}
##  ##
OddEvenRMSE_scatter <- ggplot(sorted_merged_summary_table, 
                       aes(x = factor(Clade,cladeOrder), 
                           y = log10(`Odd vs Even RMSE`), 
                           size = `Non-Genic SNPs (log10)`, 
                           color = Clade)) +
  geom_point(position = position_jitter(width = 0.2, height = 0), alpha = 0.7) +  # Jitter to spread out points
  theme(axis.text.x = element_text(angle = 30, hjust = 1, face = "italic", size = 12, color = "black"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 15, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        panel.grid.minor = element_line(color = "snow3", linewidth = 0.45, linetype = 1),
        panel.grid.major.y = element_line(color = "snow4", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_line(color = "tan", linewidth = 0.2, linetype = 2),
        panel.background = element_rect(fill = "white"),
        plot.background = element_rect(color = "transparent", size = 2),
        plot.margin = margin(0, 0, 0, 0), # plot.margin = margin(-8, 3, -2, 8)
        legend.position = "bottom",
        legend.background = element_rect(fill = "transparent", color = "transparent"),
        legend.direction = "horizontal",
        legend.text = element_text(size=12, color="black"),
        legend.title = element_text(size=12, color="black"),
        legend.key = element_rect(fill="transparent")) +
  scale_size_continuous(name = expression(paste("No. SNPs (",log[10],")")), range = c(2, 10)) +  # Adjust the size of the points
  scale_color_manual(values = ColoredClades) +
  scale_x_discrete(expand = expansion(mult = c(0.04, 0.1))) + #labels=customNumberedLabels
  guides(color = guide_none()) +
  labs(title = "",
       x = "Clade",
       y = expression(paste("Odd vs. Even RMSE (",log[10],")")))

plot(OddEvenRMSE_scatter)

pdf("~/Downloads/MathiesonLab/100SpeciesPaperFigures/OddEvenRMSEPlot.pdf", 8, 5)
OddEvenRMSE_scatter
dev.off()
```
## Selected Contexts Plot ##
```{r echo=TRUE}
ContextProbabilities <- read.delim('AllSpecies_ProportionOfContextsWithInclusionProbabilitiesAbove95.txt', sep = "\t", header = FALSE)

colnames(ContextProbabilities) <- c("Species", "Layer", "Proportion")


#changing species names
ContextProbabilities$Species <- sapply(ContextProbabilities$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})
##

#merge with class names
ContextProbabilities <- merge(ContextProbabilities, classified_species, by.x="Species",by.y="Species", by.all=TRUE)

ContextProbabilities$Layer <- replace(ContextProbabilities$Layer, ContextProbabilities$Layer==1, "2-mer")
ContextProbabilities$Layer <- replace(ContextProbabilities$Layer, ContextProbabilities$Layer==2, "3-mer")
ContextProbabilities$Layer <- replace(ContextProbabilities$Layer, ContextProbabilities$Layer==3, "4-mer")
ContextProbabilities$Layer <- replace(ContextProbabilities$Layer, ContextProbabilities$Layer==4, "5-mer")
ContextProbabilities$Layer <- replace(ContextProbabilities$Layer, ContextProbabilities$Layer==5, "6-mer")
ContextProbabilities$Layer <- replace(ContextProbabilities$Layer, ContextProbabilities$Layer==6, "7-mer")
ContextProbabilities$Layer <- replace(ContextProbabilities$Layer, ContextProbabilities$Layer==7, "8-mer")
ContextProbabilities$Layer <- replace(ContextProbabilities$Layer, ContextProbabilities$Layer==8, "9-mer")

#
Mer_medianProportions <- NULL
Mer_medianProportions <- append(Mer_medianProportions,median(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="2-mer"])))
Mer_medianProportions <- append(Mer_medianProportions,median(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="3-mer"])))
Mer_medianProportions <- append(Mer_medianProportions,median(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="4-mer"])))
Mer_medianProportions <- append(Mer_medianProportions,median(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="5-mer"])))
Mer_medianProportions <- append(Mer_medianProportions,median(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="6-mer"])))
Mer_medianProportions <- append(Mer_medianProportions,median(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="7-mer"])))
Mer_medianProportions <- append(Mer_medianProportions,median(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="8-mer"])))
Mer_medianProportions <- append(Mer_medianProportions,median(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="9-mer"])))

Mer_5thQuantiles <- NULL
Mer_5thQuantiles <- append(Mer_5thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="2-mer"]), 0.05))
Mer_5thQuantiles <- append(Mer_5thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="3-mer"]), 0.05))
Mer_5thQuantiles <- append(Mer_5thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="4-mer"]), 0.05))
Mer_5thQuantiles <- append(Mer_5thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="5-mer"]), 0.05))
Mer_5thQuantiles <- append(Mer_5thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="6-mer"]), 0.05))
Mer_5thQuantiles <- append(Mer_5thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="7-mer"]), 0.05))
Mer_5thQuantiles <- append(Mer_5thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="8-mer"]), 0.05))
Mer_5thQuantiles <- append(Mer_5thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="9-mer"]), 0.05))

Mer_95thQuantiles <- NULL
Mer_95thQuantiles <- append(Mer_95thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="2-mer"]), 0.95))
Mer_95thQuantiles <- append(Mer_95thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="3-mer"]), 0.95))
Mer_95thQuantiles <- append(Mer_95thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="4-mer"]), 0.95))
Mer_95thQuantiles <- append(Mer_95thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="5-mer"]), 0.95))
Mer_95thQuantiles <- append(Mer_95thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="6-mer"]), 0.95))
Mer_95thQuantiles <- append(Mer_95thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="7-mer"]), 0.95))
Mer_95thQuantiles <- append(Mer_95thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="8-mer"]), 0.95))
Mer_95thQuantiles <- append(Mer_95thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="9-mer"]), 0.95))

Mer_10thQuantiles <- NULL
Mer_10thQuantiles <- append(Mer_10thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="2-mer"]), 0.1))
Mer_10thQuantiles <- append(Mer_10thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="3-mer"]), 0.1))
Mer_10thQuantiles <- append(Mer_10thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="4-mer"]), 0.1))
Mer_10thQuantiles <- append(Mer_10thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="5-mer"]), 0.1))
Mer_10thQuantiles <- append(Mer_10thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="6-mer"]), 0.1))
Mer_10thQuantiles <- append(Mer_10thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="7-mer"]), 0.1))
Mer_10thQuantiles <- append(Mer_10thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="8-mer"]), 0.1))
Mer_10thQuantiles <- append(Mer_10thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="9-mer"]), 0.1))

Mer_90thQuantiles <- NULL
Mer_90thQuantiles <- append(Mer_90thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="2-mer"]), 0.90))
Mer_90thQuantiles <- append(Mer_90thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="3-mer"]), 0.90))
Mer_90thQuantiles <- append(Mer_90thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="4-mer"]), 0.90))
Mer_90thQuantiles <- append(Mer_90thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="5-mer"]), 0.90))
Mer_90thQuantiles <- append(Mer_90thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="6-mer"]), 0.90))
Mer_90thQuantiles <- append(Mer_90thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="7-mer"]), 0.90))
Mer_90thQuantiles <- append(Mer_90thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="8-mer"]), 0.90))
Mer_90thQuantiles <- append(Mer_90thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="9-mer"]), 0.90))

Mer_100thQuantiles <- NULL
Mer_100thQuantiles <- append(Mer_100thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="2-mer"]), 1))
Mer_100thQuantiles <- append(Mer_100thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="3-mer"]), 1))
Mer_100thQuantiles <- append(Mer_100thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="4-mer"]), 1))
Mer_100thQuantiles <- append(Mer_100thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="5-mer"]), 1))
Mer_100thQuantiles <- append(Mer_100thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="6-mer"]), 1))
Mer_100thQuantiles <- append(Mer_100thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="7-mer"]), 1))
Mer_100thQuantiles <- append(Mer_100thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="8-mer"]), 1))
Mer_100thQuantiles <- append(Mer_100thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="9-mer"]), 1))

Mer_75thQuantiles <- NULL
Mer_75thQuantiles <- append(Mer_75thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="2-mer"]), 0.75))
Mer_75thQuantiles <- append(Mer_75thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="3-mer"]), 0.75))
Mer_75thQuantiles <- append(Mer_75thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="4-mer"]), 0.75))
Mer_75thQuantiles <- append(Mer_75thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="5-mer"]), 0.75))
Mer_75thQuantiles <- append(Mer_75thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="6-mer"]), 0.75))
Mer_75thQuantiles <- append(Mer_75thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="7-mer"]), 0.75))
Mer_75thQuantiles <- append(Mer_75thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="8-mer"]), 0.75))
Mer_75thQuantiles <- append(Mer_75thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="9-mer"]), 0.75))

Mer_25thQuantiles <- NULL
Mer_25thQuantiles <- append(Mer_25thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="2-mer"]), 0.25))
Mer_25thQuantiles <- append(Mer_25thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="3-mer"]), 0.25))
Mer_25thQuantiles <- append(Mer_25thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="4-mer"]), 0.25))
Mer_25thQuantiles <- append(Mer_25thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="5-mer"]), 0.25))
Mer_25thQuantiles <- append(Mer_25thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="6-mer"]), 0.25))
Mer_25thQuantiles <- append(Mer_25thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="7-mer"]), 0.25))
Mer_25thQuantiles <- append(Mer_25thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="8-mer"]), 0.25))
Mer_25thQuantiles <- append(Mer_25thQuantiles, quantile(sort(ContextProbabilities$Proportion[ContextProbabilities$Layer=="9-mer"]), 0.25))

Mer_layers <- c("2-mer","3-mer","4-mer","5-mer","6-mer","7-mer","8-mer","9-mer")
Mer_total_contexts <- c(12, 96, 192, 1536, 3072, 24576, 49152, 393216)
forContextLayerSizePlot <- data.frame(Mer_layers, Mer_total_contexts)
colnames(forContextLayerSizePlot) <- c("Context layer", "Number of contexts")

contextsizesplot_for_presentation <- ggplot(forContextLayerSizePlot) + geom_line(aes(x=`Context layer`, y=log10(`Number of contexts`), group=1), color="#7000d2") + 
  geom_point(aes(x=`Context layer`, y=log10(`Number of contexts`), group=1), color="#7000d2", size=5) + 
  theme(panel.background = element_rect(fill = "white"),
        panel.grid = element_line(color = "snow3", linetype = 1),
        panel.grid.minor.y = element_blank(),
        axis.title.y = element_text(size = 14, color = "black"),
        axis.title.x = element_text(size = 14, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        axis.text.x = element_text(size = 12, angle = 30, color = "black"),
        title = element_text(size = 20, color = "snow4"),
        legend.text = element_text(size = 14),
        legend.title = element_text(size = 14),
        legend.position = "none",
        aspect.ratio = 1) + 
  labs(title = "", x = "Context layer", y = expression(paste("Number of contexts (",log[10],")")))

simplified_pip_data <- data.frame(Mer_layers, Mer_medianProportions, Mer_5thQuantiles, Mer_10thQuantiles, Mer_25thQuantiles, Mer_75thQuantiles, Mer_90thQuantiles, Mer_95thQuantiles, Mer_100thQuantiles, Mer_total_contexts)
colnames(simplified_pip_data) <- c("Layer", "Median", "Q5th", "Q10th", "Q25th", "Q75th", "Q90th", "Q95th", "Q100th", "Contexts")

includedContextsPlot <- ggplot() + geom_ribbon(data = simplified_pip_data, aes(x=Layer, ymin=Q5th, ymax=Q95th, group=1), fill="#f0dbff", alpha=0.5) +
  geom_ribbon(data = simplified_pip_data, aes(x=Layer, ymin=Q10th, ymax=Q90th, group=1), fill="#cb9cf1", alpha=0.5) +
  geom_ribbon(data = simplified_pip_data, aes(x=Layer, ymin=Q25th, ymax=Q75th, group=1), fill="#a25ce2", alpha=0.5) +
  geom_line(data = simplified_pip_data, aes(x=Layer, y=Median, group=1), color="#7000d2", lwd=1.4) +
  labs(y="Proportion of contexts with PIP >0.95", x="Context layer", title = "") +
  theme(panel.background = element_rect(fill = "white"),
        panel.grid = element_line(color = "tan", linetype = 2),
        panel.grid.minor.y = element_blank(),
        #panel.grid.major.x = element_blank(),
        #panel.border = element_rect(color = "black", fill = "transparent"),
        axis.title.y = element_text(size = 15, color = "black"),
        axis.title.x = element_text(size = 15, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        axis.text.x = element_text(size = 12, angle = 30, color = "black"),
        title = element_text(size = 20, color = "snow4"),
        legend.text = element_text(size = 14),
        legend.title = element_text(size = 14),
        legend.position = "none",
        aspect.ratio = 1) +
  geom_label_repel(data = simplified_pip_data,
                   aes(label = paste0(round(Q95th * Contexts)), x=Layer, y = Q95th),
             label.padding = unit(0.35, "lines"),
             label.size = 0.25,
             size = 4,
             color = "black",
             fill="white",
             nudge_y = 0.065,
             nudge_x = 0.04,
             segment.size = 0.45) +
  geom_line(data = simplified_pip_data,
            aes(x=Layer, y=Q95th, group=1),
            color="snow4", lwd=0.75, linetype=4, alpha=0.5) +
  geom_line(data = simplified_pip_data,
            aes(x=Layer, y=Q75th, group=1),
            color="snow4", lwd=0.75, linetype=4, alpha=0.5) +
  geom_line(data = simplified_pip_data,
            aes(x=Layer, y=Q25th, group=1),
            color="snow4", lwd=0.75, linetype=4, alpha=0.5) +
  geom_line(data = simplified_pip_data,
            aes(x=Layer, y=Q5th, group=1),
            color="snow4", lwd=0.75, linetype=4, alpha=0.5) +
  geom_line(data = simplified_pip_data,
            aes(x=Layer, y=Q10th, group=1),
            color="snow4", lwd=0.75, linetype=4, alpha=0.5) +
  geom_line(data = simplified_pip_data,
            aes(x=Layer, y=Q90th, group=1),
            color="snow4", lwd=0.75, linetype=4, alpha=0.5) +
  annotate("rect", xmin = 6.3, xmax = 8.5, ymin = 0.51, ymax = 0.77, fill = "white", color = "black") +
  annotate("rect", xmin = 6.4, xmax = 6.7, ymin = 0.7, ymax = 0.74, fill = "#f0dbff", color = "black", alpha = 0.8) +
  annotate("text", x = 6.8, y = 0.72, label = "5th-95th", hjust = 0) +
  annotate("rect", xmin = 6.4, xmax = 6.7, ymin = 0.62, ymax = 0.66, fill = "#cb9cf1", color = "black", alpha = 0.8) +
  annotate("text", x = 6.8, y = 0.64, label = "10th-90th", hjust = 0) +
  annotate("rect", xmin = 6.4, xmax = 6.7, ymin = 0.54, ymax = 0.58, fill = "#a25ce2", color = "black", alpha = 0.8) +
  annotate("text", x = 6.8, y = 0.56, label = "25th-75th", hjust = 0)

pdf("~/Downloads/MathiesonLab/100SpeciesPaperFigures/includedContextsPlot.pdf", 5, 5)
includedContextsPlot
dev.off()

trelis_chart_pip95 <- ggplot() + geom_line(data = ContextProbabilities,
                      aes(x=Layer, y=Proportion, group=Species, color=Class)) +
  labs(y="Proportion of contexts with inclusion probability >0.95", x="") +
  theme(panel.background = element_rect(fill = "white"),
        panel.grid = element_line(color = "snow2", linetype = 1),
        axis.text = element_text(size = 15),
        aspect.ratio = 1,
        legend.position = "none",
        legend.text = element_text(size = 12),
        legend.title = element_text(size = 14),
        axis.text.x = element_text(angle = 70, hjust = 1, size = 12),
        strip.text = element_text(size = 8, color = "black"),
        strip.background = element_rect(fill = "white", color = "black")) +
  scale_color_manual(values = ColoredClades) +
  facet_wrap(~Class)

pdf("~/Downloads/MathiesonLab/100SpeciesPaperFigures/TrelisChartincludedContextsPlot.pdf", 8, 8)
trelis_chart_pip95
dev.off()
```
#figure 1 plotting
```{r}
pdf("~/Downloads/MathiesonLab/100SpeciesPaperFigures/Figure1_Panel_summary_and_includedContextsPlot.pdf", 12, 8)
plot_grid(treeForPowerpoint, plot_grid(scatter_plot2, PerClade_RMSECurvesPlot, includedContextsPlot, ncol = 3, rel_widths = c(1.2,1,1), axis = "tb", align = "h", labels = c("B", "C", "D"), label_size = 15, label_fontface = "bold", label_colour = "black"), nrow = 2, rel_heights = c(0.5,1), labels = c("A", ""), label_size = 15, label_fontface = "bold", label_colour = "black")
dev.off()
```
# Data filter
```{r}
# remove outliers from scatter plots
excluded_species <- c("Clupea_harengus_(Atlantic_Herring)",
                      "Vitis_vinifera_(Wine_Grape)",
                      "Camellia_sinensis_(Lincang_Tea)",
                      "Eurytemora_affinis_(Copepod)",
                      "Trichaptum_abietinum_(Purple_Tooth_Fungia)") 
# Filter the data to remove the specific species
sorted_merged_summary_table_filteredSpecs <- sorted_merged_summary_table %>%
  filter(!Species %in% excluded_species)

# remove bad species from the tree
tree <- drop.tip(tree, excluded_species)

ape::write.tree(tree, file='MutationEvolution_108Species.nwk')
```
# Phylogenetic regressions
```{r}
CpG_to_T_phi_vec <- sorted_merged_summary_table_filteredSpecs$CpGtoTPhi
names(CpG_to_T_phi_vec) <- sorted_merged_summary_table_filteredSpecs$Species

mu1_vec <- log(sorted_merged_summary_table_filteredSpecs$CGtoTProb/sorted_merged_summary_table_filteredSpecs$CHtoTProb)
names(mu1_vec) <- sorted_merged_summary_table_filteredSpecs$Species

mu2_vec <- log(sorted_merged_summary_table_filteredSpecs$CGtoTProb/sorted_merged_summary_table_filteredSpecs$CHHtoTProb)
names(mu2_vec) <- sorted_merged_summary_table_filteredSpecs$Species

mu3_vec <- log(sorted_merged_summary_table_filteredSpecs$CHGtoTProb/sorted_merged_summary_table_filteredSpecs$CHHtoTProb)
names(mu3_vec) <- sorted_merged_summary_table_filteredSpecs$Species

CpG_to_T_prob_vec <- sorted_merged_summary_table_filteredSpecs$CpGtoTProb
names(CpG_to_T_prob_vec) <- sorted_merged_summary_table_filteredSpecs$Species

CpG_methylation_vec <- sorted_merged_summary_table_filteredSpecs$`Average CpG Methylation`
names(CpG_methylation_vec) <- sorted_merged_summary_table_filteredSpecs$Species

BodyTemp_vec <- sorted_merged_summary_table_filteredSpecs$`Body Temperature`
names(BodyTemp_vec) <- sorted_merged_summary_table_filteredSpecs$Species

OECpG_vec <- sorted_merged_summary_table_filteredSpecs$`O/E CpG in Non-Genic`
names(OECpG_vec) <- sorted_merged_summary_table_filteredSpecs$Species

CpGFreq_vec <- sorted_merged_summary_table_filteredSpecs$`CpG Frequency`
names(CpGFreq_vec) <- sorted_merged_summary_table_filteredSpecs$Species

GC_cont_vec <- sorted_merged_summary_table_filteredSpecs$`Non-Genic GC Content`
names(GC_cont_vec) <- sorted_merged_summary_table_filteredSpecs$Species

CHG_phi_vec <- sorted_merged_summary_table_filteredSpecs$CHGtoTPhi
names(CHG_phi_vec) <- sorted_merged_summary_table_filteredSpecs$Species

OECHG_vec <- sorted_merged_summary_table_filteredSpecs$`O/E CHG content`
names(OECHG_vec) <- sorted_merged_summary_table_filteredSpecs$Species

plant_OECHG_vec <- sorted_merged_summary_table_filteredSpecs[sorted_merged_summary_table_filteredSpecs$Clade=="Plantae",]$`O/E CHG content`
names(plant_OECHG_vec) <- sorted_merged_summary_table_filteredSpecs[sorted_merged_summary_table_filteredSpecs$Clade=="Plantae",]$Species

CHG_methylation_vec <- sorted_merged_summary_table_filteredSpecs$`Average CHG Methylation`
names(CHG_methylation_vec) <- sorted_merged_summary_table_filteredSpecs$Species

CHH_methylation_vec <- sorted_merged_summary_table_filteredSpecs$`Average CHH Methylation`
names(CHH_methylation_vec) <- sorted_merged_summary_table_filteredSpecs$Species

Clade_vec <- sorted_merged_summary_table_filteredSpecs$Clade
names(Clade_vec) <- sorted_merged_summary_table_filteredSpecs$Species

PhylogeneticRegression_merged_data <- data.frame(trait_CpGMethyl = CpG_methylation_vec,
                          pred_CpGPhi = CpG_to_T_phi_vec,
                          pred_CpGProb = CpG_to_T_prob_vec,
                          trait_bodytemp = BodyTemp_vec,
                          pred_OECpG = OECpG_vec,
                          trait_GCCont = GC_cont_vec,
                          pred_CHGPhi = CHG_phi_vec,
                          trait_OE_CHG = OECHG_vec,
                          trait_CHG_methyl = CHG_methylation_vec,
                          Trait_CHH_methyl = CHH_methylation_vec,
                          Trait_mu1 = mu1_vec,
                          Trait_mu2 = mu2_vec,
                          Trait_mu3 = mu3_vec,
                          Clade = Clade_vec,
                          CpGFreq = CpGFreq_vec)
PhylogeneticRegression_merged_data$expCpG <- 1/(exp(PhylogeneticRegression_merged_data$Trait_mu1) + 1)

#SpeciesWithBodyTemp_Mu1_on_Methyl_and_temp_lambdaModel <- phylolm(Trait_mu1 ~ trait_bodytemp, data = PhylogeneticRegression_merged_data[PhylogeneticRegression_merged_data$Clade %in% c("Mammalia", "Reptilia", "Aves", "Osteichthyes"),], phy = tree, model = "lambda", boot = 1000)

#Plants_OECHG_on_CHGPhi_lambdaModel <- phylolm(trait_OE_CHG~pred_CHGPhi, data = PhylogeneticRegression_merged_data[PhylogeneticRegression_merged_data$Clade=="Plantae",], phy = tree, model = "lambda", boot = 1000)
#OECHG_on_CHGPhi_lambdaModel <- phylolm(trait_OE_CHG~pred_CHGPhi, data = PhylogeneticRegression_merged_data, phy = tree, model = "lambda", boot = 1000)
#CHGPhi_on_methylation_lambdaModel <- phylolm(pred_CHGPhi~trait_CHG_methyl, data = PhylogeneticRegression_merged_data, phy = tree, model = "lambda", boot = 1000)

#CpGPhi_on_methylation_lambdaModel <- phylolm(pred_CpGPhi~trait_CpGMethyl, data = PhylogeneticRegression_merged_data, phy = tree, model = "lambda", boot = 1000)
#OECpG_on_CpGPhi_lambdaModel <- phylolm(pred_OECpG~pred_CpGPhi, data = PhylogeneticRegression_merged_data, phy = tree, model = "lambda", boot = 1000)

#CombinedPhi_on_methylation_lambdaModel <- phylolm(pred_Combined_CpGPhi~trait_CpGMethyl, data = PhylogeneticRegression_merged_data, phy = tree, model = "lambda", boot = 1000)
#OECpG_on_InversePhi_lampdaModel <- phylolm(pred_OECpG~pred_inversePhi, data = PhylogeneticRegression_merged_data, phy = tree, model = "lambda", boot = 1000)
#OECpG_on_CombinedPhi_lambdaModel <- phylolm(pred_OECpG~pred_Combined_CpGPhi, data = PhylogeneticRegression_merged_data, phy = tree, model = "lambda", boot = 1000)

#GCContent_on_CpGPhi_lambdaModel <- phylolm(trait_GCCont~pred_CpGPhi, data = PhylogeneticRegression_merged_data, phy = tree, model = "lambda", boot = 1000)

#CpGPhi_on_bodytemp_lambdaModel <- phylolm(pred_CpGPhi~trait_bodytemp, data = PhylogeneticRegression_merged_data, phy = tree, model = "lambda", boot = 1000)

#CpGPhi_on_methylationAndBodytemp_lambdaModel <- phylolm(pred_CpGPhi~trait_CpGMethyl + trait_bodytemp, data = PhylogeneticRegression_merged_data, phy = tree, model = "lambda", boot = 1000)

#mu1_on_CpGMethyl_lambdaModel <- phylolm(Trait_mu1~trait_CpGMethyl, data = PhylogeneticRegression_merged_data, phy = tree, model = "lambda", boot = 1000)
#OECpG_on_mu1_lambdaModel <- phylolm(pred_OECpG~Trait_mu1, data = PhylogeneticRegression_merged_data, phy = tree, model = "lambda", boot = 1000)
#CpGFreq_on_mu1_lambdaModel <- phylolm(CpGFreq~Trait_mu1, data = PhylogeneticRegression_merged_data, phy = tree, model = "lambda", boot = 1000)

#Plants_mu2_on_CpGminusCHHMethyl_lambdaModel <- phylolm(Trait_mu2~(trait_CpGMethyl - Trait_CHH_methyl), data = PhylogeneticRegression_merged_data[PhylogeneticRegression_merged_data$Clade=="Plantae",], phy = tree, model = "lambda", boot = 1000)
#Plants_mu3_on_CHGminusCHHMethyl_lambdaModel <- phylolm(Trait_mu3~(trait_CHG_methyl - Trait_CHH_methyl), data = PhylogeneticRegression_merged_data[PhylogeneticRegression_merged_data$Clade=="Plantae",], phy = tree, model = "lambda", boot = 1000)
```
# CpG plots: methylation, OE CpG, GC content
```{r}
sorted_merged_summary_table_filteredSpecs$CorrectedCpGPhi <- sorted_merged_summary_table_filteredSpecs$CpGtoTPhi - log((exp(sorted_merged_summary_table_filteredSpecs$CAtoTPhi) + exp(sorted_merged_summary_table_filteredSpecs$CCtoTPhi) + exp(sorted_merged_summary_table_filteredSpecs$CTtoTPhi))/3)

sorted_merged_summary_table_filteredSpecs$mu1 <- log(sorted_merged_summary_table_filteredSpecs$CGtoTProb/sorted_merged_summary_table_filteredSpecs$CHtoTProb)

sorted_merged_summary_table_filteredSpecs$mu2 <- log(sorted_merged_summary_table_filteredSpecs$CGtoTProb/sorted_merged_summary_table_filteredSpecs$CHHtoTProb)

sorted_merged_summary_table_filteredSpecs$mu3 <- log(sorted_merged_summary_table_filteredSpecs$CHGtoTProb/sorted_merged_summary_table_filteredSpecs$CHHtoTProb)

## CpG methylation and mutational shift plot ##
mCpG_CpG_scatter <- ggplot() + 
  geom_point(data = sorted_merged_summary_table_filteredSpecs, 
             aes(x=`Average CpG Methylation`, y=mu1, color=Clade), 
             size=5, pch=16, alpha=0.8) +
scale_color_manual(values = ColoredClades) +
  theme_minimal() +
  theme(panel.background = element_rect(fill = "white"),
        panel.grid = element_line(color = "tan", linetype = 2, linewidth = 0.2),
        axis.text=element_text(size = 16, color = "black"),
        title = element_blank(),
        axis.title.x = element_text(size = 18, color = "black"),
        axis.title.y = element_text(size = 18, color = "black"),
        aspect.ratio = 1,
        legend.title = element_blank(),
        legend.position = "none",
        legend.direction = "horizontal",
        legend.text = element_text(size = 12, color = "black"),
        legend.background = element_rect(color = "transparent", fill = "transparent"),
        plot.margin = unit(c(0,0,0.3,0), "cm")) +
  labs(title = "", y = "CpG > T mutational shift (log)", x = "mCpG%") +
  annotate("text", x = 0.65, y = 0, label = expression(paste(p == 0.21)), hjust = 0, size = 6)

Plant_CpG_scatter <- ggplot() + 
  geom_point(data = sorted_merged_summary_table_filteredSpecs[sorted_merged_summary_table_filteredSpecs$Clade=="Plantae",], 
             aes(x=(`Average CpG Methylation` - `Average CHH Methylation`), y=mu2, color=Clade), 
             size=5, pch=16, alpha=0.8) +
scale_color_manual(values = ColoredClades) +
  theme_minimal() +
  theme(panel.background = element_rect(fill = "white"),
        panel.grid = element_line(color = "tan", linetype = 2, linewidth = 0.2),
        axis.text=element_text(size = 16, color = "black"),
        title = element_blank(),
        axis.title.x = element_text(size = 18, color = "black"),
        axis.title.y = element_text(size = 18, color = "black"),
        aspect.ratio = 1,
        legend.title = element_blank(),
        legend.position = "none",
        legend.direction = "horizontal",
        legend.text = element_text(size = 12, color = "black"),
        legend.background = element_rect(color = "transparent", fill = "transparent"),
        plot.margin = unit(c(0,0,0.3,0), "cm")) +
  labs(title = "", y = "CpG>T/CHH>T mutational shift (log)", x = "mCpG% - mCHH%") +
  annotate("text", x = 0.65, y = 1.9, label = expression(paste(p == 0.97)), hjust = 0, size = 6)

##Observed/Epected CpG and CpG mutational shift plot ## 
CpG_OE_CpG_scatter <- ggplot() +
  geom_point(data = sorted_merged_summary_table_filteredSpecs, 
             aes(x=mu1, y=`O/E CpG in Non-Genic`, color=Clade), 
             size=5, pch=16, alpha=0.8) +
scale_color_manual(values = ColoredClades) +
  theme_minimal() +
  theme(panel.background = element_rect(fill = "white"),
        panel.grid = element_line(color = "tan", linetype = 2, linewidth = 0.2),
        axis.text=element_text(size = 16, color = "black"),
        title = element_blank(),
        axis.title.x = element_text(size = 18, color = "black"),
        axis.title.y = element_text(size = 18, color = "black"), #angle = 0, vjust = 0.5
        aspect.ratio = 1,
        legend.title = element_blank(),
        legend.position = "none",
        legend.direction = "horizontal",
        legend.text = element_text(size = 12, color = "black"),
        legend.background = element_rect(color = "transparent", fill = "transparent"),
        plot.margin = unit(c(0,0,0.3,0), "cm")) +
  labs(title = "", y = "O/E CpG", x = expression(paste("CpG > T mutational shift (log)"))) +
  annotate("text", x = 1.5, y = 1.7, label = expression(paste(p == 4.03 %*% 10^-4)), hjust = 0, size = 6)

```
# plotting tree with CpG data
```{r}
library(treeio)
library(ggstar)
library(ggnewscale)
library(ggpattern)
knitr::opts_knit$set(root.dir = "~/Downloads/MathiesonLab/Temporary_100Species_Downloads")

# The rectangular layout tree.
p <- ggtree(tree, layout="rectangular", size=2, color = "snow4")

p <- p %<+% sorted_merged_summary_table_filteredSpecs + 
  geom_star(mapping=aes(fill=Clade, color = Clade, x = 1645), size = 4, starshape=13,
                        position="identity",starstroke=0.1) +
  scale_fill_manual(values=ColoredClades,
                          name ="",
                           guide=guide_legend(keywidth = 0.5, 
                                        keyheight = 0.5, order=1),
                           na.translate=FALSE) +
  scale_color_manual(values=ColoredClades,
                          name ="",
                           guide=guide_legend(keywidth = 0.5, 
                                        keyheight = 0.5, order=1),
                           na.translate=FALSE)

p <- p  + new_scale_fill() + 
  geom_fruit(geom=geom_tile,
                    mapping=aes(y=label, fill=mu1),
             offset = 0.1, size = 0.02, width = 80) +
         scale_fill_continuous(name = "CpG > T mutational shift (log)",
                               low = "white",
                               high = "goldenrod",
                               #breaks = c(0.2, 0.4, 0.6, 0.8, 1),
                               #limits = c(0.2, 1),
                               #oob = scales::squish,
                               guide=guide_legend(keywidth = 0.5, 
                                        keyheight = 0.3, order=2)) +
         geom_treescale(fontsize=2, linesize=0.3, x=4.9, y=0.1) +
         theme(legend.position=c(0.93, 0.5),
               legend.background=element_rect(fill=NA),
               legend.title=element_text(size=6.5),
               legend.text=element_text(size=4.5),
               legend.spacing.y = unit(0.02, "cm"),
             )

p <- p  + new_scale_fill() + 
  geom_fruit(geom=geom_tile,
                    mapping=aes(y=label, fill=`O/E CpG in Non-Genic`),
             offset = 0.1, size = 0.02, width = 80) +
         scale_fill_continuous(name = "O/E CpG",
                               low = "white",
                               high = "#21b479",
                               breaks = c(0.2, 0.4, 0.6, 0.8, 1),
                               limits = c(0.2, 1),
                               oob = scales::squish,
                               guide=guide_legend(keywidth = 0.5, 
                                        keyheight = 0.3, order=3)) +
         geom_treescale(fontsize=2, linesize=0.3, x=4.9, y=0.1) +
         theme(legend.position=c(0.93, 0.5),
               legend.background=element_rect(fill=NA),
               legend.title=element_text(size=6.5),
               legend.text=element_text(size=4.5),
               legend.spacing.y = unit(0.02, "cm"),
             )

p <- p + new_scale_fill() +
         geom_fruit(geom=geom_tile_pattern,
                  mapping=aes(y=label, fill=`Average CpG Methylation`, pattern = ifelse(is.na(`Average CpG Methylation`), "stripe", NA)),
                  offset = 0.1,size = 0.02, width = 80,
                  pattern = "stripe",
                  pattern_color = "transparent",
                  pattern_fill = "transparent") +
         scale_fill_continuous(na.value = "gray15",
                               low = "white",
                               high = "mediumpurple",
                                name = "mCpG%",
                             guide=guide_legend(keywidth = 0.5, 
                                             keyheight = 0.3, order=4)) +
  scale_pattern_color_manual(na.value = "white") +
  scale_pattern_fill_manual(na.value = "white")

p <- p + theme(legend.position=c(.25, .75), legend.title = element_text(size = 16), legend.text = element_text(size = 15))
```
# figure 3 plotting
```{r}
pdf("Figure3_CpG.pdf", 15, 13)
plot_grid(p, NULL, plot_grid(mCpG_CpG_scatter, Plant_CpG_scatter, CpG_OE_CpG_scatter, ncol = 1, align = "v", hjust = 1, labels = c("B", "C", "D"), label_x = 0.32, label_size = 16), ncol = 3, rel_widths = c(1.2, -0.7, 2.6), labels = c("A", "", ""), label_x = 0.4, label_size = 16)
dev.off()
```
# NCHG > NTHG phi plot
```{r}
mCHG_CHG_scatter <- ggplot() + 
  geom_point(data = sorted_merged_summary_table_filteredSpecs[sorted_merged_summary_table_filteredSpecs$Clade=="Plantae",], 
             aes(x=`Average CHG Methylation` - `Average CHH Methylation`, y=mu3, color=Clade), 
             size=5, pch=16, alpha=0.8) +
scale_color_manual(values = ColoredClades) +
  theme_minimal() +
  theme(panel.background = element_rect(fill = "white"),
        panel.grid = element_line(color = "tan", linetype = 2, linewidth = 0.2),
        axis.text=element_text(size = 16, color = "black"),
        title = element_blank(),
        axis.title.x = element_text(size = 18, color = "black"),
        axis.title.y = element_text(size = 18, color = "black"),
        aspect.ratio = 1,
        legend.title = element_blank(),
        legend.position = "none",
        legend.direction = "horizontal",
        legend.text = element_text(size = 12, color = "black"),
        legend.background = element_rect(color = "transparent", fill = "transparent")) +
  labs(title = "", y = "CHG>T/CHH>T mutational shift", x = "mCHG% - mCHH%") +
  annotate("text", x = 0.5, y = 1.4, label = expression(paste(p == 0.57)), hjust = 0, size = 7, color = "black")

OE_CHG_CHG_scatter <- ggplot() + 
  geom_point(data = sorted_merged_summary_table_filteredSpecs[sorted_merged_summary_table_filteredSpecs$Clade=="Plantae",], 
             aes(x=mu3, y=(`O/E CHG content`), color=Clade), 
             size=5, pch=16, alpha=0.8) +
scale_color_manual(values = ColoredClades) +
  theme_minimal() +
  theme(panel.background = element_rect(fill = "white"),
        panel.grid = element_line(color = "tan", linetype = 2, linewidth = 0.2),
        axis.text=element_text(size = 16, color = "black"),
        title = element_blank(),
        axis.title.x = element_text(size = 18, color = "black"),
        axis.title.y = element_text(size = 18, color = "black"),
        aspect.ratio = 1,
        legend.title = element_blank(),
        legend.position = "none",
        legend.direction = "horizontal",
        legend.text = element_text(size = 12, color = "black"),
        legend.background = element_rect(color = "transparent", fill = "transparent")) +
  labs(title = "", y = "O/E CHG", x = "CHG>T / CHH>T mutational shift") +
  annotate("text", x = 0.8, y = 1.5, label = expression(paste(p == 6.75%*% 10^-7)), hjust = 0, size = 7, color = "black")

pdf("CHGPhi_vs_OECHG.pdf", 6, 6)
OE_CHG_CHG_scatter
dev.off()
```
# CHG and CpG methylation model residuals
```{r}
Plant_sorted_merged_summary_table_filteredSpecs <- sorted_merged_summary_table_filteredSpecs[sorted_merged_summary_table_filteredSpecs$Clade=="Plantae",]
Plant_sorted_merged_summary_table_filteredSpecs$mu2_on_Methylation_residuals <- Plants_mu2_on_CpGminusCHHMethyl_lambdaModel$residuals[Plant_sorted_merged_summary_table_filteredSpecs$Species]
Plant_sorted_merged_summary_table_filteredSpecs$mu3_on_Methylation_residuals <- Plants_mu3_on_CHGminusCHHMethyl_lambdaModel$residuals[Plant_sorted_merged_summary_table_filteredSpecs$Species]

#cor(Plant_sorted_merged_summary_table_filteredSpecs$mu2_on_Methylation_residuals, Plant_sorted_merged_summary_table_filteredSpecs$mu3_on_Methylation_residuals, use = "complete.obs")

Residuals_CHG_CpG_MethylModels_Plants_scatter <- ggplot() + 
  geom_point(data = Plant_sorted_merged_summary_table_filteredSpecs, 
             aes(x=mu2_on_Methylation_residuals, y=mu3_on_Methylation_residuals, color=Clade), 
             size=5, pch=16, alpha=0.8) +
scale_color_manual(values = ColoredClades) +
  theme_minimal() +
  theme(panel.background = element_rect(fill = "white"),
        panel.grid = element_line(color = "tan", linetype = 2, linewidth = 0.2),
        axis.text=element_text(size = 16, color = "black"),
        title = element_blank(),
        axis.title.x = element_text(size = 18, color = "black"),
        axis.title.y = element_text(size = 18, color = "black"),
        aspect.ratio = 1,
        legend.title = element_blank(),
        legend.position = "none",
        legend.direction = "horizontal",
        legend.text = element_text(size = 12, color = "black"),
        legend.background = element_rect(color = "transparent", fill = "transparent")) +
  labs(title = "", y = "e: CHG>T mutational shift ~ mCHG%", x = "e: CpG>T mutational shift ~ mCpG%") +
  annotate("text", x = -0.5, y = 0.75, label = expression(paste(r == 0.829)), hjust = 0, size = 7, color = "black")

pdf("CHG_vs_CpG_plant_residuals.pdf", 6, 6)
Residuals_CHG_CpG_MethylModels_Plants_scatter
dev.off()
```
# Plant phylogeny
```{r}
knitr::opts_knit$set(root.dir = "~/Downloads/MathiesonLab/Temporary_100Species_Downloads")

non_plants <- sorted_merged_summary_table_filteredSpecs[sorted_merged_summary_table_filteredSpecs$Clade!="Plantae","Species"]
Plant_tree <- drop.tip(tree, non_plants)

# The rectangular layout tree.
pp <- ggtree(Plant_tree, layout="rectangular", size=2, color = "snow4")

pp <- pp %<+% sorted_merged_summary_table_filteredSpecs + 
  geom_star(mapping=aes(fill=Clade, color = Clade, x = 165), size = 4, starshape=13,
                        position="identity",starstroke=0.1) +
  scale_fill_manual(values=ColoredClades,
                          name ="",
                           guide=guide_legend(keywidth = 0.5, 
                                        keyheight = 0.5, order=1),
                           na.translate=FALSE) +
  scale_color_manual(values=ColoredClades,
                          name ="",
                           guide=guide_legend(keywidth = 0.5, 
                                        keyheight = 0.5, order=1),
                           na.translate=FALSE)

pp <- pp  + new_scale_fill() + 
  geom_fruit(geom=geom_tile,
                    mapping=aes(y=label, fill=mu3),
             offset = 0.1, size = 0.02, width = 10) +
         scale_fill_continuous(name = "CHG > T mutational shift (log)",
                               low = "white",
                               high = "goldenrod",
                               #breaks = c(0.2, 0.4, 0.6, 0.8, 1),
                               #limits = c(0.2, 1),
                               #oob = scales::squish,
                               guide=guide_legend(keywidth = 0.5, 
                                        keyheight = 0.3, order=2)) +
         geom_treescale(fontsize=2, linesize=0.3, x=4.9, y=0.1) +
         theme(legend.position=c(0.93, 0.5),
               legend.background=element_rect(fill=NA),
               legend.title=element_text(size=6.5),
               legend.text=element_text(size=4.5),
               legend.spacing.y = unit(0.02, "cm"),
             )

pp <- pp  + new_scale_fill() + 
  geom_fruit(geom=geom_tile,
                    mapping=aes(y=label, fill=`O/E CHG content`),
             offset = 0.1, size = 0.02, width = 10) +
         scale_fill_continuous(name = "O/E CHG",
                               low = "white",
                               high = "#21b479",
                               breaks = c(0.6, 0.7, 0.8, 0.9, 1),
                               limits = c(0.6, 1),
                               oob = scales::squish,
                               guide=guide_legend(keywidth = 0.5, 
                                        keyheight = 0.3, order=3)) +
         geom_treescale(fontsize=2, linesize=0.3, x=4.9, y=0.1) +
         theme(legend.position=c(0.93, 0.5),
               legend.background=element_rect(fill=NA),
               legend.title=element_text(size=6.5),
               legend.text=element_text(size=4.5),
               legend.spacing.y = unit(0.02, "cm"),
             )

pp <- pp + new_scale_fill() +
         geom_fruit(geom=geom_tile_pattern,
                  mapping=aes(y=label, fill=`Average CHG Methylation`, pattern = ifelse(is.na(`Average CHG Methylation`), "stripe", NA)),
                  offset = 0.1,size = 0.02, width = 10,
                  pattern = "stripe",
                  pattern_color = "transparent",
                  pattern_fill = "transparent") +
         scale_fill_continuous(na.value = "gray15",
                               low = "white",
                               high = "mediumpurple",
                                name = "mCHG%",
                             guide=guide_legend(keywidth = 0.5, 
                                             keyheight = 0.3, order=4)) +
  scale_pattern_color_manual(na.value = "white") +
  scale_pattern_fill_manual(na.value = "white")

pp <- pp + theme(legend.position=c(.25, .82), legend.title = element_text(size = 16), legend.text = element_text(size = 15))
```

# Figure 4
```{r}
pdf("Figure4_Plants.pdf", 16, 10)
plot_grid(pp, NULL, plot_grid(mCHG_CHG_scatter, Residuals_CHG_CpG_MethylModels_Plants_scatter, ncol = 1, align = "v", hjust = 1, labels = c("B", "C"), label_x = 0.3, label_size = 24), ncol = 3, rel_widths = c(1.2, -0.75, 2.6), labels = c("A", "", ""), label_x = 0.4, label_size = 24)
dev.off()
```

# RMSE of k/9-mer
```{r}
PredictRMSECurve_data <- read.delim('PredKmer_ODDvs9merEVEN_RMSEs.txt', sep = "\t", header = FALSE)
colnames(PredictRMSECurve_data) <- c("Species","Mer1","Mer2","Mer3","Mer4","Mer5","Mer6","Mer7","Mer8","Mer9")
PredictRMSECurve_data$Species <- sapply(PredictRMSECurve_data$Species, function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})

ForRMSECurves <- merge(sorted_merged_summary_table_filteredSpecs, PredictRMSECurve_data, by = "Species")
ForRMSECurves$NormedMer1 <- ForRMSECurves$Mer1/ForRMSECurves$Mer1
ForRMSECurves$NormedMer2 <- ForRMSECurves$Mer2/ForRMSECurves$Mer1
ForRMSECurves$NormedMer3 <- ForRMSECurves$Mer3/ForRMSECurves$Mer1
ForRMSECurves$NormedMer4 <- ForRMSECurves$Mer4/ForRMSECurves$Mer1
ForRMSECurves$NormedMer5 <- ForRMSECurves$Mer5/ForRMSECurves$Mer1
ForRMSECurves$NormedMer6 <- ForRMSECurves$Mer6/ForRMSECurves$Mer1
ForRMSECurves$NormedMer7 <- ForRMSECurves$Mer7/ForRMSECurves$Mer1
ForRMSECurves$NormedMer8 <- ForRMSECurves$Mer8/ForRMSECurves$Mer1
ForRMSECurves$NormedMer9 <- ForRMSECurves$Mer9/ForRMSECurves$Mer1

longer_ForRMSECurves <- pivot_longer(ForRMSECurves, cols = contains("NormedMer"), names_to = "ContextLayer", values_to = "RMSEk")

cladeAveraged_longer_ForRMSECurves <- aggregate(RMSEk ~ Clade + ContextLayer, data = longer_ForRMSECurves, FUN = mean)
colnames(cladeAveraged_longer_ForRMSECurves) <- c("Clade", "ContextLayer", "MeanRMSE")
cladeAveraged_longer_ForRMSECurvesSD <- aggregate(RMSEk ~ Clade + ContextLayer, data = longer_ForRMSECurves, FUN = sd)
colnames(cladeAveraged_longer_ForRMSECurvesSD) <- c("Clade", "ContextLayer", "SdRMSE")

cladeAveraged_longer_ForRMSECurves$SdRMSE <- cladeAveraged_longer_ForRMSECurvesSD$SdRMSE

cladeAveraged_longer_ForRMSECurves$ContextLayer <- case_when(
  grepl("NormedMer1", cladeAveraged_longer_ForRMSECurves$ContextLayer) ~ "1-mer",
  grepl("NormedMer2", cladeAveraged_longer_ForRMSECurves$ContextLayer) ~ "2-mer",
  grepl("NormedMer3", cladeAveraged_longer_ForRMSECurves$ContextLayer) ~ "3-mer",
  grepl("NormedMer4", cladeAveraged_longer_ForRMSECurves$ContextLayer) ~ "4-mer",
  grepl("NormedMer5", cladeAveraged_longer_ForRMSECurves$ContextLayer) ~ "5-mer",
  grepl("NormedMer6", cladeAveraged_longer_ForRMSECurves$ContextLayer) ~ "6-mer",
  grepl("NormedMer7", cladeAveraged_longer_ForRMSECurves$ContextLayer) ~ "7-mer",
  grepl("NormedMer8", cladeAveraged_longer_ForRMSECurves$ContextLayer) ~ "8-mer",
  grepl("NormedMer9", cladeAveraged_longer_ForRMSECurves$ContextLayer) ~ "9-mer",
  TRUE ~ "NA"
)

RMSECurvesPlot <- ggplot(longer_ForRMSECurves) + geom_line(aes(x = ContextLayer, y = RMSEk, group = Species, color = Clade), alpha = 0.75) +
  scale_color_manual(values = ColoredClades) +
  scale_x_discrete(labels = c("NormedMer1" = "1-mer",
                              "NormedMer2" = "2-mer",
                              "NormedMer3" = "3-mer",
                              "NormedMer4" = "4-mer",
                              "NormedMer5" = "5-mer",
                              "NormedMer6" = "6-mer",
                              "NormedMer7" = "7-mer",
                              "NormedMer8" = "8-mer",
                              "NormedMer9" = "9-mer")) +
  theme_minimal() +
  theme(panel.background = element_rect(fill = "white"),
        panel.grid = element_line(color = "tan", linetype = 2, linewidth = 0.2),
        axis.text=element_text(size = 16, color = "black"),
        axis.text.x = element_text(angle = 70, hjust = 1, size = 12, color = "black"),
        title = element_blank(),
        axis.title.x = element_text(size = 18, color = "black"),
        axis.title.y = element_text(size = 18, color = "black"),
        aspect.ratio = 1,
        legend.title = element_blank(),
        legend.position = "none",
        legend.direction = "vertical",
        legend.text = element_text(size = 12, color = "black"),
        legend.background = element_rect(color = "transparent", fill = "transparent"),
        strip.text = element_text(size = 12, color = "black"),
        strip.background = element_rect(fill = "white", color = "black")) +
  labs(title = "", y = "Relative 9-mer RMSE", x = "") +
  facet_wrap(~Clade)

PerClade_RMSECurvesPlot <- ggplot(cladeAveraged_longer_ForRMSECurves %>% filter(Clade %in% c("Mammalia", "Invertebrata", "Aves", "Osteichthyes", "Plantae"))) + 
  geom_ribbon(aes(x = ContextLayer, ymin = (MeanRMSE - SdRMSE), ymax = (MeanRMSE + SdRMSE), group=Clade, fill=Clade), alpha=0.25) +
  geom_line(aes(x = ContextLayer, y = MeanRMSE, group = Clade, color = Clade), alpha = 1) +
  geom_point(aes(x = ContextLayer, y = MeanRMSE, group = Clade, color = Clade), alpha = 1) +
  scale_color_manual(values = ColoredClades) +
  scale_fill_manual(values = ColoredClades) +
  theme_minimal() +
  theme(panel.background = element_rect(fill = "white"),
        panel.border = element_blank(),
        panel.grid = element_line(color = "tan", linetype = 2, linewidth = 0.2),
        axis.text=element_text(size = 16, color = "black"),
        axis.text.x = element_text(angle = 45, hjust = 1, size = 12, color = "black"),
        title = element_blank(),
        axis.title.x = element_text(size = 18, color = "black"),
        axis.title.y = element_text(size = 18, color = "black"),
        aspect.ratio = 1,
        legend.title = element_blank(),
        legend.position = "none",
        legend.direction = "vertical",
        legend.text = element_text(size = 12, color = "black"),
        legend.background = element_rect(color = "transparent", fill = "transparent")) +
  labs(title = "", y = "Relative 9-mer RMSE", x = "")
#plot(PerClade_RMSECurvesPlot)

pdf("RMSEkCurvePlots.pdf", 8, 8)
RMSECurvesPlot
dev.off()
```

# PCA 3mers
```{r}
library(ggfortify)
Mer3_species_data <- read.delim('Merged_layer2_Probabilities.txt', sep = "\t", header = FALSE)
Mer3_species_data[1,2:(length(Mer3_species_data[1,]))] <- sapply(Mer3_species_data[1,2:(length(Mer3_species_data[1,]))], function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})
df_Mer3_species_data <- Mer3_species_data[2:(length(Mer3_species_data[,1])),2:(length(Mer3_species_data[1,]))]
df_Mer3_species_data <- sapply(df_Mer3_species_data, as.numeric)

####
mat_sigfit_species_data <- as.matrix(df_Mer3_species_data)
# Normalize each column
for (i in 1:ncol(mat_sigfit_species_data)) {
  mat_sigfit_species_data[, i] <- mat_sigfit_species_data[, i] / sum(mat_sigfit_species_data[, i], na.rm = TRUE)
}

# Convert back to data frame
df_Mer3_species_data <- as.data.frame(mat_sigfit_species_data)
####

colnames(df_Mer3_species_data) <- c(Mer3_species_data[1,2:(length(Mer3_species_data[1,]))])
rownames(df_Mer3_species_data) <- c(Mer3_species_data[2:(length(Mer3_species_data[,1])),1])

df_Mer3_species_data <- t(df_Mer3_species_data)

df_Mer3_species_data <- as.data.frame(df_Mer3_species_data)

df_Mer3_species_data$Species <- as.vector(rownames(df_Mer3_species_data))
for_pca_plot <- merge(df_Mer3_species_data, classified_species, by.x="Species",by.y="Species", by.all=TRUE)
# remove bad species
for_pca_plot <- for_pca_plot[!(for_pca_plot$Species %in% excluded_species),]

for_pca_run <- for_pca_plot[,2:(length(Mer3_species_data[,1]))]

for_pca_run <- sapply(for_pca_run, as.numeric)
rownames(for_pca_run) <- for_pca_plot$Species


pca_test <- prcomp(for_pca_run)
pca_test_3mer <- pca_test

### manual plotting
pca_3mer_summary <- summary(pca_test)
var_explained <- pca_3mer_summary$importance[2, ]
pc1_var <- round(var_explained[1] * 100, 2)
pc2_var <- round(var_explained[2] * 100, 2)
pc3_var <- round(var_explained[3] * 100, 2)
pc4_var <- round(var_explained[4] * 100, 2)

pca_scores <- as.data.frame(pca_test$x)

# Combine with rate data
pca_data <- cbind(for_pca_plot, pca_scores)
pca_data$Methyl <- sorted_merged_summary_table[pca_data$Species,]$CpGtoTPhi
# Plot
#PC 1 and 2
pca_3mers_plot <- ggplot(pca_data, aes(x = PC1, y = PC2, color = Class)) +
  geom_hline(yintercept = 0, color = "tan", linetype = "dashed", linewidth = 0.5) + 
  geom_vline(xintercept = 0, color = "tan", linetype = "dashed", linewidth = 0.5) +
  geom_point(size = 5, alpha = 0.8, pch=16) +
  theme_minimal() +
  theme(panel.background = element_rect(fill = "white"),
        panel.grid = element_blank(),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        title = element_text(size = 22, face = "bold", color = "blueviolet"),
        legend.position = "bottom",
        legend.direction = "horizontal",
        legend.title = element_blank(),
        legend.text = element_text(size=16, color="black"),
        legend.background = element_rect(fill = "transparent"),
        aspect.ratio = 1,
        panel.border = element_blank()) +
  labs(title = "",
       x = paste0("PC1 (", pc1_var, "%)"),
       y = paste0("PC2 (", pc2_var, "%)")) +
  scale_color_manual(values = ColoredClades) +
  guides(color = guide_legend(nrow = 3))
# PC 3 and 4
pca_PC3_PC4_3mers_plot <- ggplot(pca_data, aes(x = PC3, y = PC4, color = Class)) +
  geom_hline(yintercept = 0, color = "tan", linetype = "dashed", linewidth = 0.5) + 
  geom_vline(xintercept = 0, color = "tan", linetype = "dashed", linewidth = 0.5) +
  geom_point(size = 5, alpha = 0.8, pch=16) +
  theme_minimal() +
  theme(panel.background = element_rect(fill = "white"),
        panel.grid = element_blank(),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, face = "bold", color = "black"),
        title = element_text(size = 18, face = "bold", color = "blueviolet"),
        legend.position = "none",
        legend.title = element_blank(),
        legend.background = element_rect(fill = "transparent"),
        aspect.ratio = 1,
        panel.border = element_blank()) +
  labs(title = "",
       x = paste0("PC3 (", pc3_var, "%)"),
       y = paste0("PC4 (", pc4_var, "%)")) +
  scale_color_manual(values = ColoredClades)
###

loadings_3mers <- as.data.frame(pca_test_3mer$rotation)
loadings_3mers$Context <- "empty"
loadings_3mers$Variable <- rownames(loadings_3mers)
loadings_3mers$Context <- case_when(
  grepl("C>A]G", rownames(loadings_3mers)) ~ "CpG",
  grepl("C>T]G", rownames(loadings_3mers)) ~ "CpG",
  grepl("C>G]G", rownames(loadings_3mers)) ~ "CpG",
  TRUE ~ "NonCpG"
)

loadings_3mers$Type <- case_when(
  grepl("C>A\\]G", rownames(loadings_3mers)) ~ "C_to_A_at_CpG",
  grepl("C>T\\]G", rownames(loadings_3mers)) ~ "C_to_T_at_CpG",
  grepl("C>G\\]G", rownames(loadings_3mers)) ~ "C_to_G_at_CpG",
  grepl("C>A", rownames(loadings_3mers)) ~ "C_to_A",
  grepl("C>T", rownames(loadings_3mers)) ~ "C_to_T",
  grepl("C>G", rownames(loadings_3mers)) ~ "C_to_G",
  grepl("A>G", rownames(loadings_3mers)) ~ "A_to_G",
  grepl("A>C", rownames(loadings_3mers)) ~ "A_to_C",
  grepl("A>T", rownames(loadings_3mers)) ~ "A_to_T",
  TRUE ~ "NA"
)

# plot PC variances
MutType_order <- c("C_to_T_at_CpG", "C_to_G_at_CpG", "C_to_A_at_CpG",
                "C_to_T", "C_to_G", "C_to_A", 
                "A_to_T", "A_to_C", "A_to_G")

loadings_3mers <- loadings_3mers %>%
  mutate(Type = factor(Type, levels = MutType_order)) %>%
  arrange(Type)

VarianceColoredLabels <- c(
    "C_to_T_at_CpG" = "#AC9ECEFF",
    "C_to_G_at_CpG" = "#6EC5ABFF",
    "C_to_A_at_CpG" = "#DAA5ACFF",
    "C_to_T" = "#B6E7E0FF",
    "C_to_G" = "#AA3F5DFF",
    "C_to_A" = "#98A54FFF",
    "A_to_T" = "#FFB651FF",
    "A_to_C" = "#D85A44FF",
    "A_to_G" = "#751C6DFF")

VarianceModifiedLabels <- c("CpG > TpG", "CpG > GpG", "CpG > ApG", "C > T", "C > G", "C > A", "A > T", "A > C", "A > G")
PC1_variance_bars_3mers <- ggplot() + 
  geom_bar(data=loadings_3mers,
           aes(x=factor(Variable, loadings_3mers$Variable), y=PC1, fill=Type), stat = "identity", width = 1) +
  scale_fill_manual(labels = VarianceModifiedLabels,
                    values = VarianceColoredLabels) +
  theme(axis.text.x = element_blank(),
        title = element_text(size = 20, color = "black"),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.position = c(0.7, 0.87),
        legend.title = element_blank(),
        legend.background = element_rect(fill = "white", color = "snow4"),
        legend.direction = "horizontal",
        legend.text = element_text(size=11, color="black"),
        legend.box = "horizontal",
        legend.frame = element_rect(fill = "black")) +
  labs(title = "",
       x = "",
       y = paste0("PC1 (", pc1_var, "%)"))

PC2_variance_bars_3mers <- ggplot() + 
  geom_bar(data=loadings_3mers,
           aes(x=factor(Variable, loadings_3mers$Variable), y=PC2, fill=Type), stat = "identity", width = 1) +
  scale_fill_manual(labels = VarianceModifiedLabels,
                    values = VarianceColoredLabels) +
  theme(axis.text.x = element_blank(),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.position = "none",
        legend.title = element_blank(),
        legend.background = element_rect(fill = "transparent")) +
  labs(title = "",
       x = "",
       y = paste0("PC2 (", pc2_var, "%)"))

PC3_variance_bars_3mers <- ggplot() + 
  geom_bar(data=loadings_3mers,
           aes(x=factor(Variable, loadings_3mers$Variable), y=PC3, fill=Type), stat = "identity") +
  scale_fill_manual(labels = c("CpG > TpG", "CpG > GpG", "CpG > ApG", "C > T", "C > G", "C > A", "A > T", "A > C", "A > G"), 
                    values = c(
    "C_to_T_at_CpG" = "cornflowerblue",
    "C_to_G_at_CpG" = "indianred1",
    "C_to_A_at_CpG" = "seagreen",
    "C_to_T" = "aquamarine",
    "C_to_G" = "brown",
    "C_to_A" = "darkseagreen",
    "A_to_T" = "plum1",
    "A_to_C" = "palevioletred",
    "A_to_G" = "mediumpurple1")) +
  theme(axis.text.x = element_text(angle = 70, hjust = 1),
        panel.grid.minor = element_line(color = "snow4", linewidth = 0.5, linetype = 1),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_line(color = "tan", linewidth = 0.2, linetype = 2),
        panel.background = element_rect(fill = "white"),
        legend.position = "none",
        legend.title = element_blank(),
        legend.background = element_rect(fill = "transparent")) +
  labs(title = "PC3 Variance Explained by Mutation Type",
       x = "",
       y = paste0("PC3 (", pc3_var, "%)"))


PC4_variance_bars_3mers <- ggplot() + 
  geom_bar(data=loadings_3mers,
           aes(x=factor(Variable, loadings_3mers$Variable), y=PC4, fill=Type), stat = "identity") +
  scale_fill_manual(labels = c("CpG > TpG", "CpG > GpG", "CpG > ApG", "C > T", "C > G", "C > A", "A > T", "A > C", "A > G"),
                    values = c(
    "C_to_T_at_CpG" = "cornflowerblue",
    "C_to_G_at_CpG" = "indianred1",
    "C_to_A_at_CpG" = "seagreen",
    "C_to_T" = "aquamarine",
    "C_to_G" = "brown",
    "C_to_A" = "darkseagreen",
    "A_to_T" = "plum1",
    "A_to_C" = "palevioletred",
    "A_to_G" = "mediumpurple1")) +
  theme(axis.text.x = element_text(angle = 70, hjust = 1),
        panel.grid.minor = element_line(color = "snow4", linewidth = 0.5, linetype = 1),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_line(color = "tan", linewidth = 0.2, linetype = 2),
        panel.background = element_rect(fill = "white"),
        legend.position = "none",
        legend.title = element_blank(),
        legend.background = element_rect(fill = "transparent")) +
  labs(title = "PC4 Variance Explained by Mutation Type",
       x = "",
       y = paste0("PC4 (", pc4_var, "%)"))


pdf("PCA_3mers.pdf", 18, 9)
plot_grid(pca_3mers_plot, plot_grid(PC1_variance_bars_3mers, PC2_variance_bars_3mers, ncol = 1), ncol = 2)
dev.off()
```
# 5-mer PCA
```{r}
library(ggfortify)
Mer5_species_data <- read.delim('Merged_layer4_Probabilities.txt', sep = "\t", header = FALSE)
Mer5_species_data[1,2:(length(Mer5_species_data[1,]))] <- sapply(Mer5_species_data[1,2:(length(Mer5_species_data[1,]))], function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})
df_Mer5_species_data <- Mer5_species_data[2:(length(Mer5_species_data[,1])),2:(length(Mer5_species_data[1,]))]
df_Mer5_species_data <- sapply(df_Mer5_species_data, as.numeric)

####
mat_sigfit_species_data <- as.matrix(df_Mer5_species_data)
# Normalize each column
for (i in 1:ncol(mat_sigfit_species_data)) {
  mat_sigfit_species_data[, i] <- mat_sigfit_species_data[, i] / sum(mat_sigfit_species_data[, i], na.rm = TRUE)
}

# Convert back to data frame
df_Mer5_species_data <- as.data.frame(mat_sigfit_species_data)
####

colnames(df_Mer5_species_data) <- c(Mer5_species_data[1,2:(length(Mer5_species_data[1,]))])
rownames(df_Mer5_species_data) <- c(Mer5_species_data[2:(length(Mer5_species_data[,1])),1])

df_Mer5_species_data <- t(df_Mer5_species_data)

df_Mer5_species_data <- as.data.frame(df_Mer5_species_data)

df_Mer5_species_data$Species <- as.vector(rownames(df_Mer5_species_data))
for_pca_plot <- merge(df_Mer5_species_data, classified_species, by.x="Species",by.y="Species", by.all=TRUE)
# remove bad species
for_pca_plot <- for_pca_plot[!(for_pca_plot$Species %in% excluded_species),]

for_pca_run <- for_pca_plot[,2:(length(Mer5_species_data[,1]))]

for_pca_run <- sapply(for_pca_run, as.numeric)
rownames(for_pca_run) <- for_pca_plot$Species


pca_test <- prcomp(for_pca_run, scale = FALSE, center = TRUE)
pca_test_5mer <- pca_test

### manual plotting
pca_5mer_summary <- summary(pca_test)
var_explained <- pca_5mer_summary$importance[2, ]
pc1_var <- round(var_explained[1] * 100, 2)
pc2_var <- round(var_explained[2] * 100, 2)
pc3_var <- round(var_explained[3] * 100, 2)
pc4_var <- round(var_explained[4] * 100, 2)

pca_scores <- as.data.frame(pca_test$x)

pc_explained_variance_1_to_10 <- pca_5mer_summary$importance[2, 1:10]
# Combine with rate data
pca_data <- cbind(for_pca_plot, pca_scores)
pca_data$Methyl <- sorted_merged_summary_table_filteredSpecs[pca_data$Species,]$`Average CpG Methylation`
pca_data$CpGPhi <- sorted_merged_summary_table_filteredSpecs[pca_data$Species,]$CpGtoTPhi
pca_data$CHGPhi <- sorted_merged_summary_table_filteredSpecs[pca_data$Species,]$CHGtoTPhi
pca_data$N50 <- sorted_merged_summary_table_filteredSpecs[pca_data$Species,]$ContigN50
pca_data$CHGMethyl <- sorted_merged_summary_table_filteredSpecs[pca_data$Species,]$`Average CHG Methylation`
pca_data$BodyTemp <- sorted_merged_summary_table_filteredSpecs[pca_data$Species,]$`Body Temperature`
pca_data$NoSNPs <- sorted_merged_summary_table_filteredSpecs[pca_data$Species,]$`Non-Genic SNPs (log10)`
pca_data$Thermy <- sorted_merged_summary_table_filteredSpecs[pca_data$Species,]$Thermy
pca_data$L50 <- sorted_merged_summary_table_filteredSpecs[pca_data$Species,]$ContigL50
pca_data$RMSE <- sorted_merged_summary_table_filteredSpecs[pca_data$Species,]$`Odd vs Even RMSE`
pca_data$Random1 <- runif(108)
pca_data$Random2 <- runif(108)
# Plot
#PC 1 and 2
pca_5mers_plot <- ggplot(pca_data, aes(x = PC1, y = PC2, color = Class, fill = Class)) +
  geom_hline(yintercept = 0, color = "tan", linetype = "dashed", linewidth = 0.5) + 
  geom_vline(xintercept = 0, color = "tan", linetype = "dashed", linewidth = 0.5) +
  geom_point(size = 7, alpha = 0.75, pch=21) +
  theme_minimal() +
  theme(panel.background = element_rect(fill = "white"),
        panel.grid = element_blank(),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        title = element_text(size = 22, face = "bold", color = "blueviolet"),
        legend.position = "none",
        legend.direction = "horizontal",
        legend.title = element_blank(),
        legend.text = element_text(size=16, color="black"),
        legend.background = element_rect(fill = "transparent"),
        aspect.ratio = 1,
        panel.border = element_blank()) +
  labs(title = "",
       x = paste0("PC1 (", pc1_var, "%)"),
       y = paste0("PC2 (", pc2_var, "%)")) +
  scale_color_manual(values = ColoredClades) +
  scale_fill_manual(values = ColoredClades) +
  guides(color = guide_legend(nrow = 3))

# PC 3 and 4
pca_PC3_PC4_5mers_plot <- ggplot(pca_data, aes(x = PC3, y = PC4, color = Class)) +
  geom_hline(yintercept = 0, color = "tan", linetype = "dashed", linewidth = 0.5) + 
  geom_vline(xintercept = 0, color = "tan", linetype = "dashed", linewidth = 0.5) +
  geom_point(size = 5, alpha = 0.8, pch=16) +
  theme_minimal() +
  theme(panel.background = element_rect(fill = "white"),
        panel.grid = element_blank(),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, face = "bold", color = "black"),
        title = element_text(size = 22, face = "bold", color = "blueviolet"),
        legend.position = "bottom",
        legend.title = element_blank(),
        legend.text = element_text(size=16, color="black"),
        legend.background = element_rect(fill = "transparent"),
        aspect.ratio = 1,
        panel.border = element_blank()) +
  labs(title = "",
       x = paste0("PC3 (", pc3_var, "%)"),
       y = paste0("PC4 (", pc4_var, "%)")) +
  scale_color_manual(values = ColoredClades)

pca_PC5_PC6_5mers_plot <- ggplot(pca_data, aes(x = PC5, y = PC6, color = Class)) +
  geom_hline(yintercept = 0, color = "tan", linetype = "dashed", linewidth = 0.5) + 
  geom_vline(xintercept = 0, color = "tan", linetype = "dashed", linewidth = 0.5) +
  geom_point(size = 5, alpha = 0.8, pch=16) +
  theme_minimal() +
  theme(panel.background = element_rect(fill = "white"),
        panel.grid = element_blank(),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, face = "bold", color = "black"),
        title = element_text(size = 22, face = "bold", color = "blueviolet"),
        legend.position = "none",
        legend.title = element_blank(),
        legend.background = element_rect(fill = "transparent"),
        aspect.ratio = 1,
        panel.border = element_blank()) +
  labs(title = "",
       x = paste0("PC5 (", "%)"),
       y = paste0("PC6 (", "%)")) +
  scale_color_manual(values = ColoredClades)
###

loadings_5mers <- as.data.frame(pca_test_5mer$rotation)
loadings_5mers$Context <- "empty"
loadings_5mers$Variable <- rownames(loadings_5mers)
loadings_5mers$Context <- case_when(
  grepl("C>A]G", rownames(loadings_5mers)) ~ "CpG",
  grepl("C>T]G", rownames(loadings_5mers)) ~ "CpG",
  grepl("C>G]G", rownames(loadings_5mers)) ~ "CpG",
  TRUE ~ "NonCpG"
)

loadings_5mers$Type <- case_when(
  grepl("C>A\\]G", rownames(loadings_5mers)) ~ "CpG > ApG",
  grepl("C>T\\]G", rownames(loadings_5mers)) ~ "CpG > TpG",
  grepl("C>G\\]G", rownames(loadings_5mers)) ~ "CpG > GpG",
  grepl("C>T\\]AG", rownames(loadings_5mers)) ~ "CHG > THG",
  grepl("C>T\\]CG", rownames(loadings_5mers)) ~ "CHG > THG",
  grepl("C>T\\]TG", rownames(loadings_5mers)) ~ "CHG > THG",
  grepl("C>A", rownames(loadings_5mers)) ~ "C > A",
  grepl("C>T", rownames(loadings_5mers)) ~ "C > T",
  grepl("C>G", rownames(loadings_5mers)) ~ "C > G",
  grepl("A>G", rownames(loadings_5mers)) ~ "A > G",
  grepl("A>C", rownames(loadings_5mers)) ~ "A > C",
  grepl("A>T", rownames(loadings_5mers)) ~ "A > T",
  TRUE ~ "NA"
)

loadings_5mers$TypeOfCpG_Binary <- case_when(
  grepl("AA\\[C>T\\]GA", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AA\\[C>T\\]GC", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AA\\[C>T\\]GT", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AA\\[C>T\\]GG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("AC\\[C>T\\]GA", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AC\\[C>T\\]GC", rownames(loadings_5mers)) ~ "2+ S",
  grepl("AC\\[C>T\\]GT", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AC\\[C>T\\]GG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("AT\\[C>T\\]GA", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AT\\[C>T\\]GC", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AT\\[C>T\\]GT", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AT\\[C>T\\]GG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("AG\\[C>T\\]GA", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AG\\[C>T\\]GC", rownames(loadings_5mers)) ~ "2+ S",
  grepl("AG\\[C>T\\]GT", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AG\\[C>T\\]GG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("CA\\[C>T\\]GA", rownames(loadings_5mers)) ~ "2+ W",
  grepl("CA\\[C>T\\]GC", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CA\\[C>T\\]GT", rownames(loadings_5mers)) ~ "2+ W",
  grepl("CA\\[C>T\\]GG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("CC\\[C>T\\]GA", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CC\\[C>T\\]GC", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CC\\[C>T\\]GT", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CC\\[C>T\\]GG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("CT\\[C>T\\]GA", rownames(loadings_5mers)) ~ "2+ W",
  grepl("CT\\[C>T\\]GC", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CT\\[C>T\\]GT", rownames(loadings_5mers)) ~ "2+ W",
  grepl("CT\\[C>T\\]GG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("CG\\[C>T\\]GA", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CG\\[C>T\\]GC", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CG\\[C>T\\]GT", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CG\\[C>T\\]GG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("TA\\[C>T\\]GA", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TA\\[C>T\\]GC", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TA\\[C>T\\]GT", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TA\\[C>T\\]GG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("TC\\[C>T\\]GA", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TC\\[C>T\\]GC", rownames(loadings_5mers)) ~ "2+ S",
  grepl("TC\\[C>T\\]GT", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TC\\[C>T\\]GG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("TT\\[C>T\\]GA", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TT\\[C>T\\]GC", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TT\\[C>T\\]GT", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TT\\[C>T\\]GG", rownames(loadings_5mers)) ~ "2+ W",

  grepl("TG\\[C>T\\]GA", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TG\\[C>T\\]GC", rownames(loadings_5mers)) ~ "2+ S",
  grepl("TG\\[C>T\\]GT", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TG\\[C>T\\]GG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("GA\\[C>T\\]GA", rownames(loadings_5mers)) ~ "2+ W",
  grepl("GA\\[C>T\\]GC", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GA\\[C>T\\]GT", rownames(loadings_5mers)) ~ "2+ W",
  grepl("GA\\[C>T\\]GG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("GC\\[C>T\\]GA", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GC\\[C>T\\]GC", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GC\\[C>T\\]GT", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GC\\[C>T\\]GG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("GT\\[C>T\\]GA", rownames(loadings_5mers)) ~ "2+ W",
  grepl("GT\\[C>T\\]GC", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GT\\[C>T\\]GT", rownames(loadings_5mers)) ~ "2+ W",
  grepl("GT\\[C>T\\]GG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("GG\\[C>T\\]GA", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GG\\[C>T\\]GC", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GG\\[C>T\\]GT", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GG\\[C>T\\]GG", rownames(loadings_5mers)) ~ "2+ S",
  TRUE ~ "NotCpG"
)

loadings_5mers$TypeOfCHG_Binary <- case_when(
  grepl("AA\\[C>T\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AA\\[C>T\\]CG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AA\\[C>T\\]TG", rownames(loadings_5mers)) ~ "2+ W",

  grepl("AC\\[C>T\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AC\\[C>T\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("AC\\[C>T\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("AT\\[C>T\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AT\\[C>T\\]CG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AT\\[C>T\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("AG\\[C>T\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AG\\[C>T\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("AG\\[C>T\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("CA\\[C>T\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("CA\\[C>T\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CA\\[C>T\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("CC\\[C>T\\]AG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CC\\[C>T\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CC\\[C>T\\]TG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("CT\\[C>T\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("CT\\[C>T\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CT\\[C>T\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("CG\\[C>T\\]AG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CG\\[C>T\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CG\\[C>T\\]TG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("TA\\[C>T\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TA\\[C>T\\]CG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TA\\[C>T\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("TC\\[C>T\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TC\\[C>T\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("TC\\[C>T\\]TG", rownames(loadings_5mers)) ~ "2+ W",
 
  grepl("TT\\[C>T\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TT\\[C>T\\]CG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TT\\[C>T\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("TG\\[C>T\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TG\\[C>T\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("TG\\[C>T\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("GA\\[C>T\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("GA\\[C>T\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GA\\[C>T\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("GC\\[C>T\\]AG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GC\\[C>T\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GC\\[C>T\\]TG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("GT\\[C>T\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("GT\\[C>T\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GT\\[C>T\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("GG\\[C>T\\]AG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GG\\[C>T\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GG\\[C>T\\]TG", rownames(loadings_5mers)) ~ "2+ S",
  TRUE ~ "NotCHG"
)

loadings_5mers$TypeOfCH_to_A_Binary <- case_when(
  grepl("AA\\[C>A\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AA\\[C>A\\]CG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AA\\[C>A\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("AC\\[C>A\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AC\\[C>A\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("AC\\[C>A\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("AT\\[C>A\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AT\\[C>A\\]CG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AT\\[C>A\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("AG\\[C>A\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("AG\\[C>A\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("AG\\[C>A\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("CA\\[C>A\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("CA\\[C>A\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CA\\[C>A\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("CC\\[C>A\\]AG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CC\\[C>A\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CC\\[C>A\\]TG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("CT\\[C>A\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("CT\\[C>A\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CT\\[C>A\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("CG\\[C>A\\]AG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CG\\[C>A\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("CG\\[C>A\\]TG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("TA\\[C>A\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TA\\[C>A\\]CG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TA\\[C>A\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("TC\\[C>A\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TC\\[C>A\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("TC\\[C>A\\]TG", rownames(loadings_5mers)) ~ "2+ W",
 
  grepl("TT\\[C>A\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TT\\[C>A\\]CG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TT\\[C>A\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("TG\\[C>A\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("TG\\[C>A\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("TG\\[C>A\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("GA\\[C>A\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("GA\\[C>A\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GA\\[C>A\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("GC\\[C>A\\]AG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GC\\[C>A\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GC\\[C>A\\]TG", rownames(loadings_5mers)) ~ "2+ S",
  
  grepl("GT\\[C>A\\]AG", rownames(loadings_5mers)) ~ "2+ W",
  grepl("GT\\[C>A\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GT\\[C>A\\]TG", rownames(loadings_5mers)) ~ "2+ W",
  
  grepl("GG\\[C>A\\]AG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GG\\[C>A\\]CG", rownames(loadings_5mers)) ~ "2+ S",
  grepl("GG\\[C>A\\]TG", rownames(loadings_5mers)) ~ "2+ S",
  TRUE ~ "NotCHToA"
)

loadings_5mers$ForWilcoxonTestOfCpG <- case_when(
  grepl("AA\\[C>T\\]GA", rownames(loadings_5mers)) ~ 0,
  grepl("AA\\[C>T\\]GC", rownames(loadings_5mers)) ~ 0,
  grepl("AA\\[C>T\\]GT", rownames(loadings_5mers)) ~ 0,
  grepl("AA\\[C>T\\]GG", rownames(loadings_5mers)) ~ 0,
  
  grepl("AC\\[C>T\\]GA", rownames(loadings_5mers)) ~ 0,
  grepl("AC\\[C>T\\]GC", rownames(loadings_5mers)) ~ 1,
  grepl("AC\\[C>T\\]GT", rownames(loadings_5mers)) ~ 0,
  grepl("AC\\[C>T\\]GG", rownames(loadings_5mers)) ~ 1,
  
  grepl("AT\\[C>T\\]GA", rownames(loadings_5mers)) ~ 0,
  grepl("AT\\[C>T\\]GC", rownames(loadings_5mers)) ~ 0,
  grepl("AT\\[C>T\\]GT", rownames(loadings_5mers)) ~ 0,
  grepl("AT\\[C>T\\]GG", rownames(loadings_5mers)) ~ 0,
  
  grepl("AG\\[C>T\\]GA", rownames(loadings_5mers)) ~ 0,
  grepl("AG\\[C>T\\]GC", rownames(loadings_5mers)) ~ 1,
  grepl("AG\\[C>T\\]GT", rownames(loadings_5mers)) ~ 0,
  grepl("AG\\[C>T\\]GG", rownames(loadings_5mers)) ~ 1,
  
  grepl("CA\\[C>T\\]GA", rownames(loadings_5mers)) ~ 0,
  grepl("CA\\[C>T\\]GC", rownames(loadings_5mers)) ~ 1,
  grepl("CA\\[C>T\\]GT", rownames(loadings_5mers)) ~ 0,
  grepl("CA\\[C>T\\]GG", rownames(loadings_5mers)) ~ 1,
  
  grepl("CC\\[C>T\\]GA", rownames(loadings_5mers)) ~ 1,
  grepl("CC\\[C>T\\]GC", rownames(loadings_5mers)) ~ 1,
  grepl("CC\\[C>T\\]GT", rownames(loadings_5mers)) ~ 1,
  grepl("CC\\[C>T\\]GG", rownames(loadings_5mers)) ~ 1,
  
  grepl("CT\\[C>T\\]GA", rownames(loadings_5mers)) ~ 0,
  grepl("CT\\[C>T\\]GC", rownames(loadings_5mers)) ~ 1,
  grepl("CT\\[C>T\\]GT", rownames(loadings_5mers)) ~ 0,
  grepl("CT\\[C>T\\]GG", rownames(loadings_5mers)) ~ 1,
  
  grepl("CG\\[C>T\\]GA", rownames(loadings_5mers)) ~ 1,
  grepl("CG\\[C>T\\]GC", rownames(loadings_5mers)) ~ 1,
  grepl("CG\\[C>T\\]GT", rownames(loadings_5mers)) ~ 1,
  grepl("CG\\[C>T\\]GG", rownames(loadings_5mers)) ~ 1,
  
  grepl("TA\\[C>T\\]GA", rownames(loadings_5mers)) ~ 0,
  grepl("TA\\[C>T\\]GC", rownames(loadings_5mers)) ~ 0,
  grepl("TA\\[C>T\\]GT", rownames(loadings_5mers)) ~ 0,
  grepl("TA\\[C>T\\]GG", rownames(loadings_5mers)) ~ 0,
  
  grepl("TC\\[C>T\\]GA", rownames(loadings_5mers)) ~ 0,
  grepl("TC\\[C>T\\]GC", rownames(loadings_5mers)) ~ 1,
  grepl("TC\\[C>T\\]GT", rownames(loadings_5mers)) ~ 0,
  grepl("TC\\[C>T\\]GG", rownames(loadings_5mers)) ~ 1,
  
  grepl("TT\\[C>T\\]GA", rownames(loadings_5mers)) ~ 0,
  grepl("TT\\[C>T\\]GC", rownames(loadings_5mers)) ~ 0,
  grepl("TT\\[C>T\\]GT", rownames(loadings_5mers)) ~ 0,
  grepl("TT\\[C>T\\]GG", rownames(loadings_5mers)) ~ 0,
  
  grepl("TG\\[C>T\\]GA", rownames(loadings_5mers)) ~ 0,
  grepl("TG\\[C>T\\]GC", rownames(loadings_5mers)) ~ 1,
  grepl("TG\\[C>T\\]GT", rownames(loadings_5mers)) ~ 0,
  grepl("TG\\[C>T\\]GG", rownames(loadings_5mers)) ~ 1,
  
  grepl("GA\\[C>T\\]GA", rownames(loadings_5mers)) ~ 0,
  grepl("GA\\[C>T\\]GC", rownames(loadings_5mers)) ~ 1,
  grepl("GA\\[C>T\\]GT", rownames(loadings_5mers)) ~ 0,
  grepl("GA\\[C>T\\]GG", rownames(loadings_5mers)) ~ 1,
  
  grepl("GC\\[C>T\\]GA", rownames(loadings_5mers)) ~ 1,
  grepl("GC\\[C>T\\]GC", rownames(loadings_5mers)) ~ 1,
  grepl("GC\\[C>T\\]GT", rownames(loadings_5mers)) ~ 1,
  grepl("GC\\[C>T\\]GG", rownames(loadings_5mers)) ~ 1,
  
  grepl("GT\\[C>T\\]GA", rownames(loadings_5mers)) ~ 0,
  grepl("GT\\[C>T\\]GC", rownames(loadings_5mers)) ~ 1,
  grepl("GT\\[C>T\\]GT", rownames(loadings_5mers)) ~ 0,
  grepl("GT\\[C>T\\]GG", rownames(loadings_5mers)) ~ 1,
  
  grepl("GG\\[C>T\\]GA", rownames(loadings_5mers)) ~ 1,
  grepl("GG\\[C>T\\]GC", rownames(loadings_5mers)) ~ 1,
  grepl("GG\\[C>T\\]GT", rownames(loadings_5mers)) ~ 1,
  grepl("GG\\[C>T\\]GG", rownames(loadings_5mers)) ~ 1,
  TRUE ~ 0
)

loadings_5mers$ForWilcoxonTestOfCHG <- case_when(
  grepl("AA\\[C>T\\]AG", rownames(loadings_5mers)) ~ 0,
  grepl("AA\\[C>T\\]CG", rownames(loadings_5mers)) ~ 0,
  grepl("AA\\[C>T\\]TG", rownames(loadings_5mers)) ~ 0,
  
  grepl("AC\\[C>T\\]AG", rownames(loadings_5mers)) ~ 0,
  grepl("AC\\[C>T\\]CG", rownames(loadings_5mers)) ~ 1,
  grepl("AC\\[C>T\\]TG", rownames(loadings_5mers)) ~ 0,
  
  grepl("AT\\[C>T\\]AG", rownames(loadings_5mers)) ~ 0,
  grepl("AT\\[C>T\\]CG", rownames(loadings_5mers)) ~ 0,
  grepl("AT\\[C>T\\]TG", rownames(loadings_5mers)) ~ 0,
  
  grepl("AG\\[C>T\\]AG", rownames(loadings_5mers)) ~ 0,
  grepl("AG\\[C>T\\]CG", rownames(loadings_5mers)) ~ 1,
  grepl("AG\\[C>T\\]TG", rownames(loadings_5mers)) ~ 0,
  
  grepl("CA\\[C>T\\]AG", rownames(loadings_5mers)) ~ 0,
  grepl("CA\\[C>T\\]CG", rownames(loadings_5mers)) ~ 1,
  grepl("CA\\[C>T\\]TG", rownames(loadings_5mers)) ~ 0,
  
  grepl("CC\\[C>T\\]AG", rownames(loadings_5mers)) ~ 1,
  grepl("CC\\[C>T\\]CG", rownames(loadings_5mers)) ~ 1,
  grepl("CC\\[C>T\\]TG", rownames(loadings_5mers)) ~ 1,
  
  grepl("CT\\[C>T\\]AG", rownames(loadings_5mers)) ~ 0,
  grepl("CT\\[C>T\\]CG", rownames(loadings_5mers)) ~ 1,
  grepl("CT\\[C>T\\]TG", rownames(loadings_5mers)) ~ 0,
  
  grepl("CG\\[C>T\\]AG", rownames(loadings_5mers)) ~ 1,
  grepl("CG\\[C>T\\]CG", rownames(loadings_5mers)) ~ 1,
  grepl("CG\\[C>T\\]TG", rownames(loadings_5mers)) ~ 1,
  
  grepl("TA\\[C>T\\]AG", rownames(loadings_5mers)) ~ 0,
  grepl("TA\\[C>T\\]CG", rownames(loadings_5mers)) ~ 0,
  grepl("TA\\[C>T\\]TG", rownames(loadings_5mers)) ~ 0,
  
  grepl("TC\\[C>T\\]AG", rownames(loadings_5mers)) ~ 0,
  grepl("TC\\[C>T\\]CG", rownames(loadings_5mers)) ~ 1,
  grepl("TC\\[C>T\\]TG", rownames(loadings_5mers)) ~ 0,
 
  grepl("TT\\[C>T\\]AG", rownames(loadings_5mers)) ~ 0,
  grepl("TT\\[C>T\\]CG", rownames(loadings_5mers)) ~ 0,
  grepl("TT\\[C>T\\]TG", rownames(loadings_5mers)) ~ 0,
  
  grepl("TG\\[C>T\\]AG", rownames(loadings_5mers)) ~ 0,
  grepl("TG\\[C>T\\]CG", rownames(loadings_5mers)) ~ 1,
  grepl("TG\\[C>T\\]TG", rownames(loadings_5mers)) ~ 0,
  
  grepl("GA\\[C>T\\]AG", rownames(loadings_5mers)) ~ 0,
  grepl("GA\\[C>T\\]CG", rownames(loadings_5mers)) ~ 1,
  grepl("GA\\[C>T\\]TG", rownames(loadings_5mers)) ~ 0,
  
  grepl("GC\\[C>T\\]AG", rownames(loadings_5mers)) ~ 1,
  grepl("GC\\[C>T\\]CG", rownames(loadings_5mers)) ~ 1,
  grepl("GC\\[C>T\\]TG", rownames(loadings_5mers)) ~ 1,
  
  grepl("GT\\[C>T\\]AG", rownames(loadings_5mers)) ~ 0,
  grepl("GT\\[C>T\\]CG", rownames(loadings_5mers)) ~ 1,
  grepl("GT\\[C>T\\]TG", rownames(loadings_5mers)) ~ 0,
  
  grepl("GG\\[C>T\\]AG", rownames(loadings_5mers)) ~ 1,
  grepl("GG\\[C>T\\]CG", rownames(loadings_5mers)) ~ 1,
  grepl("GG\\[C>T\\]TG", rownames(loadings_5mers)) ~ 1,
  TRUE ~ 0
)
###
loadings_5mers %<>%
  rownames_to_column(var = "context") %>%      # move rownames into a column
  mutate(
    # extract bases from `context` exactly as before...
    left2   = str_sub(context, 1, 2),
    refBase = str_sub(context, 4, 4),
    right2  = str_sub(context, 8, 9),
    # build the 5-base matrix and classify
    all_bases = map2(left2, right2, ~ c(str_split(.x, "")[[1]], str_sub(context, 4,4), str_split(.y, "")[[1]])),
    W_counts  = map_int(all_bases, ~ sum(.x %in% c("A","T"))),
    TypeOfCH_to_A_Binary = case_when(
      W_counts >= 2     ~ "2+ W",
      (5 - W_counts) >= 2 ~ "2+ S",
      TRUE               ~ "NotImportant"
    )
  ) %>%
  select(-left2, -refBase, -right2, -all_bases, -W_counts) %>%
  column_to_rownames(var = "context")
###
# plot PC variances
MutType_order <- c("CpG > TpG", "CpG > GpG", "CpG > ApG", "CHG > THG",
                "C > T", "C > G", "C > A", 
                "A > T", "A > C", "A > G")

loadings_5mers <- loadings_5mers %>%
  mutate(Type = factor(Type, levels = MutType_order)) %>%
  arrange(Type)

VarianceColoredLabels <- c(
    "CpG > TpG" = "#AC9ECEFF",
    "CpG > GpG" = "#6EC5ABFF",
    "CpG > ApG" = "#DAA5ACFF",
    "CHG > THG" = "#CB6BCEFF",
    "C > T" = "#B6E7E0FF",
    "C > G" = "#AA3F5DFF",
    "C > A" = "#98A54FFF",
    "A > T" = "#FFB651FF",
    "A > C" = "#D85A44FF",
    "A > G" = "#751C6DFF")

PC1_variance_bars_5mers <- ggplot() + 
  geom_bar(data=loadings_5mers,
           aes(x=factor(Variable, loadings_5mers$Variable), y=PC1, fill=Type), stat = "identity", width = 1) +
  scale_fill_manual(values = VarianceColoredLabels) +
  theme(axis.text.x = element_blank(),
        title = element_text(size = 20, color = "black"),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        axis.ticks.x = element_blank(),
        aspect.ratio = 0.2,
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.position = "none", #c(0.5, 0.87),
        legend.title = element_blank(),
        legend.background = element_rect(fill = "white", color = "snow4"),
        legend.direction = "horizontal",
        legend.text = element_text(size=15, color="black"),
        legend.box = "horizontal",
        legend.frame = element_rect(fill = "black")) +
  labs(title = "",
       x = "",
       y = paste0("PC1 (", pc1_var, "%)"))

PC2_variance_bars_5mers <- ggplot() + 
  geom_bar(data=loadings_5mers,
           aes(x=factor(Variable, loadings_5mers$Variable), y=PC2, fill=Type), stat = "identity", width = 1) +
  scale_fill_manual(values = VarianceColoredLabels) +
  theme(axis.text.x = element_blank(),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        axis.ticks.x = element_blank(),
        aspect.ratio = 0.2,
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.position = "none",
        legend.title = element_blank(),
        legend.background = element_rect(fill = "transparent")) +
  labs(title = "",
       x = "",
       y = paste0("PC2 (", pc2_var, "%)"))

PC2_variance_bars_5mers_C_to_T_nonCpG <- ggplot() + 
  geom_bar(data=loadings_5mers[loadings_5mers$Type == "C > T", ],
           aes(x=reorder(Variable, PC2), y=PC2, fill=Type), stat = "identity", width = 0.5) +
  scale_fill_manual(values = VarianceColoredLabels) +
  theme(axis.text.x = element_text(angle = 90, size = 8),
        panel.grid.minor = element_line(color = "snow4", linewidth = 0.5, linetype = 1),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.position = "none",
        legend.title = element_blank(),
        legend.background = element_rect(fill = "transparent")) +
  labs(title = "",
       x = "",
       y = paste0("PC2 (", pc2_var, "%)"))

PC2_variance_bars_5mers_C_to_T_atCpG <- ggplot() + 
  geom_bar(data=loadings_5mers[loadings_5mers$Type == "CpG > TpG", ],
           aes(x=reorder(Variable, PC2), y=PC2, fill=TypeOfCpG), stat = "identity", width = 0.5) +
  theme(axis.text.x = element_text(angle = 90, size = 8),
        panel.grid.minor = element_line(color = "snow4", linewidth = 0.5, linetype = 1),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.position = "top",
        legend.direction = "horizontal",
        legend.title = element_blank(),
        legend.background = element_rect(fill = "transparent")) +
  labs(title = "",
       x = "",
       y = paste0("PC2 (", pc2_var, "%)"))

PC2_variance_bars_5mers_A_to_G <- ggplot() + 
  geom_bar(data=loadings_5mers[loadings_5mers$Type == "A > G", ],
           aes(x=reorder(Variable, PC2), y=PC2, fill=Type), stat = "identity", width = 0.5) +
  scale_fill_manual(values = VarianceColoredLabels) +
  theme(axis.text.x = element_text(angle = 90, size = 8),
        panel.grid.minor = element_line(color = "snow4", linewidth = 0.5, linetype = 1),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.position = "none",
        legend.title = element_blank(),
        legend.background = element_rect(fill = "transparent")) +
  labs(title = "", #PC2 Variance Explained by Mutation Type
       x = "",
       y = paste0("PC2 (", pc2_var, "%)"))

PC3_variance_bars_5mers <- ggplot() + 
  geom_bar(data=loadings_5mers,
           aes(x=factor(Variable, loadings_5mers$Variable), y=PC3, fill=Type), stat = "identity", width = 1) +
  scale_fill_manual(values = VarianceColoredLabels) +
  theme(axis.text.x = element_blank(),
        title = element_text(size = 20, color = "black"),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        axis.ticks.x = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.position = c(0.5, 0.87),
        legend.title = element_blank(),
        legend.background = element_rect(fill = "white", color = "snow4"),
        legend.direction = "horizontal",
        legend.text = element_text(size=15, color="black"),
        legend.box = "horizontal",
        legend.frame = element_rect(fill = "black")) +
  labs(title = "",
       x = "",
       y = paste0("PC3 (", pc3_var, "%)"))

#c("#FF3200FF", "#E9A17CFF", "#E9E4A6FF", "#1BB6AFFF", "#0076BBFF", "#172869FF")

PC3_wilcoxon_test_CpG <- wilcox.test(loadings_5mers[loadings_5mers$Type == "CpG > TpG" & loadings_5mers$ForWilcoxonTestOfCpG==1, c("PC3")], loadings_5mers[loadings_5mers$Type == "CpG > TpG" & loadings_5mers$ForWilcoxonTestOfCpG==0, c("PC3")])
PC4_wilcoxon_test_CpG <- wilcox.test(loadings_5mers[loadings_5mers$Type == "CpG > TpG" & loadings_5mers$ForWilcoxonTestOfCpG==1, c("PC4")], loadings_5mers[loadings_5mers$Type == "CpG > TpG" & loadings_5mers$ForWilcoxonTestOfCpG==0, c("PC4")])

PC3_wilcoxon_test_CHG <- wilcox.test(loadings_5mers[loadings_5mers$Type == "CHG > THG" & loadings_5mers$ForWilcoxonTestOfCHG==1, c("PC3")], loadings_5mers[loadings_5mers$Type == "CHG > THG" & loadings_5mers$ForWilcoxonTestOfCHG==0, c("PC3")])
PC4_wilcoxon_test_CHG <- wilcox.test(loadings_5mers[loadings_5mers$Type == "CHG > THG" & loadings_5mers$ForWilcoxonTestOfCHG==1, c("PC4")], loadings_5mers[loadings_5mers$Type == "CHG > THG" & loadings_5mers$ForWilcoxonTestOfCHG==0, c("PC4")])

PC3_variance_bars_5mers_C_to_T_atCpG <- ggplot() + 
  geom_bar(data=loadings_5mers[loadings_5mers$Type == "CpG > TpG", ],
           aes(x=reorder(Variable, PC3), y=PC3, fill=TypeOfCpG_Binary), stat = "identity", width = 0.5) +
  theme(axis.text.x = element_blank(),
        title = element_text(size = 20, color = "black"),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        axis.ticks.x = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.position = c(0.5, 0.87),
        legend.title = element_text(size=16, color="black"),
        legend.background = element_rect(fill = "white", color = "snow4"),
        legend.direction = "horizontal",
        legend.text = element_text(size=15, color="black"),
        legend.box = "horizontal",
        legend.frame = element_rect(fill = "black")) +
  labs(title = "",
       x = "",
       y = paste0("PC3 (", pc3_var, "%)")) +
  scale_fill_manual(name = "CpG > TpG Type",
                    values = c("2+ S" = "#FF3200FF",
                               "2+ W" = "#172869FF")) +
  annotate("text", x = 48, y = -0.15, label = expression(paste(p == 2.98 %*% 10^-7)), hjust = 0, size = 7, color = "black")

PC3_variance_bars_5mers_C_to_T_atCHG <- ggplot() + 
  geom_bar(data=loadings_5mers[loadings_5mers$Type == "CHG > THG", ],
           aes(x=reorder(Variable, PC3), y=PC3, fill=TypeOfCHG_Binary), stat = "identity", width = 0.5) +
  theme(axis.text.x = element_blank(),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        axis.ticks.x = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.position = c(0.5, 0.87),
        legend.direction = "horizontal",
        legend.title = element_text(size=16, color="black"),
        legend.text = element_text(size=15, color="black"),
        legend.box = "horizontal",
        legend.background = element_rect(fill = "white", color = "snow4"),
        legend.frame = element_rect(fill = "black")) +
  labs(title = "",
       x = "",
       y = paste0("PC3 (", pc3_var, "%)")) +
  scale_fill_manual(name = "CHG > THG Type",
                    values = c("2+ S" = "#FF3200FF",
                               "2+ W" = "#172869FF")) +
  annotate("text", x = 36, y = -0.025, label = expression(paste(p == 5.19 %*% 10^-3)), hjust = 0, size = 7, color = "black")

PC3_variance_bars_5mers_C_to_A_atCH <- ggplot() + 
  geom_bar(data=loadings_5mers[loadings_5mers$Type == "C > A", ],
           aes(x=reorder(Variable, PC3), y=PC3, fill=TypeOfCH_to_A_Binary), stat = "identity", width = 0.5) +
  theme(axis.text.x = element_blank(),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        axis.ticks.x = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.position = c(0.5, 0.87),
        legend.direction = "horizontal",
        legend.title = element_text(size=16, color="black"),
        legend.text = element_text(size=15, color="black"),
        legend.box = "horizontal",
        legend.background = element_rect(fill = "white", color = "snow4"),
        legend.frame = element_rect(fill = "black")) +
  labs(title = "",
       x = "",
       y = paste0("PC3 (", pc3_var, "%)")) +
  scale_fill_manual(name = "CH > AH Type",
                    values = c("2+ S" = "#FF3200FF",
                               "2+ W" = "#172869FF"))

PC4_variance_bars_5mers <- ggplot() + 
  geom_bar(data=loadings_5mers,
           aes(x=factor(Variable, loadings_5mers$Variable), y=PC4, fill=Type), stat = "identity", width = 1) +
  scale_fill_manual(values = VarianceColoredLabels) +
  theme(axis.text.x = element_blank(),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        axis.ticks.x = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.position = "none",
        legend.title = element_blank(),
        legend.background = element_rect(fill = "transparent")) +
  labs(title = "",
       x = "",
       y = paste0("PC4 (", pc4_var, "%)"))

PC4_variance_bars_5mers_C_to_T_atCpG <- ggplot() + 
  geom_bar(data=loadings_5mers[loadings_5mers$Type == "CpG > TpG", ],
           aes(x=reorder(Variable, PC4), y=PC4, fill=TypeOfCpG_Binary), stat = "identity", width = 0.5) +
  theme(axis.text.x = element_blank(),
        title = element_text(size = 20, color = "black"),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        axis.ticks.x = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.background = element_rect(fill = "white", color = "snow4"),
        legend.direction = "horizontal",
        legend.text = element_text(size=15, color="black"),
        legend.box = "horizontal",
        legend.frame = element_rect(fill = "black"),
        legend.position = c(0.5, 0.87),
        legend.title = element_text(size=16, color="black")) +
  labs(title = "",
       x = "",
       y = paste0("PC4 (", pc4_var, "%)")) +
  scale_fill_manual(name = "CpG > TpG Type",
                    values = c("2+ S" = "#FF3200FF",
                               "2+ W" = "#172869FF")) +
  annotate("text", x = 48, y = -0.07, label = expression(paste(p == 9.79 %*% 10^-5)), hjust = 0, size = 7, color = "black")

PC4_variance_bars_5mers_C_to_T_atCHG <- ggplot() + 
  geom_bar(data=loadings_5mers[loadings_5mers$Type == "CHG > THG", ],
           aes(x=reorder(Variable, PC4), y=PC4, fill=TypeOfCHG_Binary), stat = "identity", width = 0.5) +
  theme(axis.text.x = element_blank(),
        title = element_text(size = 20, color = "black"),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        axis.ticks.x = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.background = element_rect(fill = "white", color = "snow4"),
        legend.direction = "horizontal",
        legend.text = element_text(size=15, color="black"),
        legend.box = "horizontal",
        legend.frame = element_rect(fill = "black"),
        legend.position = c(0.5, 0.87),
        legend.title = element_text(size=16, color="black")) +
  labs(title = "",
       x = "",
       y = paste0("PC4 (", pc4_var, "%)")) +
  scale_fill_manual(name = "CHG > THG Type",
                    values = c("2+ S" = "#FF3200FF",
                               "2+ W" = "#172869FF")) +
  annotate("text", x = 36, y = -0.03, label = expression(paste(p == 1.85 %*% 10^-3)), hjust = 0, size = 7, color = "black")

```
# mutation rates between weak and strong surrounding dna contexts
```{r}
longer_PCALoadings_data <- loadings_5mers

library(purrr)
library(magrittr)

longer_PCALoadings_data %<>%
  mutate(
    # Extract flanking context
    left_flank  = str_sub(Variable, 1, 2),
    right_flank = str_sub(Variable, 8, 9),
    # Split flanks into individual characters
    left_bases  = str_split(left_flank, "", simplify = FALSE),
    right_bases = str_split(right_flank, "", simplify = FALSE),
    # Choose number of right bases to include depending on BroadType
    flank_bases = pmap(
      list(left_bases, right_bases, Type),
      function(left, right, type) {
        if (type %in% c("C > G", "C > A", "A > C", "A > G", "A > T", "C > T")) {
          c(left, right)  # only use the first base on the right
        } else  if (type %in% c("CpG > GpG", "CpG > ApG", "CpG > TpG")) {
          c(left, right[2])
        } else if (type %in% c("CHG > THG")) {
          c(left, right[1])
        }
      }
    ),
    # Count weak bases (A or T) in the flanks only (excluding the reference base)
    W_count = map_int(flank_bases, ~ sum(.x %in% c("A", "T"))),
    # Assign based on W/S counts
    weakBaseCount = as.character(W_count)
  ) %>%
  select(-left_flank, -right_flank, -left_bases, -right_bases, -flank_bases, -W_count)

New_PC3_plots <- ggplot() + 
  geom_bar(data=longer_PCALoadings_data,
           aes(x=reorder(Variable, PC3), y=PC3, fill=weakBaseCount), alpha = 1, stat = "identity", width = 1) + #reorder(Variable, PC3)
  theme(strip.text = element_text(size = 14, color = "black", face = "bold"),
        strip.background = element_rect(fill = "white", color = "black"),
        axis.text.x = element_blank(),
        aspect.ratio = 0.5,
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        axis.ticks.x = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.position = "top",
        legend.direction = "horizontal",
        legend.title = element_text(size=16, color="black"),
        legend.text = element_text(size=15, color="black"),
        legend.box = "horizontal",
        legend.background = element_rect(fill = "white", color = "snow4"),
        legend.frame = element_rect(fill = "black")) +
  labs(title = "",
       x = "",
       y = paste0("PC3 (", pc3_var, "%)")) +
  scale_fill_manual(name = "Number of Weak Bases",
                    values = c("0" = "#FF3200FF",
                               "1" = "#FF3200FF",
                               "2" = "#172869FF",
                               "3" = "#172869FF",
                               "4" = "#172869FF")
                    ) +
  facet_wrap(~Type, scales = "free", ncol = 1)

New_PC4_plots <- ggplot() + 
  geom_bar(data=longer_PCALoadings_data,
           aes(x=reorder(Variable, PC4), y=PC4, fill=weakBaseCount), alpha = 1, stat = "identity", width = 1) + #reorder(Variable, PC4)
  theme(strip.text = element_text(size = 14, color = "black", face = "bold"),
        strip.background = element_rect(fill = "white", color = "black"),
        axis.text.x = element_blank(),
        aspect.ratio = 0.5,
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        axis.ticks.x = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.position = "top",
        legend.direction = "horizontal",
        legend.title = element_text(size=16, color="black"),
        legend.text = element_text(size=15, color="black"),
        legend.box = "horizontal",
        legend.background = element_rect(fill = "white", color = "snow4"),
        legend.frame = element_rect(fill = "black")) +
  labs(title = "",
       x = "",
       y = paste0("PC4 (", pc4_var, "%)")) +
  scale_fill_manual(name = "Number of Weak Bases",
                    values = c("0" = "#FF3200FF",
                               "1" = "#FF3200FF",
                               "2" = "#172869FF",
                               "3" = "#172869FF",
                               "4" = "#172869FF")
                    ) +
  facet_wrap(~Type, scales = "free", ncol = 1)

Longer_pca_data <- pivot_longer(data = pca_data, cols = contains(">"), names_to = "MutationContext", values_to = "MutationProbability")
Longer_pca_data$Type <- case_when(
  grepl("C>A\\]G", Longer_pca_data$MutationContext) ~ "CpG > ApG",
  grepl("C>T\\]G", Longer_pca_data$MutationContext) ~ "CpG > TpG",
  grepl("C>G\\]G", Longer_pca_data$MutationContext) ~ "CpG > GpG",
  grepl("C>T\\]AG", Longer_pca_data$MutationContext) ~ "CHG > THG",
  grepl("C>T\\]CG", Longer_pca_data$MutationContext) ~ "CHG > THG",
  grepl("C>T\\]TG", Longer_pca_data$MutationContext) ~ "CHG > THG",
  grepl("C>A", Longer_pca_data$MutationContext) ~ "C > A",
  grepl("C>T", Longer_pca_data$MutationContext) ~ "C > T",
  grepl("C>G", Longer_pca_data$MutationContext) ~ "C > G",
  grepl("A>G", Longer_pca_data$MutationContext) ~ "A > G",
  grepl("A>C", Longer_pca_data$MutationContext) ~ "A > C",
  grepl("A>T", Longer_pca_data$MutationContext) ~ "A > T",
  TRUE ~ "NA"
)

Longer_pca_data %<>%
  mutate(
    # Extract flanking context
    left_flank  = str_sub(MutationContext, 1, 2),
    right_flank = str_sub(MutationContext, 8, 9),
    # Split flanks into individual characters
    left_bases  = str_split(left_flank, "", simplify = FALSE),
    right_bases = str_split(right_flank, "", simplify = FALSE),
    # Choose number of right bases to include depending on BroadType
    flank_bases = pmap(
      list(left_bases, right_bases, Type),
      function(left, right, type) {
        if (type %in% c("C > G", "C > A", "A > C", "A > G", "A > T", "C > T", "CHG > THG", "CpG > GpG", "CpG > ApG")) {
          c(left, right)  # only use the first base on the right
        } else  {
          c(left, right[2])
        }
      }
    ),
    # Count weak bases (A or T) in the flanks only (excluding the reference base)
    W_count = map_int(flank_bases, ~ sum(.x %in% c("A", "T"))),
    # Assign based on W/S counts
    weakBaseCount = as.character(W_count)
  ) %>%
  select(-left_flank, -right_flank, -left_bases, -right_bases, -flank_bases, -W_count)

Longer_pca_data_mean <- aggregate(MutationProbability ~ weakBaseCount + Species + Type + Class, data = Longer_pca_data, FUN = mean)
Longer_pca_data_sd <- aggregate(MutationProbability ~ weakBaseCount + Species + Type + Class, data = Longer_pca_data, FUN = sd)
Longer_pca_data_mean$ProbSD <- Longer_pca_data_sd$MutationProbability

Longer_pca_data_mean$WvsS <- case_when(
  Longer_pca_data_mean$Type =="CpG > TpG" & Longer_pca_data_mean$weakBaseCount == "0" ~ "2+ S",
  Longer_pca_data_mean$Type =="CpG > TpG" & Longer_pca_data_mean$weakBaseCount == "1" ~ "2+ S",
  Longer_pca_data_mean$Type =="CpG > TpG" & Longer_pca_data_mean$weakBaseCount == "2" ~ "2+ W",
  Longer_pca_data_mean$Type =="CpG > TpG" & Longer_pca_data_mean$weakBaseCount == "3" ~ "2+ W",
  Longer_pca_data_mean$Type =="CpG > TpG" & Longer_pca_data_mean$weakBaseCount == "4" ~ "2+ W",
  
  Longer_pca_data_mean$Type =="CpG > ApG" & Longer_pca_data_mean$weakBaseCount == "0" ~ "2+ S",
  Longer_pca_data_mean$Type =="CpG > ApG" & Longer_pca_data_mean$weakBaseCount == "1" ~ "2+ S",
  Longer_pca_data_mean$Type =="CpG > ApG" & Longer_pca_data_mean$weakBaseCount == "2" ~ "2+ W",
  Longer_pca_data_mean$Type =="CpG > ApG" & Longer_pca_data_mean$weakBaseCount == "3" ~ "2+ W",
  Longer_pca_data_mean$Type =="CpG > ApG" & Longer_pca_data_mean$weakBaseCount == "4" ~ "2+ W",
  
  Longer_pca_data_mean$Type =="CpG > GpG" & Longer_pca_data_mean$weakBaseCount == "0" ~ "2+ S",
  Longer_pca_data_mean$Type =="CpG > GpG" & Longer_pca_data_mean$weakBaseCount == "1" ~ "2+ S",
  Longer_pca_data_mean$Type =="CpG > GpG" & Longer_pca_data_mean$weakBaseCount == "2" ~ "2+ W",
  Longer_pca_data_mean$Type =="CpG > GpG" & Longer_pca_data_mean$weakBaseCount == "3" ~ "2+ W",
  Longer_pca_data_mean$Type =="CpG > GpG" & Longer_pca_data_mean$weakBaseCount == "4" ~ "2+ W",
  
  Longer_pca_data_mean$Type =="CHG > THG" & Longer_pca_data_mean$weakBaseCount == "0" ~ "2+ S",
  Longer_pca_data_mean$Type =="CHG > THG" & Longer_pca_data_mean$weakBaseCount == "1" ~ "2+ S",
  Longer_pca_data_mean$Type =="CHG > THG" & Longer_pca_data_mean$weakBaseCount == "2" ~ "2+ W",
  Longer_pca_data_mean$Type =="CHG > THG" & Longer_pca_data_mean$weakBaseCount == "3" ~ "2+ W",
  Longer_pca_data_mean$Type =="CHG > THG" & Longer_pca_data_mean$weakBaseCount == "4" ~ "2+ W",
  
  Longer_pca_data_mean$Type =="C > T" & Longer_pca_data_mean$weakBaseCount == "0" ~ "2+ S",
  Longer_pca_data_mean$Type =="C > T" & Longer_pca_data_mean$weakBaseCount == "1" ~ "2+ S",
  Longer_pca_data_mean$Type =="C > T" & Longer_pca_data_mean$weakBaseCount == "2" ~ "2+ W",
  Longer_pca_data_mean$Type =="C > T" & Longer_pca_data_mean$weakBaseCount == "3" ~ "2+ W",
  Longer_pca_data_mean$Type =="C > T" & Longer_pca_data_mean$weakBaseCount == "4" ~ "2+ W",
  
  Longer_pca_data_mean$Type =="C > A" & Longer_pca_data_mean$weakBaseCount == "0" ~ "2+ S",
  Longer_pca_data_mean$Type =="C > A" & Longer_pca_data_mean$weakBaseCount == "1" ~ "2+ S",
  Longer_pca_data_mean$Type =="C > A" & Longer_pca_data_mean$weakBaseCount == "2" ~ "2+ W",
  Longer_pca_data_mean$Type =="C > A" & Longer_pca_data_mean$weakBaseCount == "3" ~ "2+ W",
  Longer_pca_data_mean$Type =="C > A" & Longer_pca_data_mean$weakBaseCount == "4" ~ "2+ W",
  
  Longer_pca_data_mean$Type =="C > G" & Longer_pca_data_mean$weakBaseCount == "0" ~ "2+ S",
  Longer_pca_data_mean$Type =="C > G" & Longer_pca_data_mean$weakBaseCount == "1" ~ "2+ S",
  Longer_pca_data_mean$Type =="C > G" & Longer_pca_data_mean$weakBaseCount == "2" ~ "2+ W",
  Longer_pca_data_mean$Type =="C > G" & Longer_pca_data_mean$weakBaseCount == "3" ~ "2+ W",
  Longer_pca_data_mean$Type =="C > G" & Longer_pca_data_mean$weakBaseCount == "4" ~ "2+ W",
  
  Longer_pca_data_mean$Type =="A > T" & Longer_pca_data_mean$weakBaseCount == "0" ~ "2+ S",
  Longer_pca_data_mean$Type =="A > T" & Longer_pca_data_mean$weakBaseCount == "1" ~ "2+ S",
  Longer_pca_data_mean$Type =="A > T" & Longer_pca_data_mean$weakBaseCount == "2" ~ "2+ W",
  Longer_pca_data_mean$Type =="A > T" & Longer_pca_data_mean$weakBaseCount == "3" ~ "2+ W",
  Longer_pca_data_mean$Type =="A > T" & Longer_pca_data_mean$weakBaseCount == "4" ~ "2+ W",
  
  Longer_pca_data_mean$Type =="A > C" & Longer_pca_data_mean$weakBaseCount == "0" ~ "2+ S",
  Longer_pca_data_mean$Type =="A > C" & Longer_pca_data_mean$weakBaseCount == "1" ~ "2+ S",
  Longer_pca_data_mean$Type =="A > C" & Longer_pca_data_mean$weakBaseCount == "2" ~ "2+ W",
  Longer_pca_data_mean$Type =="A > C" & Longer_pca_data_mean$weakBaseCount == "3" ~ "2+ W",
  Longer_pca_data_mean$Type =="A > C" & Longer_pca_data_mean$weakBaseCount == "4" ~ "2+ W",
  
  Longer_pca_data_mean$Type =="A > G" & Longer_pca_data_mean$weakBaseCount == "0" ~ "2+ S",
  Longer_pca_data_mean$Type =="A > G" & Longer_pca_data_mean$weakBaseCount == "1" ~ "2+ S",
  Longer_pca_data_mean$Type =="A > G" & Longer_pca_data_mean$weakBaseCount == "2" ~ "2+ W",
  Longer_pca_data_mean$Type =="A > G" & Longer_pca_data_mean$weakBaseCount == "3" ~ "2+ W",
  Longer_pca_data_mean$Type =="A > G" & Longer_pca_data_mean$weakBaseCount == "4" ~ "2+ W",
  TRUE ~ "NA"
)

summarized_Longer_pca_data_mean <- aggregate(MutationProbability ~ WvsS + Species + Type + Class, data = Longer_pca_data_mean, FUN = mean)
summarized_wider_pca_data_mean <- pivot_wider(data = summarized_Longer_pca_data_mean, names_from = "WvsS", values_from = "MutationProbability")
summarized_wider_pca_data_mean$W_over_S_ratio <- summarized_wider_pca_data_mean$`2+ W` / summarized_wider_pca_data_mean$`2+ S`

Mammal_CpGtoT_test <- t.test(x = log((summarized_wider_pca_data_mean[summarized_wider_pca_data_mean$Class=="Mammalia" & summarized_wider_pca_data_mean$Type=="CpG > TpG",])$W_over_S_ratio), mu = 0)
#wilcox.test(x = log((summarized_wider_pca_data_mean[summarized_wider_pca_data_mean$Class=="Mammalia" & summarized_wider_pca_data_mean$Type=="CpG > TpG",])$W_over_S_ratio), mu = 0, alternative = "two.sided")

WvsS_Ttest_df <- data.frame(matrix(ncol = 10, nrow = 50))
colnames(WvsS_Ttest_df) <- c("Clade", "MutationType", "Mean_W_over_Strong", "LowerBound", "UpperBound", "Pvalue", "WilcoxonEstimate", "WilcoxonLower", "WilcoxonUpper", "WilcoxonPvalue")
RowCounter <- 1
for (i in c("Mammalia", "Plantae", "Aves", "Osteichthyes", "Invertebrata")) {
  for (j in c("CpG > TpG", "CHG > THG", "CpG > ApG", "CpG > GpG", "C > T", "C > A", "C > G", "A > C", "A > G", "A > T")) {
    ttest_results <- t.test(x = log((summarized_wider_pca_data_mean[summarized_wider_pca_data_mean$Class==i & summarized_wider_pca_data_mean$Type==j,])$W_over_S_ratio), mu = 0)
    wolcoxonresults <- wilcox.test(x = log((summarized_wider_pca_data_mean[summarized_wider_pca_data_mean$Class==i & summarized_wider_pca_data_mean$Type==j,])$W_over_S_ratio), mu = 0, alternative = "two.sided", conf.int = TRUE)
    
    WvsS_Ttest_df[RowCounter,"Clade"] <- i
    WvsS_Ttest_df[RowCounter, "MutationType"] <- j
    WvsS_Ttest_df[RowCounter,"Mean_W_over_Strong"] <- ttest_results$estimate
    WvsS_Ttest_df[RowCounter, "LowerBound"] <- ttest_results$conf.int[1]
    WvsS_Ttest_df[RowCounter, "UpperBound"] <- ttest_results$conf.int[2]
    WvsS_Ttest_df[RowCounter, "Pvalue"] <- ttest_results$p.value
    WvsS_Ttest_df[RowCounter, "WilcoxonEstimate"] <- wolcoxonresults$estimate
    WvsS_Ttest_df[RowCounter, "WilcoxonLower"] <- wolcoxonresults$conf.int[1]
    WvsS_Ttest_df[RowCounter, "WilcoxonUpper"] <- wolcoxonresults$conf.int[2]
    WvsS_Ttest_df[RowCounter, "WilcoxonPvalue"] <- wolcoxonresults$p.value
    
    RowCounter = RowCounter + 1
  }
}

testPC_plot_rates <- ggplot() +
  geom_point(data = summarized_wider_pca_data_mean, aes(x = factor(Class, rev(CladeOrder)), y = W_over_S_ratio, color = Class), alpha=0.6, pch = 16) +
  theme(aspect.ratio = 0.5,
        axis.text.x = element_text(angle = 35, hjust = 1, face = "italic", size = 12, color = "black"),
        axis.title.x = element_text(size = 12, color = "black"),
        axis.title.y = element_text(size = 12, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        strip.text = element_text(size = 16, color = "black", face = "bold"),
        strip.background = element_rect(fill = "white", color = "black"),
        panel.grid.minor = element_line(color = "snow3", linewidth = 0.45, linetype = 1),
        panel.grid.major.y = element_line(color = "snow4", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_line(color = "tan", linewidth = 0.2, linetype = 2),
        panel.background = element_rect(fill = "white"),
        plot.background = element_rect(color = "transparent", size = 2),
        legend.position = "none",
        legend.background = element_rect(fill = "white", color = "black"),
        legend.direction = "horizontal",
        legend.text = element_text(size=10, color="black"),
        legend.title = element_text(size=12, color="black"),
        legend.key = element_rect(fill="transparent")) +
  scale_color_manual(values = ColoredClades) +
  labs(title = "",
       x = "",
       y = "Mutation Probability") +
  facet_wrap(~Type, scales = "free")

quickmutationtypeorder <- c("CpG > TpG", "CHG > THG")

w_versus_s_plot_rates <- ggplot() +
  geom_hline(yintercept = 1, linetype = 2, color = "black") +
  geom_errorbar(data = WvsS_Ttest_df[WvsS_Ttest_df$MutationType %in% c("CpG > TpG", "CHG > THG"), ], aes(x = factor(Clade, rev(CladeOrder)), ymin = exp(WilcoxonLower), ymax = exp(WilcoxonUpper), color = Clade), width = 0.2, alpha = 1) +
  scale_color_manual(values = ColoredClades) +
  geom_point(data = WvsS_Ttest_df[WvsS_Ttest_df$MutationType %in% c("CpG > TpG", "CHG > THG"), ], aes(x = factor(Clade, rev(CladeOrder)), y = exp(WilcoxonEstimate), fill = Clade, size = -log10(WilcoxonPvalue)), alpha=1, pch = 21) +
  theme(aspect.ratio = 0.6,
        axis.text.x = element_text(angle = 35, hjust = 1, face = "italic", size = 12, color = "black"),
        axis.title.x = element_text(size = 12, color = "black"),
        axis.title.y = element_text(size = 12, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        strip.text = element_text(size = 16, color = "black", face = "bold"),
        strip.background = element_rect(fill = "white", color = "black"),
        panel.grid.minor = element_line(color = "snow3", linewidth = 0.45, linetype = 1),
        panel.grid.major.y = element_line(color = "snow1", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_line(color = "tan", linewidth = 0.2, linetype = 2),
        panel.background = element_rect(fill = "white"),
        plot.background = element_rect(color = "transparent", size = 2),
        legend.position = "bottom",
        legend.box = element_blank(),
        legend.background = element_rect(fill = "white", color = "transparent"),
        legend.direction = "horizontal",
        legend.text = element_text(size=10, color="black"),
        legend.title = element_text(size=12, color="black"),
        legend.key = element_rect(fill="transparent")) +
  scale_fill_manual(values = ColoredClades) +
  labs(title = "",
       x = "",
       y = "[2+ W] / [2+ S]") +
  facet_wrap(~factor(MutationType, quickmutationtypeorder), scales = "free_x", ncol = 1) +
  scale_size(name = expression(paste("pvalue (-",log[10],")"))) +
  guides(color = guide_legend(nrow = 2), size = guide_legend(nrow=1))
```
# pca supplemental figure PC3 and 4 Fig. S 5
```{r}
panelAB <- plot_grid(pca_PC3_PC4_5mers_plot, plot_grid(PC3_variance_bars_5mers, PC4_variance_bars_5mers, ncol = 1), ncol = 2, labels = c("A", "B"), label_size = 24)
panelC1 <- plot_grid(PC3_variance_bars_5mers_C_to_T_atCpG, PC4_variance_bars_5mers_C_to_T_atCpG, ncol = 1, labels = c("C",""), label_size = 24)
panelC2 <- plot_grid(PC3_variance_bars_5mers_C_to_T_atCHG, PC4_variance_bars_5mers_C_to_T_atCHG, ncol = 1, labels = c("",""), label_size = 24)
PanelC <- plot_grid(panelC1, panelC2, ncol = 1)
panelCD <- plot_grid(PanelC, w_versus_s_plot_rates, ncol = 2, labels = c("","D"), label_size = 24)

pdf("supplemental_PCs_PCA_5mers.pdf", 18, 18)
plot_grid(panelAB, panelCD, ncol = 1, rel_heights = c(1,1.2))
dev.off()
```
# principal components and variable correlations Fig. S 8
```{r}
cor_data <- pca_data %>%
  summarise(across(starts_with("PC"), 
                   list(N50 = ~ cor(., N50, use = "complete.obs"),
                        L50 = ~ cor(., L50, use = "complete.obs"),
                        NoSNPs = ~ cor(., NoSNPs, use = "complete.obs"),
                        CpGPhi = ~ cor(., CpGPhi, use = "complete.obs"),
                        CHGPhi = ~ cor(., CHGPhi, use = "complete.obs"),
                        Methyl = ~ cor(., Methyl, use = "complete.obs"),
                        RMSE = ~ cor(., RMSE, use = "complete.obs"),
                        CHGMethyl = ~ cor(., CHGMethyl, use = "complete.obs"),
                        #BodyTemp = ~ cor(., BodyTemp, use = "complete.obs"),
                        #Thermy = ~ cor(., Thermy, use = "complete.obs"),
                        Random1 = ~ cor(., Random1, use = "complete.obs"),
                        Random2 = ~ cor(., Random2, use = "complete.obs")))) %>%
  pivot_longer(cols = everything(), 
               names_to = c("PC", "Variable"), 
               names_sep = "_", 
               values_to = "Correlation") %>%
  filter(PC %in% paste0("PC", 1:10))

PCs <- c("PC1", "PC2", "PC3", "PC4", "PC5", "PC6", "PC7", "PC8", "PC9", "PC10")
variables_order <- c("Methyl", "CHGMethyl", "CpGPhi", "CHGPhi", "NoSNPs", "RMSE", "N50", "L50", "Random1", "Random2")

pc_labels <- paste0("PC", 1:10, " (", 100*(pc_explained_variance_1_to_10), "%)")

correctvariablelabels <-  list("Methyl" = "mCpG %",
                            "CHGMethyl" = "mCHG %",
                            "CpGPhi" = "[C > T] - [CH > TH] (log)",
                            "CHGPhi" = "[C > T] - [CNH > TNH] (log)",
                            "NoSNPs" = "SNP Count (log)",
                            "RMSE" = "RMSE",
                            "N50" = "Contig N50",
                            "L50" = "Contig L50",
                            "Random1" = "Random Uniform Variable 1",
                            "Random2" = "Random Uniform Variable 2")

facet_labeller <- function(variable,value){
  return(correctvariablelabels[value])
}

correlations_to_5mers_PCs <- ggplot(data = cor_data, aes(x = factor(PC, PCs), y = Correlation, fill = Variable)) + 
  geom_bar(stat = "identity", width = 1) +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 14, color = "black"),
        title = element_text(size = 20, color = "black"),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        strip.text = element_text(size = 12, color = "black", face = "bold"),
        strip.background = element_rect(fill = "white", color = "black"),
        legend.position = "none",
        legend.title = element_blank(),
        legend.background = element_rect(fill = "white", color = "snow4"),
        legend.direction = "horizontal",
        legend.text = element_text(size=15, color="black"),
        legend.box = "horizontal",
        legend.frame = element_rect(fill = "black")) +
  labs(title = "",
       x = "",
       y = "Pearson correlation") +
  scale_x_discrete(labels = pc_labels) +
  scale_fill_manual(values = c("Methyl" = "seagreen", "CHGMethyl" = "seagreen",
                               "CpGPhi" = "goldenrod", "CHGPhi" = "goldenrod",
                               "NoSNPs" = "orchid", "RMSE" = "orchid", "N50" = "orchid", "L50" = "orchid",
                               "Random1" = "firebrick", "Random2" = "firebrick"),
                    labels = c("Methyl" = "mCpG %", "CHGMethyl" = "mCHG %", "CpGPhi" = "[C > T] - [CH > TH] (log)",
                               "CHGPhi" = "[C > T] - [CNH > TNH] (log)",
                               "NoSNPs" = "SNP Count (log)", "RMSE" = "RMSE", "N50" = "Contig N50", "L50" = "Contig L50",
                               "Random1" = "Random Uniform Variable 1", "Random2" = "Random Uniform Variable 2")) +
  facet_wrap(~factor(Variable, variables_order), ncol = 2, labeller = facet_labeller)


pdf("Correlations_with_PCs_5mers.pdf", 12, 8)
correlations_to_5mers_PCs
dev.off()
```
# 5-mer fitted probability heatmap and profile summary
```{r}
Mer5_pca_data <- pca_data
Mer5_pca_data <- Mer5_pca_data %>% mutate(Class = factor(Class, levels = rev(CladeOrder)))

MutType_order

CpGType_order <- loadings_5mers[loadings_5mers$Type == "CpG > TpG", ][order(loadings_5mers[loadings_5mers$Type == "CpG > TpG", ]$PC3),]$Variable

longer_5mer_pca_data <- pivot_longer(Mer5_pca_data,
                                     cols = contains(">"),
                                     names_to = "MutationContext",
                                     values_to = "FittedProb")

species_order <- tree$tip.label

longer_5mer_pca_data <- longer_5mer_pca_data %>%
  mutate(Species = factor(Species, levels = species_order))

longer_5mer_pca_data$MutType <- case_when(
  grepl("C>A\\]G", longer_5mer_pca_data$MutationContext) ~ "CpG > ApG",
  grepl("C>T\\]G", longer_5mer_pca_data$MutationContext) ~ "CpG > TpG",
  grepl("C>G\\]G", longer_5mer_pca_data$MutationContext) ~ "CpG > GpG",
  grepl("C>T\\]AG", longer_5mer_pca_data$MutationContext) ~ "CHG > THG",
  grepl("C>T\\]CG", longer_5mer_pca_data$MutationContext) ~ "CHG > THG",
  grepl("C>T\\]TG", longer_5mer_pca_data$MutationContext) ~ "CHG > THG",
  grepl("C>A", longer_5mer_pca_data$MutationContext) ~ "C > A",
  grepl("C>T", longer_5mer_pca_data$MutationContext) ~ "C > T",
  grepl("C>G", longer_5mer_pca_data$MutationContext) ~ "C > G",
  grepl("A>G", longer_5mer_pca_data$MutationContext) ~ "A > G",
  grepl("A>C", longer_5mer_pca_data$MutationContext) ~ "A > C",
  grepl("A>T", longer_5mer_pca_data$MutationContext) ~ "A > T",
  TRUE ~ "NA"
)

longer_5mer_pca_data <- longer_5mer_pca_data %>%
  arrange(factor(MutType, levels = MutType_order)) %>%
  mutate(MutationContext = factor(MutationContext, levels = unique(MutationContext)))

longer_5mer_pca_data <- longer_5mer_pca_data %>%
  group_by(MutationContext) %>%
  mutate(FittedProb_scaled = (FittedProb - min(FittedProb)) / (max(FittedProb) - min(FittedProb))) %>%
  ungroup()

ColoredClades
longer_5mer_pca_data <- longer_5mer_pca_data %>%
  mutate(CladeColumn = "CladeColor")

longer_5mer_pca_data <- longer_5mer_pca_data %>%
  mutate(MutColorColumn = "MutColor")
longer_5mer_pca_data <- longer_5mer_pca_data %>%
  mutate(MutColorColumnB = "MutColorB")
longer_5mer_pca_data <- longer_5mer_pca_data %>%
  mutate(MutColorColumnC = "MutColorC")
longer_5mer_pca_data <- longer_5mer_pca_data %>%
  mutate(MutColorColumnD = "MutColorD")

quickheat <- ggplot(longer_5mer_pca_data) +
  geom_tile(aes(x = CladeColumn, y = Species, fill = Class), width = 50, position = position_nudge(x = 26)) +
  scale_fill_manual(values = ColoredClades, name = "Clade") +
  new_scale_fill() +
  geom_tile(aes(x = MutationContext, y = MutColorColumn, fill = MutType), width = 1) +
  geom_tile(aes(x = MutationContext, y = MutColorColumnB, fill = MutType), width = 1) +
  geom_tile(aes(x = MutationContext, y = MutColorColumnC, fill = MutType), width = 1) +
  geom_tile(aes(x = MutationContext, y = MutColorColumnD, fill = MutType), width = 1) +
  scale_fill_manual(values = VarianceColoredLabels, name = "Mutation Type") +
  new_scale_fill() +
  geom_tile(aes(x = MutationContext, y = Species, fill = log10(FittedProb))) +
  scale_fill_gradientn(colours = colorRampPalette(c("#085AFFFF","white", "#FF0000FF"))(50) ) +
  labs(x = "Mutation Type", y = "Clade", fill = expression(paste("Fitted Probabilities (",log[10],")"))) +
  theme(axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.title = element_text(size = 20, color = "black", face = "bold"),
        legend.text = element_text(size = 16, color = "black"),
        legend.title = element_text(size = 16, color = "black"),
        aspect.ratio = 1) +
  annotate("rect",
           xmin = 0, xmax = 65,
           ymin = 19, ymax = 63,
           color = "black", fill = NA, size = 1) +
  annotate("rect",
           xmin = 192, xmax = 241,
           ymin = 74, ymax = 106,
           color = "black", fill = NA, size = 1) +
  scale_x_discrete(position = "top")
```
# Figure 2 plotting
```{r}
pdf("Figure2_PCA_5mers.pdf", 20, 12)
plot_grid(plot_grid(quickheat, pca_5mers_plot, ncol = 2, labels = c("A", "B"), label_size = 24, rel_widths = c(1,0.8)), plot_grid(PC1_variance_bars_5mers, PC2_variance_bars_5mers, ncol = 2, labels = c("C", ""), label_size = 24), ncol = 1, rel_heights = c(1,0.25))
dev.off()
```
#7-mer PCA
```{r}
library(ggfortify)
Mer7_species_data <- read.delim('Merged_layer6_Probabilities.txt', sep = "\t", header = FALSE)
Mer7_species_data[1,2:(length(Mer7_species_data[1,]))] <- sapply(Mer7_species_data[1,2:(length(Mer7_species_data[1,]))], function(old_label) {
  if (old_label %in% names(species_label_map)) {
    return(species_label_map[[old_label]])
  } else {
    return(old_label)  # Keep the original label if not found in mapping
  }
})
df_Mer7_species_data <- Mer7_species_data[2:(length(Mer7_species_data[,1])),2:(length(Mer7_species_data[1,]))]
df_Mer7_species_data <- sapply(df_Mer7_species_data, as.numeric)

####
mat_sigfit_species_data <- as.matrix(df_Mer7_species_data)
# Normalize each column
for (i in 1:ncol(mat_sigfit_species_data)) {
  mat_sigfit_species_data[, i] <- mat_sigfit_species_data[, i] / sum(mat_sigfit_species_data[, i], na.rm = TRUE)
}

# Convert back to data frame
df_Mer7_species_data <- as.data.frame(mat_sigfit_species_data)
####

colnames(df_Mer7_species_data) <- c(Mer7_species_data[1,2:(length(Mer7_species_data[1,]))])
rownames(df_Mer7_species_data) <- c(Mer7_species_data[2:(length(Mer7_species_data[,1])),1])

df_Mer7_species_data <- t(df_Mer7_species_data)

df_Mer7_species_data <- as.data.frame(df_Mer7_species_data)

df_Mer7_species_data$Species <- as.vector(rownames(df_Mer7_species_data))
for_pca_plot <- merge(df_Mer7_species_data, classified_species, by.x="Species",by.y="Species", by.all=TRUE)
# remove bad species
for_pca_plot <- for_pca_plot[!(for_pca_plot$Species %in% excluded_species),]

for_pca_run <- for_pca_plot[,2:(length(Mer7_species_data[,1]))]

for_pca_run <- sapply(for_pca_run, as.numeric)
rownames(for_pca_run) <- for_pca_plot$Species


pca_test <- prcomp(for_pca_run)
pca_test_7mer <- pca_test

### manual plotting
pca_7mer_summary <- summary(pca_test)
var_explained <- pca_7mer_summary$importance[2, ]
pc1_var <- round(var_explained[1] * 100, 2)
pc2_var <- round(var_explained[2] * 100, 2)
pc3_var <- round(var_explained[3] * 100, 2)
pc4_var <- round(var_explained[4] * 100, 2)

pca_scores <- as.data.frame(pca_test$x)

# Combine with rate data
pca_data <- cbind(for_pca_plot, pca_scores)
pca_data$Methyl <- sorted_merged_summary_table[pca_data$Species,]$CpGtoTPhi
# Plot
#PC 1 and 2
pca_7mers_plot <- ggplot(pca_data, aes(x = PC1, y = PC2, color = Class)) +
  geom_hline(yintercept = 0, color = "tan", linetype = "dashed", linewidth = 0.5) + 
  geom_vline(xintercept = 0, color = "tan", linetype = "dashed", linewidth = 0.5) +
  geom_point(size = 5, alpha = 0.8, pch=16) +
  theme_minimal() +
  theme(panel.background = element_rect(fill = "white"),
        panel.grid = element_blank(),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        title = element_text(size = 22, face = "bold", color = "blueviolet"),
        legend.position = "bottom",
        legend.direction = "horizontal",
        legend.title = element_blank(),
        legend.text = element_text(size=16, color="black"),
        legend.background = element_rect(fill = "transparent"),
        aspect.ratio = 1,
        panel.border = element_blank()) +
  labs(title = "",
       x = paste0("PC1 (", pc1_var, "%)"),
       y = paste0("PC2 (", pc2_var, "%)")) +
  scale_color_manual(values = ColoredClades) +
  guides(color = guide_legend(nrow = 3))
###


loadings_7mers <- as.data.frame(pca_test_7mer$rotation)
loadings_7mers$Context <- "empty"
loadings_7mers$Variable <- rownames(loadings_7mers)
loadings_7mers$Context <- case_when(
  grepl("C>A]G", rownames(loadings_7mers)) ~ "CpG",
  grepl("C>T]G", rownames(loadings_7mers)) ~ "CpG",
  grepl("C>G]G", rownames(loadings_7mers)) ~ "CpG",
  TRUE ~ "NonCpG"
)

loadings_7mers$Type <- case_when(
  grepl("C>A\\]G", rownames(loadings_7mers)) ~ "C_to_A_at_CpG",
  grepl("C>T\\]G", rownames(loadings_7mers)) ~ "C_to_T_at_CpG",
  grepl("C>G\\]G", rownames(loadings_7mers)) ~ "C_to_G_at_CpG",
  grepl("C>T\\]AG", rownames(loadings_7mers)) ~ "C_to_T_at_CHG",
  grepl("C>T\\]CG", rownames(loadings_7mers)) ~ "C_to_T_at_CHG",
  grepl("C>T\\]TG", rownames(loadings_7mers)) ~ "C_to_T_at_CHG",
  grepl("C>A", rownames(loadings_7mers)) ~ "C_to_A",
  grepl("C>T", rownames(loadings_7mers)) ~ "C_to_T",
  grepl("C>G", rownames(loadings_7mers)) ~ "C_to_G",
  grepl("A>G", rownames(loadings_7mers)) ~ "A_to_G",
  grepl("A>C", rownames(loadings_7mers)) ~ "A_to_C",
  grepl("A>T", rownames(loadings_7mers)) ~ "A_to_T",
  TRUE ~ "NA"
)

# plot PC variances
MutType_order <- c("C_to_T_at_CpG", "C_to_G_at_CpG", "C_to_A_at_CpG", "C_to_T_at_CHG",
                "C_to_T", "C_to_G", "C_to_A", 
                "A_to_T", "A_to_C", "A_to_G")

loadings_7mers <- loadings_7mers %>%
  mutate(Type = factor(Type, levels = MutType_order)) %>%
  arrange(Type)

VarianceColoredLabels <- c(
    "C_to_T_at_CpG" = "#AC9ECEFF",
    "C_to_G_at_CpG" = "#6EC5ABFF",
    "C_to_A_at_CpG" = "#DAA5ACFF",
    "C_to_T_at_CHG" = "#CB6BCEFF",
    "C_to_T" = "#B6E7E0FF",
    "C_to_G" = "#AA3F5DFF",
    "C_to_A" = "#98A54FFF",
    "A_to_T" = "#FFB651FF",
    "A_to_C" = "#D85A44FF",
    "A_to_G" = "#751C6DFF")

VarianceModifiedLabels <- c("CpG > TpG", "CpG > GpG", "CpG > ApG", "CHG > THG", "C > T", "C > G", "C > A", "A > T", "A > C", "A > G")
PC1_variance_bars_7mers <- ggplot() + 
  geom_bar(data=loadings_7mers,
           aes(x=factor(Variable, loadings_7mers$Variable), y=PC1, fill=Type), stat = "identity", width = 1) +
  scale_fill_manual(labels = VarianceModifiedLabels,
                    values = VarianceColoredLabels) +
  theme(axis.text.x = element_blank(),
        title = element_text(size = 20, color = "black"),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.position = c(0.7, 0.87),
        legend.title = element_blank(),
        legend.background = element_rect(fill = "white", color = "snow4"),
        legend.direction = "horizontal",
        legend.text = element_text(size=11, color="black"),
        legend.box = "horizontal",
        legend.frame = element_rect(fill = "black")) +
  labs(title = "",
       x = "",
       y = paste0("PC1 (", pc1_var, "%)"))

PC2_variance_bars_7mers <- ggplot() + 
  geom_bar(data=loadings_7mers,
           aes(x=factor(Variable, loadings_7mers$Variable), y=PC2, fill=Type), stat = "identity", width = 1) +
  scale_fill_manual(labels = VarianceModifiedLabels,
                    values = VarianceColoredLabels) +
  theme(axis.text.x = element_blank(),
        axis.text = element_text(size = 18, color = "black"),
        axis.title = element_text(size = 20, color = "black"),
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white"),
        legend.position = "none",
        legend.title = element_blank(),
        legend.background = element_rect(fill = "transparent")) +
  labs(title = "",
       x = "",
       y = paste0("PC2 (", pc2_var, "%)"))

pdf("PCA_7mers.pdf", 18, 9)
plot_grid(pca_7mers_plot, plot_grid(PC1_variance_bars_7mers, PC2_variance_bars_7mers, ncol = 1), ncol = 2)
dev.off()
```
# fig S. 4
```{r}
pca_3mers_plot <- pca_3mers_plot + theme(legend.position = "none")
pca_5mers_plot <- pca_5mers_plot + theme(legend.position = "none")
pca_7mers_plot <- pca_7mers_plot + theme(legend.position = "none")
pdf("PCA_supplementalFigure.pdf", 18, 6)
plot_grid(pca_3mers_plot, pca_5mers_plot, pca_7mers_plot, ncol=3, labels = c("3-mers", "5-mers", "7-mers"), label_size = 20)
dev.off()
```

# plotting 5mer sigfit results
```{r}
library(sigfit)
ObservedMutationsMer5_Matrix <- readRDS("5mer_MutationCounts_allSpecies.rds")
OpportunitiesMatrixMer5 <- readRDS("5mer_OpportunitiesMatrix_allSpecies.rds")
mer5_mutationSpectra <- as.data.frame(ObservedMutationsMer5_Matrix/OpportunitiesMatrixMer5)

Context_PropostionsMatrixMer5 <- OpportunitiesMatrixMer5
for (i in 1:nrow(Context_PropostionsMatrixMer5)) {
  Context_PropostionsMatrixMer5[i,] <- (Context_PropostionsMatrixMer5[i,] * 3) / sum(Context_PropostionsMatrixMer5[i,])
}

Exposures_mer5_4_extractedSigns <- readRDS("New_Exposures_4_Sigs_extractedFrom5mers.rds")
Signatures_mer5_4_extractedSigns <- readRDS("New_Signatures_4Sigs_extractedFrom5mers.rds")
Reconstructions_mer5_4_extractedSigns <- readRDS("New_Reconstructions_4Sigs_extractedFrom5mers.rds")

CosSimDF_mer5_4_extractedSigns <- data.frame(matrix(ncol = 3, nrow = 108))
colnames(CosSimDF_mer5_4_extractedSigns) <- c("MeanSim", "LowerMeanSim", "UpperMeanSim")

for (i in seq(1, length(rownames(Reconstructions_mer5_4_extractedSigns$mean)))) {
  rownames(CosSimDF_mer5_4_extractedSigns)[i] <- rownames(Reconstructions_mer5_4_extractedSigns$mean)[i]
  CosSimDF_mer5_4_extractedSigns[i,1] <- cosine_sim(as.numeric(Reconstructions_mer5_4_extractedSigns$mean[i,]), as.numeric(ObservedMutationsMer5_Matrix[i,]))
  CosSimDF_mer5_4_extractedSigns[i,2] <- cosine_sim(as.numeric(Reconstructions_mer5_4_extractedSigns$lower[i,]), as.numeric(ObservedMutationsMer5_Matrix[i,]))
  CosSimDF_mer5_4_extractedSigns[i,3] <- cosine_sim(as.numeric(Reconstructions_mer5_4_extractedSigns$upper[i,]), as.numeric(ObservedMutationsMer5_Matrix[i,]))
}
CosSimDF_mer5_4_extractedSigns$Clade <- classified_species[rownames(CosSimDF_mer5_3_extractedSigns),]$Class

CladeOrder <- c("Protozoa", "Plantae", "Fungia", "Cnidaria", "Invertebrata", "Echinodermata", "Osteichthyes", "Amphibia", "Reptilia", "Aves", "Mammalia")

IndividualReconstructionPerformances_mer5_4_extractedSigns_CosinePlot <- ggplot() +
  geom_point(data = CosSimDF_mer5_4_extractedSigns, aes(x=factor(Clade, rev(CladeOrder)), y = MeanSim, color=Clade), size=8, alpha=0.6, pch=16) + #position = position_dodge(width = 1)
  theme(aspect.ratio = 0.8,
        axis.text.x = element_text(angle = 35, hjust = 1, face = "italic", size = 12, color = "black"),
        axis.title.x = element_text(size = 12, color = "black"),
        axis.title.y = element_text(size = 12, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        strip.text = element_text(size = 16, color = "black", face = "bold"),
        strip.background = element_rect(fill = "white", color = "black"),
        panel.grid.minor = element_line(color = "snow3", linewidth = 0.45, linetype = 1),
        panel.grid.major.y = element_line(color = "snow4", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_line(color = "tan", linewidth = 0.2, linetype = 2),
        panel.background = element_rect(fill = "white"),
        plot.background = element_rect(color = "transparent", size = 2),
        legend.position = "none",
        legend.background = element_rect(fill = "white", color = "black"),
        legend.direction = "horizontal",
        legend.text = element_text(size=10, color="black"),
        legend.title = element_text(size=12, color="black"),
        legend.key = element_rect(fill="transparent")) +
  scale_color_manual(values = ColoredClades) +
  labs(title = "",
       x = "",
       y = "Cosine similarity between observed and reconstructed spectra")
```
# Plot 5mer extracted signatures (sigfit)
```{r}
library(ggh4x)
strips = strip_themed(background_x = elem_list_rect(fill = c("#8BD4F7FF", "#D4F4DDFF", "#F6955EFF", "#9B6981FF")))
SignaColors <- c("Signature D" = "#8BD4F7FF", "Signature A" = "#D4F4DDFF", "Signature B" = "#F6955EFF", "Signature C" = "#9B6981FF")

TypeColoredLabels <- c(
    "CpG > TpG" = "#AC9ECEFF",
    "CpG > GpG" = "#6EC5ABFF",
    "CpG > ApG" = "#DAA5ACFF",
    "CHG > THG" = "#CB6BCEFF",
    "C > T" = "#B6E7E0FF",
    "C > G" = "#AA3F5DFF",
    "C > A" = "#98A54FFF",
    "A > T" = "#FFB651FF",
    "A > C" = "#D85A44FF",
    "A > G" = "#751C6DFF")

toPlot_extractedSigns <- (Signatures_mer5_4_extractedSigns$mean)
toPlot_extractedSigns$Signature <- rownames(toPlot_extractedSigns)
longer_toPlot_extractedSigns <- pivot_longer(data = toPlot_extractedSigns, cols = contains(">"), names_to = "MutationContext", values_to = "Intensity")

    # for 5-mers
longer_toPlot_extractedSigns$MutationType <- case_when(
  grepl("C>A\\]G", longer_toPlot_extractedSigns$MutationContext) ~ "CpG > ApG",
  grepl("C>T\\]G", longer_toPlot_extractedSigns$MutationContext) ~ "CpG > TpG",
  grepl("C>G\\]G", longer_toPlot_extractedSigns$MutationContext) ~ "CpG > GpG",
  grepl("C>T\\]AG", longer_toPlot_extractedSigns$MutationContext) ~ "CHG > THG",
  grepl("C>T\\]CG", longer_toPlot_extractedSigns$MutationContext) ~ "CHG > THG",
  grepl("C>T\\]TG", longer_toPlot_extractedSigns$MutationContext) ~ "CHG > THG",
  grepl("C>A", longer_toPlot_extractedSigns$MutationContext) ~ "C > A",
  grepl("C>T", longer_toPlot_extractedSigns$MutationContext) ~ "C > T",
  grepl("C>G", longer_toPlot_extractedSigns$MutationContext) ~ "C > G",
  grepl("A>G", longer_toPlot_extractedSigns$MutationContext) ~ "A > G",
  grepl("A>C", longer_toPlot_extractedSigns$MutationContext) ~ "A > C",
  grepl("A>T", longer_toPlot_extractedSigns$MutationContext) ~ "A > T",
  TRUE ~ "NA"
)

SpeciesMutType_order <- c("CpG > TpG", "CpG > ApG", "CpG > GpG", "CHG > THG", "C > T", "C > A", "C > G", "A > T", "A > C", "A > G")

longer_toPlot_extractedSigns <- longer_toPlot_extractedSigns %>%
    mutate(MutationType = factor(MutationType, levels = SpeciesMutType_order)) %>%
    arrange(MutationType)

longer_toPlot_extractedSigns$Variable <- rownames(longer_toPlot_extractedSigns)

correctSignaturelabels <-  list("Signature D" = "CpG Transitions",
                            "Signature A" = "CpG & CHG Transitions",
                            "Signature B" = "Background A",
                            "Signature C" = "Background B")

facet_labeller <- function(variable,value){
  return(correctSignaturelabels[value])
}

mer5_extractedSignaturesPlot <- ggplot() + 
  geom_bar(data=longer_toPlot_extractedSigns,
           aes(x=factor(Variable, longer_toPlot_extractedSigns$Variable), y=Intensity, fill=MutationType), stat = "identity", width = 1) +
  scale_fill_manual(values = TypeColoredLabels) +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        title = element_text(size = 28, color = "black"),
        axis.text = element_text(size = 12, color = "black"),
        axis.title = element_text(size = 14, color = "black"),
        strip.text = element_text(size = 14, color = "black", face = "bold"),
        strip.background = element_rect(fill = "white", color = "black"),
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_line(color = "snow3", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill = "white", color = "black"),
        legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(fill = "white", color = "white"),
        legend.direction = "horizontal",
        legend.text = element_text(size=11, color="black"),
        legend.box = "horizontal",
        legend.frame = element_rect(fill = "black")) +
  guides(fill = guide_legend(nrow = 1)) +
  labs(title = "",
       x = "mutation type",
       y = paste0("intensity")) +
  facet_wrap2(~factor(Signature, c("Signature D", "Signature A", "Signature B", "Signature C")), scales="free_x", labeller = facet_labeller, strip = strips, ncol=4)
```
# plotting scaled exposure profiles
```{r}
ContextScaledExposuresDf <- data.frame(matrix(ncol = 4, nrow = 108))
colnames(ContextScaledExposuresDf) <- c("Signature A", "Signature B", "Signature C", "Signature D")

for (i in seq(1, length(rownames(Context_PropostionsMatrixMer5)))) {
  rownames(ContextScaledExposuresDf)[i] <- rownames(Context_PropostionsMatrixMer5)[i]
  
  ScaledSignatures <- colSums(as.data.frame(t(Signatures_mer5_4_extractedSigns$mean)) * t(Context_PropostionsMatrixMer5[i,]))
  
  ReScaledExposures <- Exposures_mer5_4_extractedSigns$mean[i,] * ScaledSignatures
  NormalizedExposures <- ReScaledExposures / sum(ReScaledExposures)
  
  ContextScaledExposuresDf[i,] <- NormalizedExposures
}
exposures <- ContextScaledExposuresDf

forExposurePlot <- as.data.frame(exposures)
forExposurePlot$Species <- rownames(forExposurePlot)
forExposurePlot$Clade <- classified_species[forExposurePlot$Species,]$Class

specOrderExpo <- forExposurePlot$Species[order(forExposurePlot$Clade)]

ScaledLonger_forExposurePlot <- pivot_longer(forExposurePlot, cols = contains(c("Sig", "SBS")), names_to = "Signature", values_to = "Exposure")

ScaledLonger_forExposurePlot$Signature <- factor(ScaledLonger_forExposurePlot$Signature, levels=rev(c('Signature D','Signature A','Signature B', 'Signature C')))

Scaled_fittedSignaturesGGplot <- ggplot() + geom_bar(data = ScaledLonger_forExposurePlot, aes(fill=Signature, x=factor(Species, (Longer_forExposurePlot[Longer_forExposurePlot$Signature=="Signature D",] %>% arrange(Exposure))$Species), y=Exposure), position = "stack", stat = "identity") +
  theme(#axis.text.x = element_text(angle = 60, hjust = 1, face = "italic", size = 12, color = "black"),
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 14, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        panel.grid.minor = element_line(color = "snow3", linewidth = 0.45, linetype = 1),
        panel.grid.major.y = element_line(color = "snow4", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_line(color = "tan", linewidth = 0.2, linetype = 2),
        panel.background = element_rect(fill = "white"),
        plot.background = element_rect(color = "transparent", size = 2),
        strip.text = element_text(size = 10, color = "black", face = "bold"),
        strip.background = element_rect(fill = "white", color = "black"),
        legend.position = "bottom",
        legend.background = element_rect(fill = "white", color = "transparent"),
        legend.direction = "horizontal",
        legend.text = element_text(size=12, color="black"),
        legend.title = element_blank(),
        legend.key = element_rect(fill="transparent")) +
  #scale_color_manual(values = ColoredClades) +
  labs(title = "",
       x = "",
       y = "fitted signature exposure") +
  facet_wrap(~factor(Clade, rev(CladeOrder)),  ncol=11, strip.position = "top", scales="free_x", labeller = ) +
  scale_fill_manual(values = SignaColors,
                    labels = c("Signature A" = "CpG & CHG Transitions",
                               "Signature D" = "CpG Transitions",
                               "Signature B" = "Background A",
                               "Signature C" = "Background B"))

```
# Paper Figure 5
```{r}
pdf("Figure5_Sigfit.pdf", 13, 10)
plot_grid(mer5_extractedSignaturesPlot, Scaled_fittedSignaturesGGplot, ncol = 1, rel_heights = c(0.6, 1), labels = c("A", "B"), label_size = 24)
dev.off()
```
# sigfit performance curve
```{r}
# Spectrum reconstructions
Rec_mer5_2_extractedSigns <- readRDS("New_Reconstructions_2Sigs_extractedFrom5mers.rds")
Rec_mer5_3_extractedSigns <- readRDS("New_Reconstructions_3Sigs_extractedFrom5mers.rds")
Rec_mer5_4_extractedSigns <- readRDS("New_Reconstructions_4Sigs_extractedFrom5mers.rds")
Rec_mer5_5_extractedSigns <- readRDS("New_Reconstructions_5Sigs_extractedFrom5mers.rds")
Rec_mer5_6_extractedSigns <- readRDS("New_Reconstructions_6Sigs_extractedFrom5mers.rds")
Rec_mer5_7_extractedSigns <- readRDS("New_Reconstructions_7Sigs_extractedFrom5mers.rds")
Rec_mer5_8_extractedSigns <- readRDS("New_Reconstructions_8Sigs_extractedFrom5mers.rds")
Rec_mer5_9_extractedSigns <- readRDS("New_Reconstructions_9Sigs_extractedFrom5mers.rds")
Rec_mer5_10_extractedSigns <- readRDS("New_Reconstructions_10Sigs_extractedFrom5mers.rds")


CosSim_mer5_k_Signs_mean <- data.frame(matrix(ncol = 9, nrow = 108))
colnames(CosSim_mer5_k_Signs_mean) <- c("MeanSim_2", "MeanSim_3", "MeanSim_4", "MeanSim_5", "MeanSim_6", "MeanSim_7", "MeanSim_8", "MeanSim_9", "MeanSim_10")


for (i in seq(1, length(rownames(Rec_mer5_2_extractedSigns$mean)))) {
  rownames(CosSim_mer5_k_Signs_mean)[i] <- rownames(Rec_mer5_2_extractedSigns$mean)[i]
  CosSim_mer5_k_Signs_mean[i,1] <- cosine_sim(as.numeric(Rec_mer5_2_extractedSigns$mean[i,]), as.numeric(ObservedMutationsMer5_Matrix[i,]))
  CosSim_mer5_k_Signs_mean[i,2] <- cosine_sim(as.numeric(Rec_mer5_3_extractedSigns$mean[i,]), as.numeric(ObservedMutationsMer5_Matrix[i,]))
  CosSim_mer5_k_Signs_mean[i,3] <- cosine_sim(as.numeric(Rec_mer5_4_extractedSigns$mean[i,]), as.numeric(ObservedMutationsMer5_Matrix[i,]))
  CosSim_mer5_k_Signs_mean[i,4] <- cosine_sim(as.numeric(Rec_mer5_5_extractedSigns$mean[i,]), as.numeric(ObservedMutationsMer5_Matrix[i,]))
  CosSim_mer5_k_Signs_mean[i,5] <- cosine_sim(as.numeric(Rec_mer5_6_extractedSigns$mean[i,]), as.numeric(ObservedMutationsMer5_Matrix[i,]))
  CosSim_mer5_k_Signs_mean[i,6] <- cosine_sim(as.numeric(Rec_mer5_7_extractedSigns$mean[i,]), as.numeric(ObservedMutationsMer5_Matrix[i,]))
  CosSim_mer5_k_Signs_mean[i,7] <- cosine_sim(as.numeric(Rec_mer5_8_extractedSigns$mean[i,]), as.numeric(ObservedMutationsMer5_Matrix[i,]))
  CosSim_mer5_k_Signs_mean[i,8] <- cosine_sim(as.numeric(Rec_mer5_9_extractedSigns$mean[i,]), as.numeric(ObservedMutationsMer5_Matrix[i,]))
  CosSim_mer5_k_Signs_mean[i,9] <- cosine_sim(as.numeric(Rec_mer5_10_extractedSigns$mean[i,]), as.numeric(ObservedMutationsMer5_Matrix[i,]))
}
CosSim_mer5_k_Signs_mean$Clade <- classified_species[rownames(CosSim_mer5_k_Signs_mean),]$Class

longer_CosSim_mer5_k_Signs_mean <- pivot_longer(data = CosSim_mer5_k_Signs_mean, cols = contains("Mean"), names_to = "Model", values_to = "MeanSim")
mean_longer_CosSim_mer5_k_Signs_mean <- aggregate(MeanSim ~ Model + Clade, data = longer_CosSim_mer5_k_Signs_mean, FUN = mean)
min_longer_CosSim_mer5_k_Signs_mean <- aggregate(MeanSim ~ Model + Clade, data = longer_CosSim_mer5_k_Signs_mean, FUN = min)
max_longer_CosSim_mer5_k_Signs_mean <- aggregate(MeanSim ~ Model + Clade, data = longer_CosSim_mer5_k_Signs_mean, FUN = max)

mean_longer_CosSim_mer5_k_Signs_mean$MinSim <- min_longer_CosSim_mer5_k_Signs_mean$MeanSim
mean_longer_CosSim_mer5_k_Signs_mean$MaxSim <- max_longer_CosSim_mer5_k_Signs_mean$MeanSim
ModelOrder <- c("MeanSim_2", "MeanSim_3", "MeanSim_4", "MeanSim_5", "MeanSim_6", "MeanSim_7", "MeanSim_8", "MeanSim_9", "MeanSim_10")

ReconstructionPerformanceCurves_mer5_k_Signs <- ggplot(data = mean_longer_CosSim_mer5_k_Signs_mean) +
  geom_errorbar(aes(x = factor(Model, ModelOrder), ymin = MinSim, ymax = MaxSim, color = Clade), width = 0.2, alpha = 0.75) +
  geom_line(aes(x = factor(Model, ModelOrder), y = MeanSim, color = Clade, group = Clade), alpha=0.5) +
  geom_point(aes(x = factor(Model, ModelOrder), y = MeanSim, color = Clade), size=6, alpha=0.75, pch=16) +
  theme(aspect.ratio = 1,
        axis.text.x = element_text(size = 12, color = "black"),
        axis.title.x = element_text(size = 12, color = "black"),
        axis.title.y = element_text(size = 12, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        strip.text = element_text(size = 16, color = "black", face = "bold"),
        strip.background = element_rect(fill = "white", color = "black"),
        panel.grid.minor = element_line(color = "snow3", linewidth = 0.45, linetype = 1),
        panel.grid.major.y = element_line(color = "snow4", linewidth = 0.6, linetype = 1),
        panel.grid.major.x = element_line(color = "tan", linewidth = 0.2, linetype = 2),
        panel.background = element_rect(fill = "white"),
        plot.background = element_rect(color = "transparent", size = 2),
        legend.position = "none",
        legend.background = element_rect(fill = "white", color = "black"),
        legend.direction = "horizontal",
        legend.text = element_text(size=10, color="black"),
        legend.title = element_text(size=12, color="black"),
        legend.key = element_rect(fill="transparent")) +
  scale_color_manual(values = ColoredClades) +
  labs(title = "",
       x = "Number of extracted signatures",
       y = "Cosine similarity between observed and reconstructed spectra") +
  scale_x_discrete(labels = c("MeanSim_2" = "2",
                              "MeanSim_3" = "3",
                              "MeanSim_4" = "4",
                              "MeanSim_5" = "5",
                              "MeanSim_6" = "6",
                              "MeanSim_7" = "7",
                              "MeanSim_8" = "8",
                              "MeanSim_9" = "9",
                              "MeanSim_10" = "10")) +
  facet_wrap(~factor(Clade, rev(CladeOrder)))
  
  plot(ReconstructionPerformanceCurves_mer5_k_Signs)
  
pdf("~/Downloads/MathiesonLab/100SpeciesPaperFigures/Mer5_SigFit_k_Sign_Model_PerformanceCurve.pdf", 10, 10)
ReconstructionPerformanceCurves_mer5_k_Signs
dev.off()
```
